---
title: "Creating -Omics Data Objects"
author: "Gabriel Odom"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Create -Omics Data Objects}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      cache = FALSE,
                      comment = "#>")
```


# Overview
This vignette is the second chapter in the "Pathway Significance Testing with `pathwayPCA`" workflow. This guide will outline the major steps needed to create a data container object for analysis with the `pathwayPCA` package. These objects are called `Omics*`-class objects.

### Outline
Before we move on, we will outline our steps. We will

1. Define the four `Omics*` classes.
2. Create a few `Omics*` objects
3. Inspect and edit individual elements contained in these objects.

First, load the `pathwayPCA` package and the [`tidyverse` package suite](https://www.tidyverse.org/). If you don't have the `tidyverse` package suite, install that first.
```{r packageLoad}
# install.packages("tidyverse")
library(tidyverse)
library(pathwayPCA)
```

### Import Data
Because this is the second chapter in the workflow, we assume that

1. Your assay is "tidy".
2. Your gene pathway set / `gmt` information is stored in a `pathwaySet` object.
3. Your response and assay data have already been row-matched.

If you are unsure about any of the three points above (or you don't know what these mean), please review the [Import and Tidy Data](https://gabrielodom.github.io/pathwayPCA/articles/Importing_Data.html) vignette first. It isn't very long, but it will help you set up your data in the right way.

For the purpose of example, we will load an assay / response data frame and a `pathwaySet` list which already fits the three criteria above. This tidy data set has 656 gene expression measurements (columns) on 250 colon cancer patients (rows). Notice that the assay and survival response information have already been merged, so we have two additional columns (for Overall Survival Time and its corresponding death indicator).
```{r tumour_data_load}
data("colonSurv_df")
colonSurv_df
```

We also have a small list of 15 pathways which correspond to our example colon cancer assay.
```{r pathway_list_load}
data("colon_pathwaySet")
colon_pathwaySet
str(colon_pathwaySet$pathways, list.len = 10)
```

The pathway set list and tidy assay (with matched patient response information) are all the information we need to create an `Omics`-class data container. 

*******************************************************************************

</br>


# `Omics*`-Class Objects Defined
Now that we have our data loaded, we can create an analysis object for the `pathwayPCA` package.

### Class Overview
In this package, all primary input data will be in an `Omics*` data container object. There are four classes `Omics*` objects. Each class contains a tidy assay and `pathwaySet` gene set list. The classes differ in the type of response information they can hold. The classes, and their responses, are

1. `OmicsSurv`---a data container for survival information, which includes event time (the time of last follow-up with a subject) and event indicator (did the subject die, or was the observation right-censored?).
2. `OmicsReg`---a data container for continuous responses (usually a linear regression response).
3. `OmicsCateg`---a data container for categorical responses, the dependent variable of a generalized linear model. Currently, we only support binary classification (through logistic regression).
4. `OmicsPathway`---a data container with no response. This container isn't particularly useful by itself, but it is the "parent" class for the other three `Omics*` classes.

### Review of Data Types in `R`
Take a quick look back at the structure of our `colonSurv_df` object. We have an expression design matrix with the first two columns as subject response information. Look at the types of the columns of this data frame (directly under the column names, notice the `<dbl>` `<int>` tags). These tags tell us that the columns contain "double / numeric" (`dbl`) and "integer" (`int`) information. The other tags we could see here are `<chr>` (character), `<lgl>` (logical), or `<fct>` (factor). These tags are important because they identify which "class" of data is in each column. Here are some examples of how to change data between types:
```{r data_types}
head(colonSurv_df$OS_event)
head(as.character(colonSurv_df$OS_event))
head(as.logical(colonSurv_df$OS_event))
head(as.factor(colonSurv_df$OS_event))
```

We must put each response into a specific class:

- Survival data is stored with a pair of numeric (time) and logical (death indicator) vectors.
- Regression data is stored in a numeric or integer vector
- Binary classification data is stored in a factor.

These restrictions are on purpose: the data creation functions in the `pathwayPCA` package have very specific requirements about the types of data they take as inputs.

*******************************************************************************

</br>


# Create New `Omics*` Objects
Each response type has a specific creation function:

  - Survival response: Use the `create_OmicsSurv()` function to create an object with class `OmicsSurv`. This object will contain:
    + `eventTime`: a numeric vector of event times
    + `eventObserved`: a logical vector of death (or other event) indicators. This format precludes the option of recurrent-event survival analysis.
    + `assayData_df`: a tidy data frame of prediction information. Rows are observations or subjects; the columns are gene, protein, transcriptome, proteome, or metabolome measures. The column names *must* match a subset of the genes provided in the gene set list (in the `pathwaySet` object).
    + `pathwaySet`: a list of pathway information, as returned by the `read_gmt` function (see [the import vignette](https://gabrielodom.github.io/pathwayPCA/articles/Importing_Data.html#the-read_gmt-function) for more details). The names of the genes in these pathways *must* match a subset of the genes recorded in the prediction data frame (in the `assayData_df` object).
  - Regression response: Use the `create_OmicsReg()` function to create an object with class `OmicsReg`. This object will contain:
    + `response`: a numeric vector of the response
    + `assayData_df`: a tidy data frame of prediction information, as described above.
    + `pathwaySet`: a list of pathway information, as described above.
  - Binary Classification response: Use the `create_OmicsCateg()` function to create an object with class `OmicsCateg`. In future versions, this function will be able to take in n-ary responses and ordered categorical responses, but we only support binary responses for now. This object will contain:
    + `response`: a factor vector of the response
    + `assayData_df`: a tidy data frame of prediction information, as described above.
    + `pathwaySet`: a list of pathway information, as described above.


### Data Types and the `create_Omics*` Functions
As we previously mentioned, the response fields of the `create_OmicsSurv`, `create_OmicsReg`, or `create_OmicsCateg` functions require specific types of data:

* `create_OmicsSurv` requires two response vectors:
    - `eventTime_num` = a *numeric* vector. The column in your imported data corresponding to the event time should have the `<int>` or `<dbl>` tag. If this column is a character column (it has the `<chr>` tag), then the most likely cause is that you had an additional column of character information in your assay data frame before transposing. Remove this extra character information and transpose your assay again.
    - `eventObserved_lgl` = a *logical* vector. The column in your imported data corresponding to the event indicator should have the `<lgl>` tag. Also note that this indicator records if an event **was observed**, not if the event was censored. If the event occured at the corresponding event time (i.e. the subject died), then this vector entry will be `TRUE`. If the subject was right-censored, then this entry will be `FALSE`.
* `create_OmicsReg` and `create_OmicsCateg` require one response vector:
    - `response_num` = regression response as a *numeric* vector. The column in your imported data corresponding to the regression response should have the `<int>` or `<dbl>` tag. See the comment above about the `eventTime_num` argument if your column is a character.
    - `response_fact` = classification response as a *factor* vector. The column in your imported data corresponding to the classification / categorical regression response should have the `<fct>` tag. It is highly likely that the classification response was imported as an integer (if the original values were coded as 0/1) or a character (if the original values were given character names, such as "benign" / "malignent"). Either way, use the `as.factor` command to transform that column from a character or integer column into a factor column. For example, the `as.factor` function can turn our logical censoring indicator from a vector of `TRUE` and `FALSE` to a categorical (factor) vector with levels `TRUE` and `FALSE`. This will not overwrite the original column, so make sure you save this new factor.
```{r factors_eg, eval = FALSE}
eventObs_fct <- as.factor(colonSurv_df$OS_event)
```

We will consider the overall patient survival time (and corresponding censoring indicator) as our survival response, the event time as our regression response, and event indicator as our binary classification response (these are just for example).


### Create an `OmicsSurv` Survival Data Container
Now we are prepared to create our first `OmicsSurv` object for later analysis with either AES-PCA or Supervised PCA. Recall that `colonSurv_df` has the survival time in the first column, the event indicator in the second column, and the assay expression data in the subsequent columns. Therefore, the four arguments to the `create_OmicsSurv` function will have the following data objects supplied to them:

- `assayData_df` will be only the *expression columns* of the `colonSurv_df` data frame (i.e. all but the first two columns). In R, we can remove the first two columns of the `colonSurv_df` data frame by negative subsetting: `colonSurv_df[, -(1:2)]`.
- `pathwaySet_ls` will be the `colon_pathwaySet` list object. Recall that you can import a `.gmt` file into a `pathwaySet` object via the `read_gmt` function, or create a `pathwaySet` list object by hand with the `create_pathwaySet` function.
- `eventTime_num` will be the `numeric` survival time stored in the `OS_time` column of the `colonSurv_df` data frame: `colonSurv_df$OS_time`.
- `eventObserved_lgl` will be the `logical` event indicator stored in the `OS_event` column of the `colonSurv_df` data frame: `colonSurv_df$OS_event`.

```{r create_OmicsSurv_object}
colon_OmicsSurv <- create_OmicsSurv(assayData_df = colonSurv_df[, -(1:2)],
                                    pathwaySet_ls = colon_pathwaySet,
                                    eventTime_num = as.numeric(colonSurv_df$OS_time),
                                    eventObserved_lgl = as.logical(colonSurv_df$OS_event))
```

In order to view a summary of the contents of the `colon_OmicsSurv` object, you need simply to print it to the `R` console.
```{r print_colonOmicsSurv}
colon_OmicsSurv
```


### Regression (`OmicsReg`) and Classification (`OmicsCateg`) Data Containers
The `OmicsReg` (regression) and `OmicsCateg` (classification) objects are created almost identically to `OmicsSurv` objects. The only exceptions are how to input the response information:

- `response_num` will be some `numeric` dependent variable for regression.
- `response_fact` will be some `factor` dependent variable for (currently binomial) classification with a generalized linear model.

```{r create_OmicsReg_object, eval = FALSE}
colon_OmicsReg <- create_OmicsReg(assayData_df = colonSurv_df[, -(1:2)],
                                  pathwaySet_ls = colon_pathwaySet,
                                  response_num = as.numeric(colonSurv_df$OS_time))

colon_OmicsCateg <- create_OmicsCateg(assayData_df = colonSurv_df[, -(1:2)],
                                      pathwaySet_ls = colon_pathwaySet,
                                      response_fact = as.logical(colonSurv_df$OS_event))
```

*******************************************************************************

</br>


# Inspecting and Editing `Omics*`-Class Objects
In order to access or edit a specific component of an `Omics*` object, we need to use specific *accessor* functions.
