---
title: "Integrative Pathway Analysis with pathwayPCA"
author: "Lily Wang and Gabriel Odom"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 2
  word_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      cache = FALSE,
                      comment = "#>")
```


# 1. Introduction

`pathwayPCA` provides implementation of the methodogly described in Chen (2011) and Chen et al. (2010). `pathwayPCA` allows users to:

1. Test pathway association with binary, continuous, or survival phenotypes.
2. Extract relevant genes in the pathways using the SPCA and AESPCA approaches.
3. Compute PCAs based on the selected genes. These estimated latent variables represent pathway activities for individual subjects, which can then be used to perform integrative pathway analysis, such as multi-omics analysis, or predicting survival time.
4. Extract relevant genes that drive pathway significance as well as data corresponding to these relevant genes for additional *ad hoc* anlaysis.
5. Perform analyses with enhanced computational efficiency with parallel computing and enhanced data safety with S4-class data objects.
6. Can be extended to studies with complex experimental designs, with multiple covariates, and with interaction effects, e.g., testing whether pathway association with clinical phenotype is different between male and female subjects.


Throughout this vignette, we will make use of the `tidyverse` suite of utility packages (https://www.tidyverse.org/). The `tidyverse` and `pathwayPCA` can be loaded into R using: 
```{r packLoad, message=FALSE}
library(tidyverse)
library(pathwayPCA)
```

</br>

*******************************************************************************

</br>


# 2. Case study: identifying significant pathways in protein expressions associated with survival outcome in ovarian cancer data

## 2.1. Overian cancer dataset 
For this example, we will use the mass spectrometry based global proteomics data for ovarian cancer recently generated by the Clinical Proteomic Tumor Analysis Consortium (CPTAC). The normalized protein abundance expression dataset can be obtained from the LinkedOmics database at http://linkedomics.org/data_download/TCGA-OV/. We used the dataset “Proteome (PNNL, Gene level)” which was generated by the Pacific Northwest National Laboratory (PNNL). We then removed one subject due to missingness, we retained the least-missing 80% of proteins, and we imputed the remaining 4.5% of missing values via k-nearest-neighbors (as implemented in the Bioconductor package `impute` under default settings). This yields 83 complete observations of 5162 proteins.

## 2.2. Creating an `Omics` data object for pathway analysis
First, we need to create an `Omics`-class data object that stores

(1) the expression dataset 
(2) phenotype information for the samples
(3) a collection of pathways 

### 2.2.1 Expression and Phenotype Data
We can obtain (1) and (2) for the ovarian cancer dataset by loading the dataset `ovarian_PNNL_survival.RDS` included in the `pathwayPCA` package.
```{r loadAssay}
dataDir_path <- system.file(
  "extdata", package = "pathwayPCA", mustWork = TRUE
)
ovSurv_df <- readRDS(paste0(dataDir_path, "/ovarian_PNNL_survival.RDS"))
```

The `ovSurv_df` dataset is a data frame with protein expression levels and survival outcome matched by sample IDs. The variables (columns) include overall survival time and censoring status, as well as expression data for 5162 proteins for each of the 83 samples.
```{r printAssay, eval=FALSE}
ovSurv_df
```

### 2.2.2 Pathway Collections
For the collection of pathways in (3), we need to specify a `.gmt` file. Gene Matrix formatted files store each pathway as a single line. Each line contains an ID (column `TERMS`), an optional description (column `description`), and the genes in the pathway (all subsequent columns). Pathway collections in `.gmt` form can be downloaded from the MSigDB database at http://software.broadinstitute.org/gsea/msigdb/collections.jsp. Additionally, for [WikiPathways](https://www.wikipathways.org/index.php/WikiPathways), one can download monthly data releases in `.gmt` format using the `dowloadPathwayArchive()` function in the `rWikiPathways` package [from Bioconductor](https://bioconductor.org/packages/release/bioc/html/rWikiPathways.html). For example, the following commands downloads to your current directory the October 2017 release of WikiPathways:
```{r rWikiPathways, eval=FALSE}
library(rWikiPathways)
downloadPathwayArchive(
  date = "20171010", organism = "Homo sapiens", format = "gmt"
) 
```
```
trying URL 'http://data.wikipathways.org/20171010/gmt/wikipathways-20171010-gmt-Homo_sapiens.gmt'
Content type '' length 137254 bytes (134 KB)
downloaded 134 KB
```
```{r printWPgmtpath, echo=FALSE}
print("wikipathways-20171010-gmt-Homo_sapiens.gmt")
```

As an example, `pathwayPCA` includes the June 2018 Wikipathways collection for *homo sapiens*, which can be loaded using the read_gmt function:  
```{r ourWikiPWC}
wikipathways_PC <- read_gmt(
  paste0(dataDir_path, "/wikipathways_human_symbol.gmt"),
  description = TRUE
)
```

###  2.2.3 Create an `OmicsSurv` Data Container
Now that we have these three data components, we create an `OmicsSurv` data container. Note that when `assayData_df` and `response` are supplied from two different files, the user need to make sure these data sets are match by sample IDs.
```{r createOmics, comment=""}
ov_OmicsSurv <- CreateOmics(
  # protein expression data
  assayData_df = ovSurv_df[, -(2:3)], 
  # pathway collection
  pathwayCollection_ls = wikipathways_PC,
  # survival phenotypes
  response = ovSurv_df[, 1:3],
  # phenotype is survival data
  respType = "survival",
  # retain pathways with > 5 proteins
  minPathSize = 5  
)
```

To see a summary  of the omics data object we just created, simply type the name of the object:
```{r inspectOmicsObj}
ov_OmicsSurv
```

## 2.3. Testing pathway association with phenotypes
### 2.3.1 Method Description
Once we have a valid `Omics` object, we can perform pathway analysis using the AES-PCA (Adaptive, Elastic-net, Sparse PCA) or Supervised PCA methodology described in Chen (2011) and Chen et al. (2010).

Briefly, in the AES-PCA method, we first extract *principal components* (PCs) representing activities within each pathway using a dimension reduction approach based on adaptive, elastic-net, sparse principal component analysis (https://doi.org/10.2202/1544-6115.1697). The estimated latent variables are then tested against phenotypes using a permutation test that permutes sample labels. Note that the AESPCA approach does not use the response information to estiamte pathway PCs, so it is an unsupervised approach.

This is in contrast to the Supervised PCA approach, where a selected subset of genes most associated with disease outcome are used to estimate the latent variable for a pathway (https://doi.org/10.1002/gepi.20532). Because of this gene selection step, the test statistics from the Supervised PCA model can no longer be approximated well using the Student's $t$-distribution. To account for the gene selection step, `pathwayPCA` estimates $p$-values from a two-component mixture of Gumbel exteme value distributions instead.

### 2.3.2 Implementation 
Because the syntax for performing Supervised PCA is nearly identical to the AES-PCA syntax, we will illustrate only the AES-PCA workflow below. Note that when the value supplied to the `numReps` argument is greater than 0, the `AESPCA_pvals()` function employs a parametric test when estimating pathway significance via the following model: "phenotype ~ PC1". Pathway $p$-values are estimated based on a likelihood ratio test that compares this model to a null model (with intercept only). Because pathways should represent a single biological effect, we recommend extracting only one PC be estimated (i.e., `numPCs = 1`). Additionally, we can specify a method to control the false discovery rate (FDR) or family-wise error rate with the `adjustment` argument. 
```{r aespcaCall, comment=""}
ovarian_aespcOut <- AESPCA_pVals(
  # The Omics data container
  object = ov_OmicsSurv,
  # One principal component per pathway
  numPCs = 1,
  # Use parallel computing with 2 cores
  parallel = TRUE,
  numCores = 2,
  # Estimate the p-values parametrically
  numReps = 0,
  # Control FDR via Benjamini-Hochberg
  adjustment = "BH"
)
```

This `ovarian_aespcOut` object contains 3 components: a table of pathway $p$-values, AESPCA-estimated PCs of each sample from each pathway, and the loadings of each protein onto the AESPCs.
```{r inspectAESPCAout}
names(ovarian_aespcOut)
```

### 2.2.3 Show the Top Pathways
For this ovarian cancer dataset, the top 10 most significant pathways identified by AES-PCA are:
```{r headPVals}
ovarian_aespcOut$pVals_df
```

In this table, `n_tested` indicates the number of genes in both the pathway and assay data, i.e. the number of genes in a pathway that we actually tested for association with phenotype in AESPCA. 

We are interested in graphing the top dozen or so pathways by their level of significance as well. This code takes in the pathway $p$-values data frame from the AESPCA method output and [gathers](https://rpubs.com/mm-c/gather-and-spread) it into a "tidy" data frame (compatible with `ggplot`). We also add on a column for the negative natural logarithm of the pathway $p$-values (called `score`) and recode the label of the $p$-value adjustment method.
```{r subset_top_genes}
ovOutGather_df <-
  # Take in the results data frame,
  ovarian_aespcOut$pVals_df %>%
  # add the score variable,
  mutate(score = -log(rawp)) %>%
  # subset the top 20 pathways, and
  slice(1:20) %>% 
  # remove the columns we don't need
  select(-pathways, -n_tested, -rawp, -FDR_BH)
```

Now we plot the pathway significance level for the top 20 pathways.
```{r pathway_bar_plots, fig.height = 6, fig.width = 10.5}
ggplot(ovOutGather_df) +
  # set overall appearance of the plot
  theme_bw() +
  # Define the dependent and independent variables
  aes(x = reorder(terms, score), y = score) +
  # From the defined variables, create a vertical bar chart
  geom_col(position = "dodge", fill = "#F47321") +
  # Add pathway labels
  geom_text(
    aes(x = reorder(terms, score), label = reorder(description, score), y = 2),
    color = "white",
    size = 3
  ) +
  # Set main and axis titles
  ggtitle("AES-PCA Significant Pathways: Ovarian Cancer") +
  xlab("Pathways") +
  ylab("Negative Log p-Value") +
  # Add a line showing the alpha = 0.01 level
  geom_hline(yintercept = -log(0.01), size = 2, color = "#005030") +
  # Flip the x and y axes
  coord_flip()
```

### 2.2.4 Extract relevant genes from significant pathways  
Because pathways are defined *a priori*---typically within each pathway---only a subset of genes are relevant to the phenotype and contribute to pathway significance. In AESPCA, these relevant genes are the genes with nonzero loadings in the first PC of AESPCs.

For example, for the "IL-1 signaling pathway" (Wikipathways [WP195](https://www.wikipathways.org/index.php/Pathway:WP195)), we can extract the principal components and their protein loadings (PC & L) using the `getPathPCLs()` function:
```{r getWP195pcl}
wp195PCLs_ls <- getPathPCLs(ovarian_aespcOut, "WP195")
wp195PCLs_ls
```

From this output list, we can see the data frame of loadings. We will delete proteins with loadings that are equal to 0.
```{r WP195loads}
wp195Loadings_df <- 
  wp195PCLs_ls$Loadings %>% 
  filter(PC1 != 0)
wp195Loadings_df
```

We can also graph these loadings in a barchart. We will first add an indicator for inhibition.
```{r loadDirections}
wp195Loadings_df <- 
  wp195Loadings_df %>% 
  # Sort Loading from Strongest to Weakest
  arrange(desc(abs(PC1))) %>% 
  # Order the Genes by Loading Strength
  mutate(featureID = factor(featureID, levels = featureID)) %>% 
  # Add Directional Indicator (for Colour)
  mutate(Direction = factor(ifelse(PC1 > 0, "Up", "Down"))) 
```

Now we will construct a **col**umn chart with `ggplot2`'s `geom_col()` function.
```{r graphLoads, fig.height = 3.6, fig.width = 6.5}
ggplot(data = wp195Loadings_df) +
  # Set overall appearance
  theme_bw() +
  # Define the dependent and independent variables
  aes(x = featureID, y = PC1, fill = Direction) +
  # From the defined variables, create a vertical bar chart
  geom_col(width = 0.5, fill = "#005030", color = "#f47321") +
  # Set main and axis titles
  labs(
    title = "Gene Loadings on IL-1 Signaling Pathway",
    x = "Protein"
  ) +
  # Remove the legend
  guides(fill = FALSE)
```

Alternatively, we can also plot the correlation of each gene with first PC for each gene. These correlations can be computed by using the `TidyCorrelation()` function in Section 3.3. of additional vignette “Chapter 5 - Visualizing the Results”.

### 2.2.5 Subject-specific PCA estimates   
In the study of complex diseases, there is often considerable heterogeneity among different subjects with regard to underlying causes of disease and benefit of particular treatment. Therefore, in addition to identifying disease-relevant pathways for the entire patient group, successful (personalized) treatment regimes will also depend upon knowing if a particular pathway is dysregulated for an individual patient.

To this end, we can also assess subject-specific pathway activity. As we saw earlier, the `getPathPCLs()` function also returns subject-specific estimates for the individual pathway PCs.
```{r wp195PCs, fig.height = 3.6, fig.width = 6.5}
ggplot(data = wp195PCLs_ls$PCs) +
  # Set overall appearance
  theme_classic() + 
  # Define the independent variable
  aes(x = V1) +
  # Add the histogram layer
  geom_histogram(bins = 10, color = "#005030", fill = "#f47321") +
  # Set main and axis titles
  labs(
    title = "Distribution of Sample-specific Estimate of Pathway Activities",
    subtitle = paste0(wp195PCLs_ls$pathway, ": ", wp195PCLs_ls$description),
    x = "PC1 Value for Each Sample",
    y = "Count"
  )
```

### 2.2.6 Extract analysis dataset for significant pathways   
Users are often also interested in examining the actual data used for analysis of the top pathways, especially for the relevant genes with the pathway. To extract this dataset, use the `SubsetPathwayData()` function on the `Omics` data container.
```{r subsetWP195data}
wp195Data_df <- SubsetPathwayData(ov_OmicsSurv, "WP195")
wp195Data_df
```

We can use this data subset to perform pasthway-specific analysis to estimate the effect of individual genes.
```{r dataSubsetMod}
library(survival)
NFKB1_df <- 
  wp195Data_df %>% 
  select(EventTime, EventObs, NFKB1)

wp195_mod <- coxph(
  Surv(EventTime, EventObs) ~ NFKB1,
  data = NFKB1_df
)

summary(wp195_mod)
```

Additionally, we can estimate Kaplan-Meier survival curves based on up-regulation or down-regulation of NFKB1.
```{r NFKB1Surv}
# Add the direction
NFKB1_df <- 
  NFKB1_df %>% 
  mutate(NFKB1_Expr = ifelse(NFKB1 > median(NFKB1), "High", "Low"))

# Fit the survival model
NFKB1_fit <- survfit(
  Surv(EventTime, EventObs) ~ NFKB1_Expr,
  data = NFKB1_df
)
```

Finally, we can plot these K-M curves over NFKB1 expression.
```{r NFKB1SurvPlot, message=FALSE}
library(survminer)

ggsurvplot(
  NFKB1_fit, conf.int = FALSE, 
  title = "Kaplan-Meier Survival Curves for Ovarian Cancer by NFKB1 Expression",
  xlab = "Time in Days",
  palette = c("#f47321", "#005030")
)
```

</br>

*******************************************************************************

</br>


# 3. Case study: integrative multi-omics pathway analysis of ovarian cancer data
While copy number alterations are common genomic aberrations in ovarian carcer, recent studies have shown these changes do not necessarily lead to concordant changes in protein expression. In section 2.2.4 above, we illustrated testing pathway activities in protein expression against survival outcome. In this section, we will additionally test pathway activities in copy number against survial outcome. Moreover, we will perform integrative analysis to identify those survial-associated protein pathways driven by copy number alterations.

## 3.1 Creating copy number `Omics` data object for pathway analysis
We can identify copy number (CNV) pathways significantly associated with survival in the same way as we did for protein expressions. This CNV data is from from **DATA SOURCE HERE**.

```{r CNVsetup}
copyNumberClean_df <- readRDS(paste0(dataDir_path, "/OV_surv_x_CNV2.RDS"))
```

And now we create an `Omics` data container.
```{r CNVomics, eval=FALSE}
ovCNV_Surv <- CreateOmics(
  assayData_df = copyNumberClean_df[, -(2:3)],
  pathwayCollection_ls = wikipathways_PC,
  response = copyNumberClean_df[, 1:3],
  respType = "survival",
  minPathSize = 5
)
```
```
  ======  Creating object of class OmicsSurv  =======
The input pathway database included 5831 unique features.
The input assay dataset included 24776 features.
Only pathways with at least 5 or more features included in the assay dataset are
  tested (specified by minPathSize parameter). There are 424 pathways which meet
  this criterion.
Because pathwayPCA is a self-contained test (PMID: 17303618), only features in
  both assay data and pathway database are considered for analysis. There are 5637
  such features shared by the input assay and pathway database.
```

Finally, we can apply the AESPCA method to this copy-number data container. Due to the large sample size, this will take a few moments.
```{r CNV_aespca, eval=FALSE}
ovCNV_aespcOut <- AESPCA_pVals(
  object = ovCNV_Surv,
  numPCs = 1,
  parallel = TRUE,
  numCores = 2,
  numReps = 0,
  adjustment = "BH"
)
```
```
Part 1: Calculate Pathway AES-PCs
Initializing Computing Cluster: DONE
Extracting Pathway PCs in Parallel: DONE

Part 2: Calculate Pathway p-Values
Initializing Computing Cluster: DONE
Extracting Pathway p-Values in Parallel: DONE

Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame
DONE
```
```{r CNV_aespca_read, echo=FALSE}
ovCNV_aespcOut <- readRDS("ovarian_copyNum_aespcOut.RDS")
```

<!-- The copy number data frame is 109.6 Mb; the copy number `OmicsSurv` is 112.3 Mb; and the copy number `aespcOut` is 3.8 Mb (*and* it takes 3 minutes to compute). We're saving the output. -->

## 3.2 Identifying significant pathways and relevant genes in both CNV and protein level
Next, we identify the intersection of significant pathways based on both CNV and protein data. First, we will create a data frame of the pathway $p$-values from both CNV and proteomics.
```{r multiOmicsPvals}
# Copy Number
CNVpvals_df <- 
  ovCNV_aespcOut$pVals_df %>% 
  mutate(rawp_CNV = rawp) %>% 
  select(description, rawp_CNV) %>% 
  filter(rawp_CNV < 0.05)

# Proteomics
PROTpvals_df <- 
  ovarian_aespcOut$pVals_df %>% 
  mutate(rawp_PROT = rawp) %>% 
  select(terms, description, rawp_PROT) %>% 
  filter(rawp_PROT < 0.05)

# Intersection
SigBoth_df <- inner_join(PROTpvals_df, CNVpvals_df, by = "description")
# WnT Signaling Pathway is listed as WP363 and WP428
```

<!-- We can also create a Venn diagram of these two sets. -->
<!-- ```{r VennOverlap, message=FALSE} -->
<!-- library(VennDiagram) -->

<!-- venn.diagram( -->
<!--   list(CopyNum = CNVpvals_df$description, Protein = PROTpvals_df$description), -->
<!--   filename = "Figures/signif_Paths_Venn.png", -->
<!--   imagetype = "png", -->
<!--   col = c("#f47321", "#005030") -->
<!-- ) -->
<!-- ``` -->

The results showed there are 23 pathways significantly associated with survival in both CNV and protein data. Here are the top-10 most significant pathways (sorted by protein data significance): 
```{r intersectPaths}
SigBoth_df
```

As discussed in section 2.2.3, we can also identify relevant genes with nonzero loadings that drives pathway significance in both CNV and protein pathway analysis. The "IL-1 signaling pathway" (WP195) makes our cutoff for both CNV and protein data.
```{r wp195}
# Copy Number Loadings
CNVwp195_ls <- getPathPCLs(ovCNV_aespcOut, "WP195")
CNV195load_df <- 
  CNVwp195_ls$Loadings %>% 
  filter(abs(PC1) > 0) %>% 
  rename(PC1_CNV = PC1)

# Protein Loadings
PROTwp195_ls <- getPathPCLs(ovarian_aespcOut, "WP195")
PROT195load_df <- 
  PROTwp195_ls$Loadings %>% 
  filter(abs(PC1) > 0) %>% 
  rename(PC1_PROT = PC1)

# Intersection
inner_join(CNV195load_df, PROT195load_df)
```

[NFKB1](https://www.ncbi.nlm.nih.gov/pubmed/26345753) contributes both to CNV and protein pathway significance.

Because the proteomics data were recorded on a subset of the subjects shown in the copy number data, we can further examine the relationship between CNV and protein expressions for this gene.
```{r mergeGene}
NFKB1data_df <- inner_join(
  copyNumberClean_df %>% 
    select(Sample, NFKB1) %>% 
    rename(CNV_NFKB1 = NFKB1),
  ovSurv_df %>% 
    select(Sample, NFKB1) %>% 
    rename(PROT_NFKB1 = NFKB1)
)
```

A correlation test shows CNV is highly correlated with protein expression for this gene.
```{r NFKB1cor, comment=""}
NFKB1_cor <- cor.test(NFKB1data_df$CNV_NFKB1, NFKB1data_df$PROT_NFKB1)
NFKB1_cor
```

We can then visualize the multi-omics relationship for the NFKB1 gene using a scatterplot.
```{r NFKB1plot, fig.height = 5.6, fig.width = 6.5}
ggplot(data = NFKB1data_df) +
  # Set overall appearance
  theme_bw() +
  # Define the dependent and independent variables
  aes(x = CNV_NFKB1, y = PROT_NFKB1) +
  # Add the scatterplot
  geom_point(size = 2) +
  # Add a trendline
  geom_smooth(method = lm, se = FALSE, size = 2) +
  # Set main and axis titles
  labs(
    title = "NFKB1 Expression by Multi-Omics Data",
    x = "Copy Number",
    y = "Proteomics"
  ) +
  # Include the correlation on the graph
  annotate(
    geom = "text", x = 0.5, y = -0.6,
    label = paste("rho = ", round(NFKB1_cor$estimate, 3))
  ) +
  annotate(
    geom = "text", x = 0.5, y = -0.65,
    label = "p-val < 10^-5"
  )
```

## 3.3 An integrative view on patient-specific pathway activities 
In section 2.2.5 above, we have also seen that there can be considerable heterogeneity in pathway activities between the patients. One possible reason could be that copy number changes might not have lead to changes in protein expression for some of the patients. `pathwayPCA` can be used to estimate pathway activities for each patient, both for protein expressions and copy number expressions separately. These estimates can then be viewed jointly using a circo plot, as illustrated in section xx of the additional vignette “Chapter 5 - Visualizing the Results”.

**paste a picture here**


</br>

*******************************************************************************

</br>


# 4. Case study: analysis of studies with complex designs
`pathwayPCA` is capable of analyzing studies with multiple experimental factors. In this section, we illustrate using `pathwayPCA` to test differential association of pathway expression with survial outcome in male and female subjects.

## 4.1 Data setup and AESPCA analysis
For this example, we use TCGA KIRP RNAseq dataset downloaded from Xena datahub. First, we load the KIRP data, create an `Omics` data container, and extract first AESPC from each pathway-specific data subset.
```{r kidneyData}
# Load Data
kidney_df <- readRDS(paste0(dataDir_path, "/KIRP_Surv_RNAseq_inner.RDS"))
```

```{r kidneyOmics, eval=FALSE}
# Create Omics Container
kidney_Omics <- CreateOmics(
  assayData_df = kidney_df[, -(2:4)],
  pathwayCollection_ls = wikipathways_PC,
  response = kidney_df[, 1:3],
  respType = "surv",
  minPathSize = 5
)
```
```
  ======  Creating object of class OmicsSurv  =======
The input pathway database included 5831 unique features.
The input assay dataset included 20211 features.
Only pathways with at least 5 or more features included in the assay dataset are
  tested (specified by minPathSize parameter). There are 423 pathways which meet
  this criterion.
Because pathwayPCA is a self-contained test (PMID: 17303618), only features in
  both assay data and pathway database are considered for analysis. There are 5566
  such features shared by the input assay and pathway database.
```
```{r kidneyAESPCA, eval=FALSE}
# AESPCA
kidney_aespcOut <- AESPCA_pVals(
  object = kidney_Omics,
  numPCs = 1,
  parallel = TRUE,
  numCores = 2,
  numReps = 0,
  adjustment = "BH"
)
```
```
Part 1: Calculate Pathway AES-PCs
Initializing Computing Cluster: DONE
Extracting Pathway PCs in Parallel: DONE

Part 2: Calculate Pathway p-Values
Initializing Computing Cluster: DONE
Extracting Pathway p-Values in Parallel: DONE

Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame
DONE
```
```{r kidneyAESPCAload, echo=FALSE}
kidney_aespcOut <- readRDS("KIRP_protein_aespcOut.RDS")
```

<!-- This takes 1 minute and 40 seconds.  -->

## 4.2 Test for sex interaction with first PC
Next we can test whether there is differential pathway association with survival outcome for males and females by the following model,
\[
h(t) = h_0(t) \exp\left[ \beta_0 + \beta_1\text{PC}_1 + \beta_2\text{male} + \beta_3(\text{PC}_1 \times \text{male}) \right].
\]

In order to test the sex interaction effect for all pathways, we will write a function which tests the interaction for a single pathway-subset of the data.
```{r sexInteractFun}
TestIntxn <- function(pathway, pcaOut, resp_df){
  
  # For the given pathway, extract the PCs and loadings from the pcaOut list
  PCL_ls <- getPathPCLs(pcaOut, pathway)
  # Select and rename the PC
  PC_df <- PCL_ls$PCs %>% select(PC1 = V1)
  # Bind this PC to the phenotype data
  data_df <- bind_cols(resp_df, PC_df)
  
  # Construct a survival model with sex interaction
  sex_mod <- coxph(
    Surv(time, status) ~ PC1 + male + PC1 * male, data = data_df
  )
  # Extract the model fit statistics for the interaction
  modStats_mat <- t(
    coef(summary(sex_mod))["PC1:maleTRUE", ]
  )
 
  # Return a data frame of the pathway and model statistics
  list(
    statistics = data.frame(
      terms = pathway,
      description = PCL_ls$description,
      modStats_mat,
      stringsAsFactors = FALSE
    ),
    model = sex_mod
  )

}
```

To make sure this function gives us the information we need, we can test it on one pathway.
```{r testNewFun}
TestIntxn("WP195", kidney_aespcOut, kidney_df[, 2:4])$model
```

Now we can apply this function across the list of pathways, and wrangle the results into a sorted data frame.
```{r applyTestSexFun}
paths_char <- kidney_aespcOut$pVals_df$terms
interactions_ls <- lapply(
    paths_char,
    FUN = TestIntxn,
    pcaOut = kidney_aespcOut,
    resp_df = kidney_df[, 2:4]
)
names(interactions_ls) <- paths_char

interactions_df <- 
  interactions_ls %>% 
  lapply(`[[`, 1) %>% 
  bind_rows() %>% 
  as.tibble() %>% 
  arrange(`Pr...z..`)

interactions_df %>% 
  filter(`Pr...z..` < 0.05)
```

The results showed the most significant pathway is [WP1559](https://www.wikipathways.org/index.php/Pathway:WP1559) (“TFs Regulate miRNAs related to cardiac hypertrophy”). We can inspect the model results for this pathway directly.
```{r modelRes}
summary(interactions_ls[["WP1559"]]$model)
```

The results showed that although sex is not significantly associated with survial outcome, the association of pathway gene expression (PC1) with survival is highly dependent on sex of the samples.

## 4.3 Survival curves by sex interaction
We can also divide subjects according to sex and high or low PC-expression, and add a group indicator for the four groups.
```{r kidneySurvPlotData}
# Bind the Pheno Data to WP1559's First PC
kidneyWP1559_df <- bind_cols(
  kidney_df[, 2:4],
  getPathPCLs(kidney_aespcOut, "WP1559")$PCs %>% select(PC1 = V1)
)

# Add Grouping Feature
kidneySurvWP1559grouped_df <- 
  kidneyWP1559_df %>% 
  # add strength indicator
  mutate(direction = ifelse(PC1 > median(PC1), "High", "Low")) %>% 
  # group by interaction of sex and strength on PC
  mutate(group = paste0(direction, ifelse(male, "_Male", "_Female"))) %>% 
  # remove summarized columns
  select(-male, -PC1, -direction)
```

Now we can plot survial curves for the four groups.
```{r plotKidneyGroupedSurv, message=FALSE, fig.height = 6, fig.width = 10.5}
# Fit the survival model
fit <- survfit(
  Surv(time, status) ~ group,
  data = kidneySurvWP1559grouped_df
)

ggsurvplot(
  fit, conf.int = FALSE, 
  title = "Kaplan-Meier Curves of Kidney Cancer Survival by Sex * PC Interaction",
  xlab = "Time in Days",
  ylim = c(0.4, 1)
)
```

As shown by these Kaplan-Meier curves, for the "TFs Regulate miRNAs related to cardiac hypertrophy” pathway, first-PC expression above the median was more strongly associated with shorter survival times than those below the median. Moreover, this effect appears stronger for females than males.

</br>

*******************************************************************************

</br>


# 5. Conclusion
Now that you have an idea of what questions can be answered with this package, please see each of our supplementary vignette chapters for detailed and thorough commentary and guiding information on each of the three topics discussed herein. The vignettes are:

- [Chapter 2: Import Data](https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html)
- [Chapter 3: Create `Omics` Data Objects](https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html)
- [Chapter 4: Test Pathway Significance](https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html)
- [Chapter 5: Visualizing the Results](https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html)

Here is the R session information for this vignette:
```{r sessionDetails}
sessionInfo()
```
