---
title: "pathwayPCA Guide for Advanced R Users"
author: "Gabriel Odom"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette:
    toc: true
  word_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Advanced User's Guide}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE,
                      cache = FALSE,
                      comment = "#>")
```

# Overview of `pathwayPCA`
## Motivation
For over six decades, the scientific community has been on a quest to find how gene expression relates to particular clinical outcomes. For many years, this research was focused on detecting single-gene markers. However, clinical results of knock-down experiments of single genes have shown mixed results. Often, when one gene is knocked down by a treatment compound, one or two other genes increase their expression to compensate. Additionally, while testing all genes in an assay, genes with biological significance but weaker statistical significance often go undetected due to controlling the Type-I error. These constraints motivate concurrent testing the clinical effect of groups of genes with similar biological effect.

These bundles of genes, called *pathways*, define *molecular signatures* related to a particular disease. In recent years, we have seen practical benefits of testing clinical or experimental pathway effects. One such approach is to group the expression features per subject by pathway, then apply *principal components analysis* (PCA) to the expression-subset data matrix in order to extract a single hidden genetic effect. This is similar in philosophy to the setup of a *structural equation model*, in the sense that we expect genes or proteins in a single pathway to be biologically related to one another. Once we have calculated the eigen-decomposition of the pathway-specific data matrix subset, we then test the association of the first few principal components (PCs) of this subset with the clinical outcome.

The effectiveness of this process depends on the quality of the pathways used to group the genes. Some well-regarded pathway collections are the [Canonical Pathways](http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=CP), [Wikipathways](https://www.wikipathways.org/), or [Hallmark](http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=H) gene sets. After we have selected a pathway appropriate to the data assay, we repeat the above process (extracting a PC from each pathway-specific subset of the assay) for each pathway in the specified pathway collection. After adjusting the $p$-values from such a test for multiple comparisons, we return a table of the "most significant" pathways based on the strength of the relationship between pathway PCs and the outcome.

In summary, given a proteomic or genomic expression data set and a collection of molecular signatures (pathway collection), we can use this package to test the association between the expression data for each pathway and a given phenotypical response. 


## Necessary Data and Packages
In order to execute the code in this vignette, you will need the following packages.
```{r load_packages, message=FALSE}
library(tidyverse)
library(pathwayPCA)
```

Additionally, you will need

1. an $n \times p$ data frame of expression data with gene / protein names as the column names,
2. a pathway collection in `.gmt` format with gene / protein names matching the measurement names of the expression data frame, and
3. a $p$-dimensional vector (or matrix) of regression / classification (or survival) response data.

</br>

*******************************************************************************

</br>


# Pathways Sets and the `.gmt` File Format
The most common form of pathway collection files, especially from the Molecular Signatures Database, is the `.gmt` file. These files adhere to a [very specific and rigid](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) data format. Use the `read_gmt` function to import a pathway collection and store it as an object of class `pathwayCollection` in `R`. We will import the June 2018 pathway collection from Wikipathways for *homo sapiens*. The `pathwayPCA` package also has a copy of the [MSigDB Canonical Pathways version 6.0](http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=CP) pathway collection, and this package will update these pathway collections with each major Bioconductor release.
```{r read_gmt}
gmt_path <- system.file("extdata", "wikipathways_human_symbol.gmt",
                         package = "pathwayPCA", mustWork = TRUE)
wikipathways_PC <- read_gmt(gmt_path, description = TRUE)
```


## Inspect a `pathwayCollection`
In a `.gmt` file, the first column maps to the `TERMS` element, the second column maps to the (optional) `description` element (this is why the `read_gmt()` function has an option to read in the description column or skip it), and all subsequent columns map to the list within the `pathways` element. This shows list structure of the imported `pathwayCollection` object.
```{r view_pathwayCollection}
wikipathways_PC
```

To view the information related to an individual pathway, use the "double-subset" operator with the pathway name. For example, we can see the details of pathway "WP23" by
```{r seeWP23_1}
wikipathways_PC[["WP23"]]
```

This shows us the "name" of the pathway ("WP23"), the gene symbols in this pathway (shown in the `IDs` element), the description of the pathway ("B Cell Receptor Signaling Pathway"), and the number of genes in the pathway (97).


## `rWikiPathways`
To download a `.gmt` file from Wikipathways, we recommend the `R` package [`rWikiPathways`](https://bioconductor.org/packages/release/bioc/html/rWikiPathways.html). From their vignette:

> WikiPathways also provides a monthly data release archived at http://data.wikipathways.org. The archive includes GPML, GMT and SVG collections by organism and timestamped. There’s an R function for grabbing files from the archive...
>
>    `downloadPathwayArchive()`
> 
> This will simply open the archive in your default browser so you can look around (in case you don’t know what you are looking for). By default, it opens to the latest collection of GPML files. However, if you provide an organism, then it will download that file to your current working directory or specified destpath. For example, here’s how you’d get the latest GMT file for mouse:
>
>    `downloadPathwayArchive(organism = "Mus musculus", format = "gmt")`
>
> And if you might want to specify an archive date so that you can easily share and reproduce your script at any time in the future and get the same result. Remember, new pathways are being added to WikiPathways every month and existing pathways are improved continuously!
>
>    `downloadPathwayArchive(date = "20171010", organism = "Mus musculus", format = "gmt")`
>


</br>

*******************************************************************************

</br>


# Creating Data Objects
In order to analyze the expression and response data with respect to a given `pathwayCollection` object, first create an object of class `Omics` to store these data components. The `ovarian_PNNL_survival.RDS` data set included with the `pathwayPCA` package is a data frame with expression data and response matched by sample IDs. The expression data are recorded protein levels, and the response is overall survival time and event-observation indicator (in that order). This data is from **DATA SOURCE HERE**.
```{r read_assay}
data_path <- system.file("extdata", "ovarian_PNNL_survival.RDS",
                         package = "pathwayPCA", mustWork = TRUE)
ovSurv_df <- readRDS(data_path)
```

The `ovSurv_df` object is a [tibble](https://tibble.tidyverse.org/) data frame (a modern `R` data frame with enhanced properties) with 83 observations (rows) and 5165 features (columns). The first column is the sample ID, the second and third columns are the overall survival information, and the other 5k+ columns are protein expression levels.
```{r view_data}
ovSurv_df
```

Next, we create an `OmicsSurv` data container. This is an S4-class object with slots for the expression data, the response vector, and a `pathwayCollection` object. In our case, the expression data are the 5,162 columns of `ovSurv_df` which measure the protein expression, the response is the second and third column of `ovSurv_df` which contains the survival data, and the pathway collection is the Wikipathways collection we imported above. 
```{r createOvarianPNNLOmics}
ov_OmicsSurv <- CreateOmics(
  assayData_df = ovSurv_df[, -(1:3)],
  pathwayCollection_ls = wikipathways_PC,
  response = ovSurv_df[, 2:3],
  respType = "survival",
  minPathSize = 5,
  centerScale = c(TRUE, TRUE)
)
```

All further data analysis will require this `Omics` object. Because the response is survival data, we see the `CreateOmics()` function print a message that the class of the object is specifically `OmicsSurv`. Based on the chosen `minPathSize` argument, this function additionally prints diagnostic messages about the set intersection between the supplied pathway collection and recorded features in the assay data at object creation. This parameter sets the threshold for pathways used for analysis in the following manner:

1. Biological features in each pathway are checked against the names of the features in the expression data. Any feature that doesn't have a matching data column is removed from the pathway. Therefore, it is imperative that the format of the feature names for the expression data match the format of the names in the pathways set---gene symbols with gene symbols, Entrez IDs with Entrez IDs, etc.
2. After the "trimming" described in Step 1, any pathway with fewer than `minPathSize` features is dropped from the `pathwayCollection` in the supplied `Omics` object. The names of the pathways that were dropped are stored as an internal attribute of the new trimmed `pathwayCollection`.

The last three messages inform you of how strong the overlap is between the genes measured in your data and the genes selected in your pathway collection. These messages tell us that 65% of the 5831 total proteins included in all pathways were not measured in the assay; 133 of the 457 pathways were removed from the pathway collection for having too few genes (less than `minPathSize = 5`) after gene trimming; and the genes in the pathways list call for 40% of the 5162 genes measured in the assay. The last number is the most important: it measures how well your pathway collection overlaps with the genes measured in your assay. This number should be as close to 100% as possible, but this is rarely the case for real data. 

Print the object to confirm that the the `eventTime` and `eventObserved` slots are the survival time and event indicator, respectively.
```{r printOmics}
ov_OmicsSurv
```

</br>

*******************************************************************************

</br>


# Pathway PCA
Now that we have a valid `Omics` object, we can perform pathway-specific PCA. We provide two PCA variants: Adaptive, Elastic-net, Sparse PCA (AES-PCA) and Supervised PCA.


## Method Descriptions
The [AES-PCA method](https://doi.org/10.2202/1544-6115.1697) of Chen (2011) is unsupervised: it does not use the response information to extract pathway PC vectors. Further, this method uses a permutation test within each pathway to estimate the pathway's significance. Therefore, this method's accuracy depends on the number of permutations taken and thus can be computationally expensive.

The [Supervised PCA](https://doi.org/10.1002/gepi.20532) method of Chen et al. (2010) tests each feature within each pathway against the response, then extracts PCs from the most significant subset of features. Further, this method estimates a mixture of Gumbel Extreme Value distributions to calculate the pathway $p$-values.


## AES-PCA Implementation
Because the syntax for performing Supervised PCA is nearly identical to the AES-PCA syntax (other than specifying the number of permutations), we will simply show the AES-PCA workflow. For 1k permutations over 18 computing cores, this takes 1.5 minutes; of this time, approximately 20% will be extracting the PCs, and about 75% will be calculating the permutation $p$-values. For 10k permutations, this takes 10 minutes (not shown). For 100k permutations, this takes roughly an hour and a half (also not shown).
<!-- ```{r seed, echo=FALSE} -->
<!-- set.seed(123456) -->
<!-- ``` -->
<!-- set.seed doesn't work with parallel computing. See Issue #15. -->
```{r AESPCA_run}
# e1 <- Sys.time()
ovarian_aespcOut <- AESPCA_pVals(
  object = ov_OmicsSurv,
  numPCs = 3,
  parallel = TRUE,
  numCores = 2,
  numReps = 10,
  adjustment = "BH"
)
# Sys.time() - e1
```

## Function Details
The above function has the following computational options (unless stated, these options are identical for the `AESPCA_pVals()` and `SuperPCA_pVals()` functions):

- `numPCs` controls the number of principal components extracted from each pathway-specific expression subset.
- `parallel` turns distributed computing on or off. Our parallelization routine works on both Windows and Unix-based systems (Mac and Linux). For AES-PCA, this option is critical (as this method is uses computationally-expensive resampling techniques). The corresponding `numCores` argument dictates the number of processors you will let `R` use for parallel computation.
- `numReps` (AESPCA only) controls the number of permutations used to calculate the $p$-values for each pathway. For low-signal data, you may need to increase this to 10k or even 100k to adequately distinguish between a $p$-value of $0/1000,\ 1/1000$, or $2/1000$. For reproducibility, you can to control the random seed via the `base::` package `set.seed` function when this function is executed serially. Because Supervised PCA is a parametric technique, this argument is unavailable for the `SuperPCA_pVals()` function.
- `adjustment` specifies which family-wise error rate or false discovery rate adjustment to apply to the $p$-values from each pathway. See `?adjustRaw_pVals` for adjustment options and citations.


The top 15 pathways with AES-PCA are (from the results for 100k permutations):
```{r echo=FALSE}
# saveRDS(ovarian_aespcOut, file = "vignettes/aespcaOut_20180928.RDS")
ovarian_aespcOut <- readRDS("aespcaOut_20180928.RDS")
```
```{r ovarian2_AESPCA_pVals_wikiP}
head(ovarian_aespcOut$pVals_df, 15)
```

This data frame has the following columns:

- `pathways`: the `pathwayPCA` internal identification labels for each pathway in the `Omics*` object (stored as the `names` attribute of the `ov_OmicsSurv@@trimPathwayCollection$$pathways` object).
- `setsize`: number of genes in each of the original pathways (as stored in the `ov_OmicsSurv@@trimPathwayCollection$$setsize` object).
- `trim_size`: number of genes in each pathway after being trimmed to match the assay data (as stored in the `ov_OmicsSurv@@trimPathwayCollection$$trim_setsize` object).
- `terms`: title of the pathway (as stored in the `ov_OmicsSurv@@trimPathwayCollection$$TERMS` object).
- `description`: (OPTIONAL) description of the pathway (if stored in the `ov_OmicsSurv@@trimPathwayCollection$$description` object).
- `rawp`: unadjusted $p$-values of each pathway.
- `...` additional columns---one for each method supplied to the `adjustment` argument.


These are the significant pathways at the $\alpha = 0.05$ level (if you are not familiar with the `tidyverse` pipe [` %>% `] operator, see [this article](https://magrittr.tidyverse.org/)):
```{r signif_pathways_0.05}
signifPaths005_df <- 
  # Take the resulting table of pathway p-values,
  ovarian_aespcOut$pVals_df %>% 
  # and filter out any rows with non-significant FDR values.
  filter(FDR_BH < 0.05) %>% 
  # Of the remaining data, select the column of pathway titles and descriptions.
  select(terms, description)

signifPaths005_df
```

We can inspect these pathways by subsetting the trimmed pathway collection by the significant pathway names. For instance, we can inspect the first pathway by
```{r inspect_path1}
trimmed_PC <- getTrimPathwayCollection(ov_OmicsSurv)

trimmed_PC[["WP3858"]]
```
This shows us the pathway name (`Pathway`), gene symbols contained (`IDs`), any relevant details (`Description`), and the number of features currently in this pathway (`Size`). Further, we can view the subset of the assay corresponding to this pathway via
```{r WP75_data}
trimmed_PC[["WP3858"]]$Description
SubsetPathwayData(ov_OmicsSurv, "WP3858")
```

The `SubsetPathwayData()` function returns a data frame of the survival response (first two columns) and the recorded protein expression levels in the assay corresponding to the pathway selected.

These are the additional pathways significant at the $\alpha = 0.2$ FDR-level:
```{r signif_pathways_0.2}
signifPaths020_df <-
  ovarian_aespcOut$pVals_df %>% 
  filter(FDR_BH < 0.2) %>% 
  filter(FDR_BH >= 0.05) %>% 
  select(terms, description)

signifPaths020_df
```

</br>

*******************************************************************************

</br>


# Visualizing the Results
Given the $p$-values from these pathways, we will now graph their score (negative log of the $p$-value) and description for the top 15 most-significant pathways.


## Trim Pathway Names
Because Wikipathways pathway collections are quite long, we truncate them for better graphical display.
```{r match_abbrev_to_name}
# Truncate names longer than 30 characters
pVals_df <- 
  ovarian_aespcOut$pVals_df %>% 
  mutate(
    description = ifelse(
      str_length(description) > 35,
      paste0(str_sub(description, 1, 33), "..."),
      description
    )
  )
```


## Clean and Wrangle the Data Results
Now we subset the top 15 pathways by unadjusted $p$-value. If we had multiple adjustment columns, then we would also have to [gather](https://rpubs.com/mm-c/gather-and-spread) the results of these columns from "wide" to "tidy" form.
```{r join_and_gather}
pValsClean_df <- 
  # Results data frame
  pVals_df %>% 
  # remove the columns we don't need
  select(-pathways, -setsize, -trim_size, -terms, -rawp) %>% 
  # take the first 15 rows
  slice(1:15) %>% 
  # add the score variable
  mutate(score = -log(FDR_BH)) %>%
  rename(
    "Pathway" = description,
    "Benjamini-Hochberg FDR" = FDR_BH
  )

pValsClean_df
```


## Plot the Tidy Results
Using this data, we can plot the top 15 pathway scores for the Benjamini-Hochberg adjustment method.
```{r bar_plots, fig.height = 6, fig.width = 10.7, out.width = "100%", out.height = "60%"}
# The modified data frame from the previous step
ggplot(pValsClean_df) +
  # set overall appearance of the plot
  theme_bw() +
  # Define the dependent and independent variables
  aes(x = reorder(Pathway, score), y = score) +
  # From the defined variables, create a vertical bar chart
  geom_bar(stat = "identity", position = "dodge", fill = "#F47321") +
  # Set main and axis titles
  ggtitle("AES-PCA Significant Wikipathways: Ovarian Cancer") +
  xlab("Pathways") +
  ylab("Negative Log BH-FDR") +
  # Add a line showing the alpha = 0.10 level
  geom_hline(yintercept = -log(0.1), size = 1, color = "#005030") +
  # Flip the x and y axes
  coord_flip()
```

</br>

*******************************************************************************

</br>

# Inspecting the Driving Proteins
Now that we have a few significant pathways, we can look at the loadings of each gene onto the first AES-PC from these pathways. Because the pathway loadings are named by the internal pathway key, we will filter the `pathMatch_df` data frame to the pathways we want.
```{r significant_pathway_keys}
pVals_df %>% 
  filter(FDR_BH < 0.05) %>% 
  select(pathways)
```

Now we can inspect the loadings from each gene onto the first PC in a few of these pathways.
```{r inspect_top_loadings}
# Toll-like Receptor Signaling
proteinLoad253 <- ovarian_aespcOut$loadings_ls[["path253"]][1, ]
sort(proteinLoad253[proteinLoad253 != 0], decreasing = TRUE)

# Toll-like Receptor Signaling Pathway
proteinLoad110 <- ovarian_aespcOut$loadings_ls[["path110"]][1, ]
sort(proteinLoad110[proteinLoad110 != 0], decreasing = TRUE)
```

The intersecting genes shared by these two pathways with non-zero AES-PC loadings are [NFKB1](https://www.ncbi.nlm.nih.gov/pubmed/26345753), [MYD88](https://journals.lww.com/greenjournal/Abstract/2015/05001/Role_of_MyD88_expression_in_Epithelial_Ovarian.259.aspx), [TLR2 & TLR3](https://www.ncbi.nlm.nih.gov/pubmed/19184006), and [IKBKB & TBK1](https://www.ncbi.nlm.nih.gov/pubmed/30084832) (external literature validating the relatioship between these genes and ovarian cancer survival are linked).

</br>

*******************************************************************************

</br>


# Conclusion
We are exploring options to connect this package to other pathway-testing software. We are further considering other summary functions and/or graphics to help display the output of our `*_pVals()` functions. If you have any questions, comments, complaints, or suggestions, please [email us](mailto:gabriel.odom@med.miami.edu) and give us some feedback. Also you can visit our [development page](https://github.com/gabrielodom/pathwayPCA).

Here is the R session information for this vignette:
```{r sessionDetails}
sessionInfo()
```
