<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>AES-PCA Walkthrough with pathwayPCA • pathwayPCA</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">pathwayPCA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AES-PCA_Walkthrough.html">AES-PCA Walkthrough with pathwayPCA</a>
    </li>
    <li>
      <a href="../articles/Importing_Data.html">Import and Tidy Data</a>
    </li>
    <li>
      <a href="../articles/Package_Dev_Notes.html">pathwayPCA Package Development Notes</a>
    </li>
    <li>
      <a href="../articles/Supervised_PCA_Walkthrough.html">Supervised PCA Walkthrough with pathwayPCA</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>AES-PCA Walkthrough with pathwayPCA</h1>
                        <h4 class="author">Gabriel Odom</h4>
            
            <h4 class="date">2018-04-04</h4>
          </div>

    
    
<div class="contents">
<p>This will provide the steps to perform pathway-level adaptive, elastic-net, sparse PCA (AES-PCA) analysis, from data entry to pathway <span class="math inline">\(p\)</span>-value graphics. This vignette will give three examples: survival, regression, and binary responses. The main goal of pathway AES-PCA is to discover potential relationships between pathways (bundles of genes) and a response. This package (in its current form) should not be used for prediction, or regressing on multiple variables, but should simply be used to screen or identify potential pathways for futher inspection.</p>
<div id="workflow-setup-and-data" class="section level2">
<h2 class="hasAnchor">
<a href="#workflow-setup-and-data" class="anchor"></a>Workflow Setup and Data</h2>
<div id="load-package" class="section level3">
<h3 class="hasAnchor">
<a href="#load-package" class="anchor"></a>Load Package</h3>
<p>Before we begin, make sure you have the latest version of the package. In order to install a package from GitHub, you will need the <code>devtools::</code> package (<a href="https://github.com/r-lib/devtools" class="uri">https://github.com/r-lib/devtools</a>) and either <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (for Windows) or <a href="https://developer.apple.com/xcode/">Xcode</a> (for Mac). Then you can install the development version of the <a href="https://github.com/gabrielodom/pathwayPCA"><code>pathwayPCA</code> package</a> from <a href="https://github.com/">GitHub</a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"gabrielodom/pathwayPCA"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(pathwayPCA)
<span class="kw">library</span>(parallel)</code></pre></div>
</div>
<div id="load-data-and-gene-pathway-set" class="section level3">
<h3 class="hasAnchor">
<a href="#load-data-and-gene-pathway-set" class="anchor"></a>Load Data and Gene Pathway Set</h3>
<p>For our pathway analysis to be meaningful, we need gene expression data (from a microarray or something similar), corresponding subject information (such as weight, type of cancer, or survival time and censoring indicator), and a gene set list. Unlike our current implementation of Supervised PCA, our current implementation of AES-PCA can handle some missingness in the data. However, we still recommend that you impute or remove any missingness.</p>
<div id="load-tumour-data" class="section level4">
<h4 class="hasAnchor">
<a href="#load-tumour-data" class="anchor"></a>Load Tumour Data</h4>
<p>For this example we have a colon cancer data set. We will load and inspect this data. We will return to the gene name discussion when we import the gene set list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"colonSurv_df"</span>)
<span class="kw">str</span>(colonSurv_df, <span class="dt">list.len =</span> <span class="dv">10</span>)
<span class="co">#&gt; Classes 'tbl_df', 'tbl' and 'data.frame':    250 obs. of  658 variables:</span>
<span class="co">#&gt;  $ OS_time : num  64.9 59.8 62.4 54.5 46.3 ...</span>
<span class="co">#&gt;  $ OS_event: int  0 0 0 0 1 0 0 0 1 0 ...</span>
<span class="co">#&gt;  $ JUN     : num  9.29 9.13 9.37 10.6 8.7 ...</span>
<span class="co">#&gt;  $ SOS2    : num  5.48 6.35 5.67 4.94 5.6 ...</span>
<span class="co">#&gt;  $ PAK3    : num  8.21 8.33 7.82 8.79 8.75 ...</span>
<span class="co">#&gt;  $ RAF1    : num  8.03 7.94 7.74 7.64 8.05 ...</span>
<span class="co">#&gt;  $ PRKCB   : num  5.49 6.26 6.05 5.37 6.07 ...</span>
<span class="co">#&gt;  $ BTC     : num  6.65 7.02 7.52 6.87 6.49 ...</span>
<span class="co">#&gt;  $ SHC1    : num  8.26 8.39 8.69 7.81 8.45 ...</span>
<span class="co">#&gt;  $ PRKCA   : num  8.94 9.61 8.4 9.8 8.21 ...</span>
<span class="co">#&gt;   [list output truncated]</span></code></pre></div>
<p>This data set has 656 gene expression measurements on 250 colon cancer patients. Notice that this data frame is in <a href="https://www.jstatsoft.org/article/view/v059i10">tidy form</a> (a tidy data matrix has <span class="math inline">\(n\)</span> rows, one for each observation, and <span class="math inline">\(p\)</span> columns, one for each gene or protein measured). If your data frame is not tidy, you should transpose it. (Note that transposition in <code>R</code> will yield an object of class <code>matrix</code>, so use <code>as.data.frame()</code> to turn the transposed matrix back into a data frame.)</p>
</div>
<div id="load-sample-information" class="section level4">
<h4 class="hasAnchor">
<a href="#load-sample-information" class="anchor"></a>Load Sample Information</h4>
<p>We have an expression design matrix with the first two columns as subject response information. We will consider the overall patient survival time (and corresponding censoring indicator) as our survival response response, the event time as our regression response, and event indicator as our binary classification response (these are just for example).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">survTime_num &lt;-<span class="st"> </span>colonSurv_df<span class="op">$</span>OS_time
dead_logi &lt;-<span class="st"> </span><span class="kw">as.logical</span>(colonSurv_df<span class="op">$</span>OS_event)
dead_fact &lt;-<span class="st"> </span><span class="kw">as.factor</span>(colonSurv_df<span class="op">$</span>OS_event)
cancers_df &lt;-<span class="st"> </span>colonSurv_df[, <span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>)]
<span class="kw">rm</span>(colonSurv_df)</code></pre></div>
<p>Notice that I put the responses in specific classes: survival data is stored in numeric and logical vectors, regression data is stored in a numeric or integer vector, and binary classification data is stored in a factor. This is on purpose: the data creation functions in the <code>pathwayPCA::</code> package have very specific requirements about the types of data they take as inputs. We’ll come back to this topic when we create our first <code>Omics*</code>-class objects.</p>
</div>
<div id="load-gene-set-list" class="section level4">
<h4 class="hasAnchor">
<a href="#load-gene-set-list" class="anchor"></a>Load Gene Set List</h4>
<p>For pathway-based analyses to work, we need a well-defined list of pathways.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"colonGenesets_ls"</span>)
geneset_ls &lt;-<span class="st"> </span>colonGenesets_ls
<span class="kw">names</span>(geneset_ls)
<span class="co">#&gt; [1] "pathways" "TERMS"</span>
<span class="kw">str</span>(geneset_ls<span class="op">$</span>TERMS)
<span class="co">#&gt;  chr [1:15] "KEGG_PENTOSE_PHOSPHATE_PATHWAY" "KEGG_RETINOL_METABOLISM" ...</span>
<span class="kw">str</span>(geneset_ls<span class="op">$</span>pathways, <span class="dt">list.len =</span> <span class="dv">10</span>)
<span class="co">#&gt; List of 15</span>
<span class="co">#&gt;  $ pathway3   : chr [1:27] "RPE" "RPIA" "PGM2" "PGLS" ...</span>
<span class="co">#&gt;  $ pathway60  : chr [1:64] "RPE65" "CYP3A5" "UGT2B28" "CYP4A11" ...</span>
<span class="co">#&gt;  $ pathway87  : chr [1:87] "JUN" "SOS2" "PAK3" "RAF1" ...</span>
<span class="co">#&gt;  $ pathway120 : chr [1:89] "HLA-DOA" "HLA-DOB" "KLRC3" "KLRD1" ...</span>
<span class="co">#&gt;  $ pathway176 : chr [1:54] "CASP9" "SOS2" "E2F1" "PRKCB" ...</span>
<span class="co">#&gt;  $ pathway177 : chr [1:30] "HLA-DRB4" "HLA-DRB5" "HLA-DOA" "HLA-DOB" ...</span>
<span class="co">#&gt;  $ pathway187 : chr [1:16] "IKBKG" "CHUK" "EP300" "RELA" ...</span>
<span class="co">#&gt;  $ pathway266 : chr [1:11] "PRF1" "DFFA" "DFFB" "HMGB2" ...</span>
<span class="co">#&gt;  $ pathway390 : chr [1:29] "JUN" "BAG4" "CASP8" "MAPK8" ...</span>
<span class="co">#&gt;  $ pathway413 : chr [1:23] "PLD1" "RAF1" "EPHB2" "VAV1" ...</span>
<span class="co">#&gt;   [list output truncated]</span>
<span class="kw">rm</span>(colonGenesets_ls)</code></pre></div>
<p>The information stored in the gene set list is very important to our analysis, and the form of this information must be discussed carefully. This gene set list has the following components</p>
<ul>
<li>
<code>pathways</code>: a list of character vectors. Each character vector contains a subset of column names in the <code>cancers_df</code> data frame. These pathways should not be too short, otherwise we devolve the problem into simply testing individual genes. Therefore, the <code>pathwayPCA::</code> package requires each pathway to have a minimum of three genes with recorded expressions in the predictor data frame. <strong>Important</strong>: some protein set lists have proteins markers recorded as character numerics, so make sure the column names of your data frame have an overlap with the gene or protein names in the set list. Ensure that there is a non-null set overlap between the names in the pathways and the column names of the predictor data frame. Not every column from the <code>cancers_df</code> data frame will be in a pathway gene set, and not every gene set in each pathway will be recorded in the data frame. <em>However, for meaningful results, there should be a significant overlap between the genes measured in the data frame and the gene names sorted into the pathways.</em> If your gene set list has very few matching genes in your data frame, then your pathway-based analysis results will be significantly degraded. <strong>Make sure your pathway set and data frame are compatible.</strong>
</li>
<li>
<code>TERMS</code>: a brief description of the pathways in the <code>pathways</code> list, stored as a character vector.</li>
<li>
<code>setsize</code>: the number of genes originally recorded in each pathway, stored as an integer vector. NOTE: <em>this information is calculated and added to the pathway set list at <code>Omics*</code>-class object creation.</em> This information is useful to measure the ratio of the number of genes from each pathway expressed in your data to the number of genes defined to be in that pathway. This ratio should be very close to 1 for best pathway analysis results results. (The <code><a href="../reference/expressedOmes.html">expressedOmes()</a></code> function is useful for extracting the expressed genes from each pathway set, and it is called automatically in the Supervised PCA workflow.)</li>
</ul>
</div>
</div>
</div>
<div id="create-an-omics-class-object-for-pathway-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#create-an-omics-class-object-for-pathway-analysis" class="anchor"></a>Create an <code>Omics*</code>-class Object for Pathway Analysis</h2>
<p>Now that we have our data loaded, and we have checked that each piece of data is in the correct form, we can create an analysis object for the <code>pathwayPCA::</code> package. In this package, all primary input data will by an <code>Omics*</code> object. Each response type has a specific creation function:</p>
<ul>
<li>Survival response: Use the <code><a href="../reference/create_OmicsPathway.html">create_OmicsSurv()</a></code> function to create an object with class <code>OmicsSurv</code>. This object will contain:
<ul>
<li>
<code>eventTime</code>: a numeric vector of event times</li>
<li>
<code>eventObserved</code>: a logical vector of death (or other event) indicators. This format precludes the option of recurrent-event survival analysis.</li>
<li>
<code>assayData_df</code>: a tidy data frame of prediction information. Rows are observations or subjects; the columns are gene, protein, transcriptome, proteome, or metabolome measures. The column names <em>must</em> match a subset of the genes provided in the gene set list (in the <code>pathwaySet</code> object).</li>
<li>
<code>pathwaySet</code>: a list of pathway information, as described in the “Load Gene Set List” section. The names of the genes in these pathways <em>must</em> match a subset of the genes recorded in the prediction data frame (in the <code>assayData_df</code> object).</li>
</ul>
</li>
<li>Regression response: Use the <code><a href="../reference/create_OmicsPathway.html">create_OmicsReg()</a></code> function to create an object with class <code>OmicsReg</code>. This object will contain:
<ul>
<li>
<code>response</code>: a numeric vector of the response</li>
<li>
<code>assayData_df</code>: a tidy data frame of prediction information, as described above.</li>
<li>
<code>pathwaySet</code>: a list of pathway information, as described above.</li>
</ul>
</li>
<li>Binary Classification response: Use the <code><a href="../reference/create_OmicsPathway.html">create_OmicsCateg()</a></code> function to create an object with class <code>OmicsCateg</code>. In future versions, this function will be able to take in n-ary responses and ordered categorical responses, but we only support binary responses for now. This object will contain:
<ul>
<li>
<code>response</code>: a factor vector of the response</li>
<li>
<code>assayData_df</code>: a tidy data frame of prediction information, as described above.</li>
<li>
<code>pathwaySet</code>: a list of pathway information, as described above.</li>
</ul>
</li>
</ul>
<div id="create-an-omicssurv-object" class="section level3">
<h3 class="hasAnchor">
<a href="#create-an-omicssurv-object" class="anchor"></a>Create an <code>OmicsSurv</code> Object</h3>
<p>Using the survival response information, we will create a valid <code>OmicsSurv</code> object for further analysis with the <code>pathwayPCA::</code> functions. Remember, it is vitally important to input your data in the forms described above. Because we are creating an <code>OmicsSurv</code> object, we will need event times as a numeric vector and death indicator as a logical vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_OmicsSurv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_OmicsPathway.html">create_OmicsSurv</a></span>(<span class="dt">assayData_df =</span> cancers_df,
                                    <span class="dt">pathwaySet_ls =</span> geneset_ls,
                                    <span class="dt">eventTime_vec =</span> survTime_num,
                                    <span class="dt">eventObserved_vec =</span> dead_logi)
<span class="kw">str</span>(colon_OmicsSurv, <span class="dt">max.level =</span> <span class="dv">2</span>)
<span class="co">#&gt; Formal class 'OmicsSurv' [package "pathwayPCA"] with 4 slots</span>
<span class="co">#&gt;   ..@ eventTime    : num [1:250] 64.9 59.8 62.4 54.5 46.3 ...</span>
<span class="co">#&gt;   ..@ eventObserved: logi [1:250] FALSE FALSE FALSE FALSE TRUE FALSE ...</span>
<span class="co">#&gt;   ..@ assayData_df :Classes 'tbl_df', 'tbl' and 'data.frame':    250 obs. of  656 variables:</span>
<span class="co">#&gt;   ..@ pathwaySet   :List of 3</span></code></pre></div>
<p>Because the <code>assayData_df</code> and <code>pathwaySet</code> objects contain <em>a lot</em> of information, we use the <code>max.level</code> option within the <code>str()</code> function to prevent recursive structure printing. Notice that the <code>pathwaySet</code> object is a list of 3 now: upon creation, the <code>setsize</code> vector is added to the <code>pathwaySet</code> object. We can also repeat this to create regression and classification objects for further analysis. However, I will mention this again: <strong>it is vitally important to create your data objects correctly.</strong> As the old computer science saying goes—garbage in, garbage out.</p>
</div>
<div id="create-an-omicsreg-object" class="section level3">
<h3 class="hasAnchor">
<a href="#create-an-omicsreg-object" class="anchor"></a>Create an <code>OmicsReg</code> Object</h3>
<p>Following our previous step in creation of a survival object, we can create a data object for regression response.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_OmicsReg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_OmicsPathway.html">create_OmicsReg</a></span>(<span class="dt">assayData_df =</span> cancers_df,
                                  <span class="dt">pathwaySet_ls =</span> geneset_ls,
                                  <span class="dt">response_num =</span> survTime_num)
<span class="kw">str</span>(colon_OmicsReg, <span class="dt">max.level =</span> <span class="dv">2</span>)
<span class="co">#&gt; Formal class 'OmicsReg' [package "pathwayPCA"] with 3 slots</span>
<span class="co">#&gt;   ..@ response    : num [1:250] 64.9 59.8 62.4 54.5 46.3 ...</span>
<span class="co">#&gt;   ..@ assayData_df:Classes 'tbl_df', 'tbl' and 'data.frame': 250 obs. of  656 variables:</span>
<span class="co">#&gt;   ..@ pathwaySet  :List of 3</span></code></pre></div>
</div>
<div id="create-an-omicscateg-object" class="section level3">
<h3 class="hasAnchor">
<a href="#create-an-omicscateg-object" class="anchor"></a>Create an <code>OmicsCateg</code> Object</h3>
<p>And we can also create our data object for binary classification response. As we mentioned previously, we do not yet support other GLM models, but we are working on it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_OmicsCateg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_OmicsPathway.html">create_OmicsCateg</a></span>(<span class="dt">assayData_df =</span> cancers_df,
                                      <span class="dt">pathwaySet_ls =</span> geneset_ls,
                                      <span class="dt">response_fact =</span> dead_fact)
<span class="kw">str</span>(colon_OmicsCateg, <span class="dt">max.level =</span> <span class="dv">2</span>)
<span class="co">#&gt; Formal class 'OmicsCateg' [package "pathwayPCA"] with 3 slots</span>
<span class="co">#&gt;   ..@ response    : Factor w/ 2 levels "0","1": 1 1 1 1 2 1 1 1 2 1 ...</span>
<span class="co">#&gt;   ..@ assayData_df:Classes 'tbl_df', 'tbl' and 'data.frame': 250 obs. of  656 variables:</span>
<span class="co">#&gt;   ..@ pathwaySet  :List of 3</span></code></pre></div>
</div>
</div>
<div id="clean-up-the-pathway-list" class="section level2">
<h2 class="hasAnchor">
<a href="#clean-up-the-pathway-list" class="anchor"></a>Clean Up the Pathway List</h2>
<p>The analysis functions within the <code>pathwayPCA::</code> package subset the feature data frame by the genes in each pathway. Therefore, if we have genes in the pathways that aren’t recorded in the data frame, then we will necessarily create missing (<code>NA</code>) predictors. To circumvent this issue, we should check if each gene in each pathway is recorded in the data frame, and remove the genes for which we have not recorded expression levels.</p>
<p>However, if we remove genes from the pathways which do not have recorded levels in the predictor data frame, we could theoretically remove all the genes from a pathway. Thus, we will also check to make sure that each pathway still has three or more genes present after we have removed the genes without corresponding expression levels. The <code><a href="../reference/expressedOmes.html">expressedOmes()</a></code> function performs these two actions simultaneously. This function takes in a valid <code>Omics*</code>-class object, removed the unexpressed genes from each pathway, subsequently trims the pathways that have fewer than three genes, and finally returns an object of the same <code>Omics*</code> class as the input. If there are any pathways that are removed due to this execution, the <code>pathways</code> list within the <code>pathwaySet</code> object within each <code>Omics*</code> object will have a character vector of the pathways removed stored as the <code>"missingPaths"</code> attribute. Access this attribute with the <code>attr()</code> function.</p>
<p>This function is called and executed within the <code><a href="../reference/superPCA_pVals.html">superPCA_pVals()</a></code> function.</p>
</div>
<div id="calculate-pathway-specific-model-p-values" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-pathway-specific-model-p-values" class="anchor"></a>Calculate Pathway-Specific Model <span class="math inline">\(p\)</span>-Values</h2>
<p>We now have a clean set of <code>Omics*</code> objects. We will</p>
<ol style="list-style-type: decimal">
<li>Extract the first principal components from each expressed pathway in the MS design matrix</li>
<li>Test their association with the response matrix or vector. This step uses a permutation test to estimate the pathway <span class="math inline">\(p\)</span>-values.</li>
<li>Adjust the pathway <span class="math inline">\(p\)</span>-values for False Discovery Rate (FDR)</li>
<li>Return a sorted data frame of the adjusted <span class="math inline">\(p\)</span>-values for each pathway</li>
</ol>
<p>The accuracy of the permuted <span class="math inline">\(p\)</span>-values will depend on how many permutations you call for. The default number in the <code><a href="../reference/AESPCA_pVals.html">AESPCA_pVals()</a></code> function is 1000. Be warned, however, that this may be too few permutations to create accurate seperation in pathway significance. You could increase the permutations to a larger value, should your computing resources allow for that. For even moderately-sized data sets (~2000 features) and 1000 pathways, this could take half an hour or more. About 20-25% of the computing costs will be extracting the AES-PCs from each pathway, and the remaining 75-80% of the cost will be the permutation test (for 1000 reps).</p>
<p>Another note: we use the FDR adjustments executed in the <code>mt.rawp2adjp</code> function from the Bioconductor package <a href="https://www.bioconductor.org/packages/3.7/bioc/manuals/multtest/man/multtest.pdf"><code>multtest</code></a>. We copied the code from this function into our package, and made some slight modifications. However, we do not depend on this package directly. We acknowledge their work in this area and express our gratitude.</p>
<div id="survival-response" class="section level3">
<h3 class="hasAnchor">
<a href="#survival-response" class="anchor"></a>Survival Response</h3>
<p>We will use all but two of our available cores, and use the Hochberg (1988) and Sidak Step-Down FDR-adjustment procedures.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">a &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()
surv_pVals_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(<span class="dt">object =</span> colon_OmicsSurv,
                              <span class="dt">numReps =</span> <span class="dv">1000</span>,
                              <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
                              <span class="dt">numCores =</span> <span class="dv">2</span>,
                              <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
                              <span class="dt">adjustment =</span> <span class="kw">c</span>(<span class="st">"Hoch"</span>, <span class="st">"SidakSD"</span>))
<span class="co">#&gt; Of the 676 unique genes in the input pathway set, 9.0% were not expressed in</span>
<span class="co">#&gt;   the input data and were therefore removed.</span>
<span class="co">#&gt; After trimming unexpressed genes from the 15 supplied pathways, we removed 0</span>
<span class="co">#&gt;   pathway(s) because they contained 3 or fewer genes.</span>
<span class="co">#&gt; Of the 656 measured genes in the input data frame, 93.8% were included in at</span>
<span class="co">#&gt;   least one pathway after trimming.</span>
<span class="co">#&gt;  Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 2: Calculate Permuted Pathway p-Values</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span>
<span class="kw">Sys.time</span>() <span class="op">-</span><span class="st"> </span>a
<span class="co">#&gt; Time difference of 25.72435 secs</span></code></pre></div>
</div>
<div id="regression-response" class="section level3">
<h3 class="hasAnchor">
<a href="#regression-response" class="anchor"></a>Regression Response</h3>
<p>We will use the Holm (1979) and Benjamini and Hochberg (1995) FDR-adjustment procedures.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">b &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()
reg_pVals_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(<span class="dt">object =</span> colon_OmicsReg,
                             <span class="dt">numReps =</span> <span class="dv">1000</span>,
                             <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
                             <span class="dt">numCores =</span> <span class="dv">2</span>,
                             <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
                             <span class="dt">adjustment =</span> <span class="kw">c</span>(<span class="st">"Holm"</span>, <span class="st">"BH"</span>))
<span class="co">#&gt; Of the 676 unique genes in the input pathway set, 9.0% were not expressed in</span>
<span class="co">#&gt;   the input data and were therefore removed.</span>
<span class="co">#&gt; After trimming unexpressed genes from the 15 supplied pathways, we removed 0</span>
<span class="co">#&gt;   pathway(s) because they contained 3 or fewer genes.</span>
<span class="co">#&gt; Of the 656 measured genes in the input data frame, 93.8% were included in at</span>
<span class="co">#&gt;   least one pathway after trimming.</span>
<span class="co">#&gt;  Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 2: Calculate Permuted Pathway p-Values</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span>
<span class="kw">Sys.time</span>() <span class="op">-</span><span class="st"> </span>b
<span class="co">#&gt; Time difference of 14.09975 secs</span></code></pre></div>
</div>
<div id="binary-classification-response" class="section level3">
<h3 class="hasAnchor">
<a href="#binary-classification-response" class="anchor"></a>Binary Classification Response</h3>
<p>We will use the Sidak Single-Step and Benjamini and Yekutieli (2001) FDR-adjustment procedures.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">d &lt;-<span class="st"> </span><span class="kw">Sys.time</span>()
categ_pVals_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(<span class="dt">object =</span> colon_OmicsCateg,
                               <span class="dt">numReps =</span> <span class="dv">1000</span>,
                               <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
                               <span class="dt">numCores =</span> <span class="dv">2</span>,
                               <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
                               <span class="dt">adjustment =</span> <span class="kw">c</span>(<span class="st">"SidakSS"</span>, <span class="st">"BY"</span>))
<span class="co">#&gt; Of the 676 unique genes in the input pathway set, 9.0% were not expressed in</span>
<span class="co">#&gt;   the input data and were therefore removed.</span>
<span class="co">#&gt; After trimming unexpressed genes from the 15 supplied pathways, we removed 0</span>
<span class="co">#&gt;   pathway(s) because they contained 3 or fewer genes.</span>
<span class="co">#&gt; Of the 656 measured genes in the input data frame, 93.8% were included in at</span>
<span class="co">#&gt;   least one pathway after trimming.</span>
<span class="co">#&gt;  Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 2: Calculate Permuted Pathway p-Values</span>
<span class="co">#&gt; Initializing Cluster</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel</span>
<span class="co">#&gt; DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span>
<span class="kw">Sys.time</span>() <span class="op">-</span><span class="st"> </span>d
<span class="co">#&gt; Time difference of 23.40692 secs</span></code></pre></div>
</div>
</div>
<div id="subset-the-top-pathways" class="section level2">
<h2 class="hasAnchor">
<a href="#subset-the-top-pathways" class="anchor"></a>Subset the Top Pathways</h2>
<p>Now that we have the pathway-specific <span class="math inline">\(p\)</span>-values, we can inspect the top pathways by significance. We will first subset the data frame returned by the <code><a href="../reference/superPCA_pVals.html">superPCA_pVals()</a></code> function, then plot the most significant <span class="math inline">\(p\)</span>-values using the <code>ggplot2::</code> package. For graphics, we will need two new packages:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="co">#&gt; -- Attaching packages ------------------------------------------------------------------------------------------------------------------- tidyverse 1.2.1 --</span>
<span class="co">#&gt; v ggplot2 2.2.1     v purrr   0.2.4</span>
<span class="co">#&gt; v tibble  1.4.2     v dplyr   0.7.4</span>
<span class="co">#&gt; v tidyr   0.8.0     v stringr 1.3.0</span>
<span class="co">#&gt; v readr   1.1.1     v forcats 0.3.0</span>
<span class="co">#&gt; -- Conflicts ---------------------------------------------------------------------------------------------------------------------- tidyverse_conflicts() --</span>
<span class="co">#&gt; x dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; x dplyr::lag()    masks stats::lag()</span>
<span class="kw">library</span>(reshape2)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'reshape2'</span>
<span class="co">#&gt; The following object is masked from 'package:tidyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     smiths</span></code></pre></div>
<p>For each pathway <span class="math inline">\(p\)</span>-value data frame, we will select the top pathways, and transform the data for better graphics. Then, we will plot the pathway <span class="math inline">\(p\)</span>-values for the most significant pathways as a horizontal bar chart. The thick vertical black line is at <span class="math inline">\(p = 0.05\)</span>, and the thin vertical black line is at <span class="math inline">\(p = 0.10\)</span>.</p>
<div id="significant-survival-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#significant-survival-pathways" class="anchor"></a>Significant Survival Pathways</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Melt the data</span>
surv_melt_df &lt;-<span class="st"> </span>surv_pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>terms, <span class="op">-</span><span class="st"> </span>setsize, <span class="op">-</span>trim_size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">melt</span>(<span class="dt">id.vars =</span> <span class="st">"pathways"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw">log</span>(value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">pathways =</span> <span class="kw">factor</span>(pathways,
                           <span class="dt">levels =</span> <span class="kw">rev</span>(<span class="kw">unique</span>(pathways)),
                           <span class="dt">ordered =</span> <span class="ot">TRUE</span>))

<span class="co"># Plot Melted Data</span>
<span class="kw">ggplot</span>(surv_melt_df) <span class="op">+</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> pathways, <span class="dt">y =</span> score, <span class="dt">fill =</span> variable) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>, <span class="dt">position =</span> <span class="st">"dodge"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name =</span> <span class="st">"FDR Adjustment"</span>,
                      <span class="dt">breaks =</span> <span class="kw">c</span>(<span class="st">"rawp"</span>, <span class="st">"Hochberg"</span>, <span class="st">"SidakSD"</span>),
                      <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">"None"</span>, <span class="st">"Hochberg"</span>, <span class="st">"Sidak SD"</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Colon Pathways by Survival Response"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Log p-Value"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.05</span>), <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.1</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="AES-PCA_Walkthrough_files/figure-html/surv_pval_plot-1.png" width="100%" height="60%"></p>
</div>
<div id="significant-regression-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#significant-regression-pathways" class="anchor"></a>Significant Regression Pathways</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Melt the data</span>
reg_melt_df &lt;-<span class="st"> </span>reg_pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>terms, <span class="op">-</span><span class="st"> </span>setsize, <span class="op">-</span>trim_size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">melt</span>(<span class="dt">id.vars =</span> <span class="st">"pathways"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw">log</span>(value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">pathways =</span> <span class="kw">factor</span>(pathways,
                           <span class="dt">levels =</span> <span class="kw">rev</span>(<span class="kw">unique</span>(pathways)),
                           <span class="dt">ordered =</span> <span class="ot">TRUE</span>))

<span class="co"># Plot Melted Data</span>
<span class="kw">ggplot</span>(reg_melt_df) <span class="op">+</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> pathways, <span class="dt">y =</span> score, <span class="dt">fill =</span> variable) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>, <span class="dt">position =</span> <span class="st">"dodge"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name =</span> <span class="st">"FDR Adjustment"</span>,
                      <span class="dt">breaks =</span> <span class="kw">c</span>(<span class="st">"rawp"</span>, <span class="st">"Holm"</span>, <span class="st">"BH"</span>),
                      <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">"None"</span>, <span class="st">"Holm"</span>, <span class="st">"BH"</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Colon Pathways by Regression Response"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Log p-Value"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.05</span>), <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.1</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="AES-PCA_Walkthrough_files/figure-html/reg_pval_plot-1.png" width="100%" height="60%"></p>
</div>
<div id="significant-binary-classification-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#significant-binary-classification-pathways" class="anchor"></a>Significant Binary Classification Pathways</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Melt the data</span>
categ_melt_df &lt;-<span class="st"> </span>categ_pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>terms, <span class="op">-</span><span class="st"> </span>setsize, <span class="op">-</span>trim_size) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">melt</span>(<span class="dt">id.vars =</span> <span class="st">"pathways"</span>) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw">log</span>(value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">pathways =</span> <span class="kw">factor</span>(pathways,
                           <span class="dt">levels =</span> <span class="kw">rev</span>(<span class="kw">unique</span>(pathways)),
                           <span class="dt">ordered =</span> <span class="ot">TRUE</span>))

<span class="co"># Plot Melted Data</span>
<span class="kw">ggplot</span>(categ_melt_df) <span class="op">+</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> pathways, <span class="dt">y =</span> score, <span class="dt">fill =</span> variable) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>, <span class="dt">position =</span> <span class="st">"dodge"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">name =</span> <span class="st">"FDR Adjustment"</span>,
                      <span class="dt">breaks =</span> <span class="kw">c</span>(<span class="st">"rawp"</span>, <span class="st">"SidakSS"</span>, <span class="st">"BY"</span>),
                      <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">"None"</span>, <span class="st">"Sidak SS"</span>, <span class="st">"BY"</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Colon Pathways by Binary Classification Response"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Log p-Value"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.05</span>), <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw">log</span>(<span class="fl">0.1</span>)) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="AES-PCA_Walkthrough_files/figure-html/categ_pval_plot-1.png" width="100%" height="60%"></p>
</div>
</div>
<div id="rank-significant-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#rank-significant-genes" class="anchor"></a>Rank Significant Genes</h2>
<p>Given that we have so far only considered collections of genes, rather than the genes themselves, we can inspect which genes show up the most often in the top-ranked pathways. Our method is to build a matrix of pathways and their genes: each column is a pathway, and each row is a gene. The <span class="math inline">\(i, j\)</span> entry of this matrix is a 1 if gene <span class="math inline">\(i\)</span> is contained in pathway <span class="math inline">\(j\)</span> (after trimming to the genes measured in the supplied assay). Now consider the adjusted <span class="math inline">\(p\)</span>-values of pathway <span class="math inline">\(j\)</span>, <span class="math inline">\(p_{1j}, p_{2j}, \ldots, p_{Aj}\)</span>, where <span class="math inline">\(a \in \{1, 2, \ldots A\}\)</span> are the FDR adjustment methods; we define the pathway score for this pathway as <span class="math display">\[
S_j := - \frac{1}{A}\sum\limits_{a = 1}^A \log(p_{aj}).
\]</span> We then mulitply each of the columns in the matrix by its respective score.</p>
<p>Now consider the rows of this matrix. Each row is a gene, and the <span class="math inline">\(j^{th}\)</span> entry in the row is equal to <span class="math inline">\(S_j\)</span>, if the gene is included in pathway <span class="math inline">\(j\)</span>, and 0 otherwise. These genes can be compared by the raw row sum of scores, or by the average of the non-zero scores. The <code><a href="../reference/topGenes.html">topGenes()</a></code> function returns the top 5% of both. As we can see, most of the genes included in the <code>summedRank</code> vector are the in the list of “usual suspects” for colon cancer.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/topGenes.html">topGenes</a></span>(<span class="dt">object =</span> colon_OmicsSurv, <span class="dt">pVals_df =</span> surv_pVals_df, <span class="dt">percentile =</span> <span class="fl">0.02</span>)
<span class="co">#&gt; $summedRank</span>
<span class="co">#&gt;   PIK3CA   PIK3R1    MAPK1   MAP2K1     NRAS     GRB2    MAPK3     HRAS </span>
<span class="co">#&gt; 16.79290 16.79290 13.17714 13.13627 13.13627 13.13627 13.13627 13.13627 </span>
<span class="co">#&gt;     KRAS   PIK3CB   PIK3R2     SHC1      CRK </span>
<span class="co">#&gt; 13.13627 13.13627 13.13627 13.09540 13.09540 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $averagedRank</span>
<span class="co">#&gt;     FRS2    THEM4      KLB     DOK1    EIF4B    EIF4E   EIF4G1     FGF1 </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;     FGF2     FGF3     FGF4     FGF5     FGF6     FGF7     FGF8     FGF9 </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;    FGF10    FGFR1    FGFR3    FGFR2    FGFR4    FGF20    FGF22    GRB10 </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;      INS     INSR     IRS1    PDE3B   PRKAG2    CAB39   PRKAG3     TLR9 </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;    PPM1A   STRADB   PRKAA1   PRKAA2   PRKAB1   PRKAB2   PRKAG1    RPTOR </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;    TRIB3     RHEB     RPS6    MLST8    STK11     TSC1     TSC2    YWHAB </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;    FGF23   CAB39L     IRS2    FGF18    FGF17   STRADA       KL     CDK1 </span>
<span class="co">#&gt; 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 3.929912 </span>
<span class="co">#&gt;    FGF19 </span>
<span class="co">#&gt; 3.929912</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#workflow-setup-and-data">Workflow Setup and Data</a></li>
      <li><a href="#create-an-omics-class-object-for-pathway-analysis">Create an <code>Omics*</code>-class Object for Pathway Analysis</a></li>
      <li><a href="#clean-up-the-pathway-list">Clean Up the Pathway List</a></li>
      <li><a href="#calculate-pathway-specific-model-p-values">Calculate Pathway-Specific Model <span class="math inline">\(p\)</span>-Values</a></li>
      <li><a href="#subset-the-top-pathways">Subset the Top Pathways</a></li>
      <li><a href="#rank-significant-genes">Rank Significant Genes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
