<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Chapter 2 - Import and Tidy Data • pathwayPCA</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="Chapter 2 - Import and Tidy Data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathwayPCA</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    User Guides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/C1-Quickstart_Guide.html">Chapter 1 - Quickstart Guide</a>
    </li>
    <li>
      <a href="../articles/C2-Importing_Data.html">Chapter 2 - Import and Tidy Data</a>
    </li>
    <li>
      <a href="../articles/C3-Create_Omics_Objects.html">Chapter 3 - Creating Data Objects</a>
    </li>
    <li>
      <a href="../articles/C4-Methods_Walkthrough.html">Chapter 4 - Test Pathway Significance</a>
    </li>
    <li>
      <a href="../articles/Package_Dev_Notes.html">Package Development Notes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Version Logs</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Chapter 2 - Import and Tidy Data</h1>
                        <h4 class="author">Gabriel Odom</h4>
            
            <h4 class="date">2018-04-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gabrielodom/pathwayPCA/blob/master/vignettes/C2-Importing_Data.Rmd"><code>vignettes/C2-Importing_Data.Rmd</code></a></small>
      <div class="hidden name"><code>C2-Importing_Data.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>1. Overview</h1>
<p>This vignette is the second chapter in the “Pathway Significance Testing with <code>pathwayPCA</code>” workflow, providing a detailed perspective to the <a href="https://gabrielodom.github.io/pathwayPCA/articles/C1-Quickstart_Guide.html#import-data">Import Data</a> section of the Quickstart Guide. This vignette will discuss using the the <code>read_gmt</code> function to import Gene Matrix Transposed (<code>.gmt</code>) <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">gene set files</a> as a list object with class <code>pathwaySet</code>. Also, we will discuss importing assay and response data, and how to make your assay data <a href="https://www.jstatsoft.org/article/view/v059i10">tidy</a>. For our pathway analysis to be meaningful, we need gene expression data (from a microarray or something similar), corresponding phenotype information (such as weight, type of cancer, or survival time and censoring indicator), and a gene set list.</p>
<p>Before we move on, we will outline our steps. After reading this vignette, you should be able to</p>
<ol style="list-style-type: decimal">
<li>Import a <code>.gmt</code> file and save the gene sets therein as a <code>pathwaySet</code> object using the <code>read_gmt</code> function.</li>
<li>Import an assay <code>.csv</code> file with the <code>read_csv</code> function from the <code>readr</code> package, and transpose this data frame into “tidy” form with the <code>transpose_assay</code> function.</li>
<li>Import phenotype information stored in a <code>.csv</code> file, and join (merge) it to the assay data frame with the <code>inner_join</code> function from the <code>dplyr</code> package.</li>
</ol>
<p>First, load the <code>pathwayPCA</code> package and the <a href="https://www.tidyverse.org/"><code>tidyverse</code> package suite</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(tidyverse)
<span class="co">#&gt; ── Attaching packages ───────────────────────────────────────────── tidyverse 1.2.1 ──</span>
<span class="co">#&gt; ✔ ggplot2 2.2.1     ✔ purrr   0.2.4</span>
<span class="co">#&gt; ✔ tibble  1.4.2     ✔ dplyr   0.7.4</span>
<span class="co">#&gt; ✔ tidyr   0.8.0     ✔ stringr 1.3.0</span>
<span class="co">#&gt; ✔ readr   1.1.1     ✔ forcats 0.3.0</span>
<span class="co">#&gt; ── Conflicts ──────────────────────────────────────────────── tidyverse_conflicts() ──</span>
<span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span>
<span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span>
<span class="kw">library</span>(pathwayPCA)</code></pre></div>
<hr>
<p></p>
</div>
<div id="gmt-files" class="section level1">
<h1 class="hasAnchor">
<a href="#gmt-files" class="anchor"></a>2. GMT Files</h1>
<p>The <code>.gmt</code> format is a commonly used file format for storing <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">pathway set information</a>. Gene sets in the Molecular Signatures Database (MSigDB) can be downloaded from the <a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp">MSigDB Collections page</a>.</p>
<div id="gmt-format-description" class="section level2">
<h2 class="hasAnchor">
<a href="#gmt-format-description" class="anchor"></a>2.1 GMT Format Description</h2>
<p>GMT-formatted files follow a very specific set of rules:</p>
<ol style="list-style-type: decimal">
<li>Each row of the file represents a gene set, and only one gene set is allowed per line.</li>
<li>The first entry in each row is the gene set name. Eg. <code>"KEGG_STEROID_BIOSYNTHESIS"</code>
</li>
<li>The second entry in each row is an optional brief description of the gene set. Eg. <code>"http://www.broadinstitute.org/gsea/msigdb/cards/KEGG_STEROID_BIOSYNTHESIS"</code>
</li>
<li>The third to the last entry on each row are the genes in the gene set. Eg. <code>"SOAT1"  "LSS" "SQLE"  "EBP" "CYP51A1" "DHCR7" "CYP27B1" "DHCR24"  "HSD17B7" "MSMO1" "FDFT1" "SC5DL" "LIPA"  "CEL" "TM7SF2"  "NSDHL" "SOAT2"</code>
</li>
<li>Each entry in each line is seperated by a tab.</li>
</ol>
</div>
<div id="import-gmt-files-with-read_gmt" class="section level2">
<h2 class="hasAnchor">
<a href="#import-gmt-files-with-read_gmt" class="anchor"></a>2.2 Import GMT files with <code>read_gmt</code>
</h2>
<p>Based on the clearly-organized <code>.gmt</code> file format, we are able to write a very fast function to read <code>.gmt</code> files into R. The <code>read_gmt</code> function takes in a path specifying where your <code>.gmt</code> file is stored, and outputs a list of gene set information.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gmt_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"c2.cp.v6.0.symbols.gmt"</span>,
                         <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
gene_set_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_gmt.html">read_gmt</a></span>(gmt_path, <span class="dt">description =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>We now carefully discuss the form of this information. This <code>gene_set_ls</code> object has class <code>pathwaySet</code> and contains the following components:</p>
<ol style="list-style-type: decimal">
<li>
<code>pathways</code>: A list of character vectors. Each character vector should contain a subset of the names of the -Omes measured in your assay data frame. These pathways should not be too short, otherwise we devolve the problem into simply testing individual genes. Therefore, the <code>pathwayPCA</code> package requires each pathway to have a minimum of three genes with recorded expressions in the predictor data frame.</li>
</ol>
<p><strong>Important</strong>: some protein set lists have proteins markers recorded as character numerics (e.g. “3”), so make sure the feature names of your assay have an overlap with the gene or protein names in the <code>pathwaySet</code> list. Ensure that there is a non-empty overlap between the names in the pathways and the feature names of the assay. Not every feature in your assay data frame will be in a pathway gene set, and not every gene in each pathway will have a measurement in the assay data frame. <em>However, for meaningful results, there should be a significant overlap between the genes measured in the data frame and the gene names sorted into the pathways.</em> If your gene set list has very few matching genes in your data frame, then your pathway-based analysis results will be significantly degraded. <strong>Make sure your pathway set and data assay are compatible.</strong></p>
<ol start="2" style="list-style-type: decimal">
<li>
<code>TERMS</code>: A character vector the same length as the pathways list with the proper names of the pathways.</li>
<li>
<code>description</code>: (OPTIONAL) A character vector the same length as the pathways list with descriptive information. For instance, the <code>.gmt</code> file included with this package has hyperlinks to the MSigDB description card for that pathway in this field. This field will be imported when <code>description = TRUE</code> (it defaults to <code>FALSE</code>).</li>
<li>
<code>setsize</code>: the number of genes originally recorded in each pathway, stored as an integer vector. NOTE: <em>this information is calculated and added to the pathway set list later, at <code>Omics</code>-class object creation.</em> This information is useful to measure the ratio of the number of genes from each pathway expressed in your assay to the number of genes defined to be in that pathway. For each pathway, this ratio should be very close to 1 for best pathway analysis results.</li>
</ol>
<p>The object itself has the following structure:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_set_ls
<span class="co">#&gt; Object with Class(es) 'pathwaySet', 'list' [package 'pathwayPCA'] with 3 elements: </span>
<span class="co">#&gt;  $ pathways   :List of 1329</span>
<span class="co">#&gt;  $ TERMS      : chr [1:1329] "KEGG_GLYCOLYSIS_GLUCONEOGENESIS" ...</span>
<span class="co">#&gt;  $ description: chr [1:1329] "http://www.broadinstitute.org/gsea/msigdb/cards/KEGG_GLYCOLYSIS_GLUCONEOGENESIS" ...</span></code></pre></div>
<p>This object will be the list supplied to the <code>pathwaySet_ls</code> argument in the <code>create_Omics</code> function.</p>
</div>
<div id="creating-your-own-pathwayset-list" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-your-own-pathwayset-list" class="anchor"></a>2.3 Creating Your Own <code>pathwaySet</code> List</h2>
<p>Additionally, you can create a <code>pathwaySet</code> object from scratch with the <code>create_pathwaySet</code> function. This may be useful to users who have their pathway information stored in some form other than a <code>.gmt</code> file. You must supply a list of -Omes to the <code>pathways</code> argument, and the proper names of the pathways to the <code>TERMS</code> argument. You could also store any other pertinant pathway information by passing a <code>&lt;name&gt; = &lt;value&gt;</code> pair to this function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">myPathways_ls &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">pathway1 =</span> <span class="kw">c</span>(<span class="st">"Gene1"</span>, <span class="st">"Gene2"</span>),
                      <span class="dt">pathway2 =</span> <span class="kw">c</span>(<span class="st">"Gene3"</span>, <span class="st">"Gene4"</span>, <span class="st">"Gene5"</span>),
                      <span class="dt">pathway3 =</span> <span class="st">"Gene6"</span>)
myPathway_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"KEGG_IMPORTANT_PATHWAY_1"</span>,
                     <span class="st">"KEGG_IMPORTANT_PATHWAY_2"</span>,
                     <span class="st">"SOME_OTHER_PATHWAY"</span>)
<span class="kw"><a href="../reference/create_pathwaySet.html">create_pathwaySet</a></span>(<span class="dt">pathways =</span> myPathways_ls,
                  <span class="dt">TERMS =</span> myPathway_names,
                  <span class="dt">website =</span> <span class="st">"URL_TO_PATHWAY_CITATION"</span>)
<span class="co">#&gt; Object with Class(es) 'pathwaySet', 'list' [package 'pathwayPCA'] with 3 elements: </span>
<span class="co">#&gt;  $ pathways:List of 3</span>
<span class="co">#&gt;  $ TERMS   : chr [1:3] "KEGG_IMPORTANT_PATHWAY_1" ...</span>
<span class="co">#&gt;  $ website : chr "URL_TO_PATHWAY_CITATION"</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="import-and-tidy-an-assay-matrix" class="section level1">
<h1 class="hasAnchor">
<a href="#import-and-tidy-an-assay-matrix" class="anchor"></a>3. Import and Tidy an Assay Matrix</h1>
<p>We assume that the assay data (e.g. transcriptomic data) is either in an Excel file or flat text file. For example, your data may look like this:</p>
<div class="figure">
<img src="example_assay_data.PNG">
</div>
<p>In this data set, the columns are individual samples. The values in each row are the -Omic expression measurements for the gene in that row.</p>
<div id="import-with-readr" class="section level2">
<h2 class="hasAnchor">
<a href="#import-with-readr" class="anchor"></a>3.1 Import with <code>readr</code>
</h2>
<p>To import these files in <code>.csv</code> (comma-separated), <code>.fwf</code> (fixed-width), or <code>.txt</code> (tab-delimited), we recommend the <a href="https://readr.tidyverse.org/"><code>readr</code> package</a>. Additionally, you can read fixed-width files with the <code>read_fwf</code> function, and general delimited files with <code>read_delim</code>. Both of these functions are from the <code>readr</code> package. For data in <code>.xls</code> or <code>.xlsx</code> format, we recommend the <a href="http://readxl.tidyverse.org/"><code>readxl</code></a> package. We would read a <code>.csv</code> data file via</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assay_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"ex_assay_subset.csv"</span>,
                          <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
assay_df &lt;-<span class="st"> </span><span class="kw">read_csv</span>(assay_path)</code></pre></div>
<pre><code>#&gt; Warning: Missing column names filled in: 'X1' [1]
#&gt; Parsed with column specification:
#&gt; cols(
#&gt;   .default = col_double(),
#&gt;   X1 = col_character()
#&gt; )
#&gt; See spec(...) for full column specifications.</code></pre>
<p>The <code>read_csv</code> function warns us that the name of the first column is missing, but then automatically fills it in as <code>X1</code>. Further, this function prints messages to the screen informing you of the assumptions it makes when importing your data. Specifically, this message tells us that all the imported data is numeric (<code>col_double()</code>) except for the gene name column (<code>X1 = col_character()</code>).</p>
<p>Let’s inspect our assay data frame. Note that the gene names were imported as a character column (see the <code>&lt;chr&gt;</code> tag at the top of the first column). This data import step stored the row names (the gene names) as the first column, and preserved the column names (sample labels) of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assay_df
<span class="co">#&gt; # A tibble: 17 x 37</span>
<span class="co">#&gt;    X1      T21101311 T21101312 T21101313 T21101314 T21101315 T21101316</span>
<span class="co">#&gt;    &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 SOAT1        5.37      5.52      5.89      5.62      5.49      5.58</span>
<span class="co">#&gt;  2 LSS          9.77      9.78      8.11      8.67      9.83      9.85</span>
<span class="co">#&gt;  3 SQLE         7.74      8.06      7.00      8.59      8.13      8.55</span>
<span class="co">#&gt;  4 EBP          4.68      5.12      5.78      5.64      5.73      5.13</span>
<span class="co">#&gt;  5 CYP51A1      8.27      8.21      8.20      8.07      9.38      9.40</span>
<span class="co">#&gt;  6 DHCR7        8.32      8.33      8.39      8.64      8.15      8.71</span>
<span class="co">#&gt;  7 CYP27B1      6.78      6.47      6.57      6.47      6.43      6.56</span>
<span class="co">#&gt;  8 DHCR24       4.70      5.06      4.89      4.98      5.03      4.87</span>
<span class="co">#&gt;  9 HSD17B7      7.63      7.63      8.15      8.13      7.70      7.75</span>
<span class="co">#&gt; 10 MSMO1        7.60      7.33      7.61      5.56      6.30      7.77</span>
<span class="co">#&gt; 11 FDFT1       10.5      10.6      10.2      10.4       9.99      9.78</span>
<span class="co">#&gt; 12 SC5DL        9.53      9.35      9.82      9.70      9.50      9.34</span>
<span class="co">#&gt; 13 LIPA         8.17      6.75      7.36      7.17      6.82      7.82</span>
<span class="co">#&gt; 14 CEL         10.5       9.85     10.0       9.80     10.6       9.66</span>
<span class="co">#&gt; 15 TM7SF2       9.83     10.5       9.83     10.0      10.3       9.46</span>
<span class="co">#&gt; 16 NSDHL        6.88      7.47      7.75      6.16      5.94      6.74</span>
<span class="co">#&gt; 17 SOAT2        7.79      7.68      8.05      7.79      7.74      7.52</span>
<span class="co">#&gt; # ... with 30 more variables: T21101317 &lt;dbl&gt;, T21101318 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101319 &lt;dbl&gt;, T21101320 &lt;dbl&gt;, T21101321 &lt;dbl&gt;, T21101322 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101323 &lt;dbl&gt;, T21101324 &lt;dbl&gt;, T21101325 &lt;dbl&gt;, T21101326 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101327 &lt;dbl&gt;, T21101328 &lt;dbl&gt;, T21101329 &lt;dbl&gt;, T21101330 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101331 &lt;dbl&gt;, T21101332 &lt;dbl&gt;, T21101333 &lt;dbl&gt;, T21101334 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101335 &lt;dbl&gt;, T21101336 &lt;dbl&gt;, T21101337 &lt;dbl&gt;, T21101338 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101339 &lt;dbl&gt;, T21101340 &lt;dbl&gt;, T21101341 &lt;dbl&gt;, T21101342 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   T21101343 &lt;dbl&gt;, T21101344 &lt;dbl&gt;, T21101345 &lt;dbl&gt;, T21101346 &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="tidy-the-assay-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#tidy-the-assay-data-frame" class="anchor"></a>3.2 Tidy the Assay Data Frame</h2>
<p>The assay input to the <code>pathwayPCA</code> package must be in <a href="https://www.jstatsoft.org/article/view/v059i10"><em>tidy data</em></a> format. The “Tidy Data” format requires that each observation be its own row, and each measurement its own column. This means that we must transpose our assay data frame, while preserving the row and column names.</p>
<p>To do this, we can use the <code>transpose_assay</code> function. This function takes in a data frame as imported by the <code>readr</code> function based on data in a format similar to that shown above: genes are the rows, gene names are the first column, samples are stored in the subsequent columns, and all values in the assay (other than the gene names in the first column) are numeric.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(assayT_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/transpose_assay.html">transpose_assay</a></span>(assay_df))
<span class="co">#&gt; # A tibble: 36 x 18</span>
<span class="co">#&gt;    Sample    SOAT1   LSS  SQLE   EBP CYP51A1 DHCR7 CYP27B1 DHCR24 HSD17B7</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 T21101311  5.37  9.77  7.74  4.68    8.27  8.32    6.78   4.70    7.63</span>
<span class="co">#&gt;  2 T21101312  5.52  9.78  8.06  5.12    8.21  8.33    6.47   5.06    7.63</span>
<span class="co">#&gt;  3 T21101313  5.89  8.11  7.00  5.78    8.20  8.39    6.57   4.89    8.15</span>
<span class="co">#&gt;  4 T21101314  5.62  8.67  8.59  5.64    8.07  8.64    6.47   4.98    8.13</span>
<span class="co">#&gt;  5 T21101315  5.49  9.83  8.13  5.73    9.38  8.15    6.43   5.03    7.70</span>
<span class="co">#&gt;  6 T21101316  5.58  9.85  8.55  5.13    9.40  8.71    6.56   4.87    7.75</span>
<span class="co">#&gt;  7 T21101317  5.32 10.0   6.99  5.86    8.08  9.25    6.86   4.72    7.88</span>
<span class="co">#&gt;  8 T21101318  5.49  9.72  7.47  5.16    6.67  7.37    6.70   4.92    7.50</span>
<span class="co">#&gt;  9 T21101319  5.57  9.88  7.97  5.40    7.91  8.06    6.58   5.06    8.16</span>
<span class="co">#&gt; 10 T21101320  5.16  9.87  7.42  5.50    7.43  8.68    6.55   4.85    8.20</span>
<span class="co">#&gt; # ... with 26 more rows, and 8 more variables: MSMO1 &lt;dbl&gt;, FDFT1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SC5DL &lt;dbl&gt;, LIPA &lt;dbl&gt;, CEL &lt;dbl&gt;, TM7SF2 &lt;dbl&gt;, NSDHL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SOAT2 &lt;dbl&gt;</span></code></pre></div>
<p>This transposed data frame has the gene names as the column names and the sample names as a column of character (<code>chr</code>) values. Notice that the data itself is 17 genes measured on 36 samples. Before transposition, we had 37 columns because the feature names were stored in the first column. After transposition, we have 36 rows but 18 columns: the first column stores the sample names. This transposed data frame (after filtering to match the response data) will be supplied to the <code>assayData_df</code> argument in the <code>create_Omics</code> function. (<em>See the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Create_Omics_Objects.html">Creating <code>Omics</code> Data Objects</a> vignette for more information on creating <code>Omics</code>-class objects.</em>)</p>
</div>
<div id="subsetting-a-tidy-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#subsetting-a-tidy-data-frame" class="anchor"></a>3.3 Subsetting a Tidy Data Frame</h2>
<p>If ever we need to extract individual components of a tidy data frame, we can use the <code>assay[row, col]</code> syntax. If we need entire measurements (columns), then we can call the column by name with the <code>assay$ColName</code> syntax. For example,</p>
<ul>
<li>If we need the second row of <code>assayT_df</code>—corresponding to Sample “T21101312”—then we type</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assayT_df[<span class="dv">2</span>, ]
<span class="co">#&gt; # A tibble: 1 x 18</span>
<span class="co">#&gt;   Sample    SOAT1   LSS  SQLE   EBP CYP51A1 DHCR7 CYP27B1 DHCR24 HSD17B7</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 T21101312  5.52  9.78  8.06  5.12    8.21  8.33    6.47   5.06    7.63</span>
<span class="co">#&gt; # ... with 8 more variables: MSMO1 &lt;dbl&gt;, FDFT1 &lt;dbl&gt;, SC5DL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   LIPA &lt;dbl&gt;, CEL &lt;dbl&gt;, TM7SF2 &lt;dbl&gt;, NSDHL &lt;dbl&gt;, SOAT2 &lt;dbl&gt;</span></code></pre></div>
<p>Notice that the <code>tibble</code> object has 1 row and 18 columns. - If we need the third column of <code>assayT_df</code>—corresponding to Gene “LSS”—then we type</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assayT_df[, <span class="dv">3</span>]
<span class="co">#&gt; # A tibble: 36 x 1</span>
<span class="co">#&gt;      LSS</span>
<span class="co">#&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1  9.77</span>
<span class="co">#&gt;  2  9.78</span>
<span class="co">#&gt;  3  8.11</span>
<span class="co">#&gt;  4  8.67</span>
<span class="co">#&gt;  5  9.83</span>
<span class="co">#&gt;  6  9.85</span>
<span class="co">#&gt;  7 10.0 </span>
<span class="co">#&gt;  8  9.72</span>
<span class="co">#&gt;  9  9.88</span>
<span class="co">#&gt; 10  9.87</span>
<span class="co">#&gt; # ... with 26 more rows</span></code></pre></div>
<p>This <code>tibble</code> object has 36 rows and 1 column. - If we need the intersection of these two (the expression level of Gene “LSS” in Sample “T21101312”), then we type</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assayT_df[<span class="dv">2</span>, <span class="dv">3</span>, drop =<span class="st"> </span><span class="ot">TRUE</span>]
<span class="co">#&gt; [1] 9.780137</span></code></pre></div>
<p>This would normally be a 1 by 1 <code>tibble</code> (which isn’t terribly helpful), so we add the <code>drop = TRUE</code> argument to “drop” the dimensions of the table. This gives us a single basic number. - If we need the third column of <code>assayT_df</code>, but we want the result back as a vector instead of a data frame, we call the column by name:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assayT_df<span class="op">$</span>LSS
<span class="co">#&gt;  [1]  9.772752  9.780137  8.109067  8.667270  9.828498  9.852128 10.004496</span>
<span class="co">#&gt;  [8]  9.719076  9.883177  9.871709  7.892335  8.263840  8.215201  9.133894</span>
<span class="co">#&gt; [15] 10.116890  9.107010  9.820948  9.057552  9.268507  9.382549  7.852329</span>
<span class="co">#&gt; [22]  8.585521  8.949471  8.931112  9.597493  9.657331  7.800871  9.858547</span>
<span class="co">#&gt; [29]  9.758955  8.383229 10.562910  9.815668 10.013560 11.094211  9.217539</span>
<span class="co">#&gt; [36] 10.667547</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="import-and-join-response-data" class="section level1">
<h1 class="hasAnchor">
<a href="#import-and-join-response-data" class="anchor"></a>4. Import and Join Response Data</h1>
<p>We now have an appropriate pathway set list and a tidy -Omics assay data frame. Let’s imagine that your phenotype data looks something like this:</p>
<div class="figure">
<img src="example_pInfo_data.PNG">
</div>
<p>We now need to import this response information. We can use the <code>read_csv</code> function once again:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pInfo_path &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">"extdata"</span>, <span class="st">"ex_pInfo_subset.csv"</span>,
                          <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
pInfo_df &lt;-<span class="st"> </span><span class="kw">read_csv</span>(pInfo_path)</code></pre></div>
<pre><code>#&gt; Parsed with column specification:
#&gt; cols(
#&gt;   Sample = col_character(),
#&gt;   eventTime = col_double(),
#&gt;   eventObserved = col_logical()
#&gt; )</code></pre>
<p>This data frame has a column for the sample labels (<code>Sample</code>) and the response information. In this case, our response is a survival response with an event time and observation indicator.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pInfo_df
<span class="co">#&gt; # A tibble: 36 x 3</span>
<span class="co">#&gt;    Sample    eventTime eventObserved</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;dbl&gt; &lt;lgl&gt;        </span>
<span class="co">#&gt;  1 T21101311     14.2  TRUE         </span>
<span class="co">#&gt;  2 T21101312      1.00 TRUE         </span>
<span class="co">#&gt;  3 T21101313      6.75 FALSE        </span>
<span class="co">#&gt;  4 T21101314      8.50 TRUE         </span>
<span class="co">#&gt;  5 T21101315      7.25 FALSE        </span>
<span class="co">#&gt;  6 T21101316      5.00 TRUE         </span>
<span class="co">#&gt;  7 T21101317     20.0  TRUE         </span>
<span class="co">#&gt;  8 T21101318     13.2  FALSE        </span>
<span class="co">#&gt;  9 T21101319      7.75 FALSE        </span>
<span class="co">#&gt; 10 T21101320      9.00 FALSE        </span>
<span class="co">#&gt; # ... with 26 more rows</span></code></pre></div>
<p>This <code>pInfo</code> data frame has the sample names as a column of character values, just like the transposed assay data frame. This is crucially important for the “joining” step. We can use the <code>inner_join</code> function from the <code>dplyr</code> library to retain only the rows of the <code>assayT_df</code> data frame which have responses in the <code>pInfo</code> data frame and vice versa. This way, every response in the phenotype data has matching predictors in the assay, and every recorded predictor in the assay matches a response in the phenotype data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">(joinedExperiment_df &lt;-<span class="st"> </span><span class="kw">inner_join</span>(pInfo_df, assayT_df, <span class="dt">by =</span> <span class="st">"Sample"</span>))
<span class="co">#&gt; # A tibble: 36 x 20</span>
<span class="co">#&gt;    Sample    eventTime eventObserved SOAT1   LSS  SQLE   EBP CYP51A1 DHCR7</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 T21101311     14.2  TRUE           5.37  9.77  7.74  4.68    8.27  8.32</span>
<span class="co">#&gt;  2 T21101312      1.00 TRUE           5.52  9.78  8.06  5.12    8.21  8.33</span>
<span class="co">#&gt;  3 T21101313      6.75 FALSE          5.89  8.11  7.00  5.78    8.20  8.39</span>
<span class="co">#&gt;  4 T21101314      8.50 TRUE           5.62  8.67  8.59  5.64    8.07  8.64</span>
<span class="co">#&gt;  5 T21101315      7.25 FALSE          5.49  9.83  8.13  5.73    9.38  8.15</span>
<span class="co">#&gt;  6 T21101316      5.00 TRUE           5.58  9.85  8.55  5.13    9.40  8.71</span>
<span class="co">#&gt;  7 T21101317     20.0  TRUE           5.32 10.0   6.99  5.86    8.08  9.25</span>
<span class="co">#&gt;  8 T21101318     13.2  FALSE          5.49  9.72  7.47  5.16    6.67  7.37</span>
<span class="co">#&gt;  9 T21101319      7.75 FALSE          5.57  9.88  7.97  5.40    7.91  8.06</span>
<span class="co">#&gt; 10 T21101320      9.00 FALSE          5.16  9.87  7.42  5.50    7.43  8.68</span>
<span class="co">#&gt; # ... with 26 more rows, and 11 more variables: CYP27B1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   DHCR24 &lt;dbl&gt;, HSD17B7 &lt;dbl&gt;, MSMO1 &lt;dbl&gt;, FDFT1 &lt;dbl&gt;, SC5DL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   LIPA &lt;dbl&gt;, CEL &lt;dbl&gt;, TM7SF2 &lt;dbl&gt;, NSDHL &lt;dbl&gt;, SOAT2 &lt;dbl&gt;</span></code></pre></div>
<p><strong>This requires you to have a <em>key</em> column in both data frames with the same name.</strong> If the key column was called “Sample” in the <code>pInfo_df</code> data set but “SampleID” in the assay, then the <code>by</code> argument should be changed to <code>c("Sample" = "SampleID")</code>. It’s much nicer to just keep them with the same names, however. Moreover, it is vitally important that you check your sample IDs. Obviously the expressed genetic data should pair with the phenotype information, but <strong>it is your responsibility as the user to confirm that the assay rows match the correct responses.</strong> We can show you the proper methods to check your data, but you are ultimately responsible to use this package properly.</p>
<hr>
<p></p>
</div>
<div id="example-tidy-assay-and-gene-set" class="section level1">
<h1 class="hasAnchor">
<a href="#example-tidy-assay-and-gene-set" class="anchor"></a>5. Example Tidy Assay and Gene Set</h1>
<p>Included in this package, we have a small tidy assay and corresponding gene subset list. We will load and inspect this assay. This data set has 656 gene expression measurements on 250 colon cancer patients. Further notice that the assay and overall survival response information have already been matched.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"colonSurv_df"</span>)
colonSurv_df
<span class="co">#&gt; # A tibble: 250 x 658</span>
<span class="co">#&gt;    OS_time OS_event   JUN  SOS2  PAK3  RAF1 PRKCB   BTC  SHC1 PRKCA  ELK1</span>
<span class="co">#&gt;      &lt;dbl&gt;    &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1  64.9          0  9.29  5.48  8.21  8.03  5.49  6.65  8.26  8.94  7.38</span>
<span class="co">#&gt;  2  59.8          0  9.13  6.35  8.33  7.94  6.26  7.02  8.39  9.61  7.53</span>
<span class="co">#&gt;  3  62.4          0  9.37  5.67  7.82  7.74  6.05  7.52  8.69  8.40  7.25</span>
<span class="co">#&gt;  4  54.5          0 10.6   4.94  8.79  7.64  5.37  6.87  7.81  9.80  7.79</span>
<span class="co">#&gt;  5  46.3          1  8.70  5.60  8.75  8.05  6.07  6.49  8.45  8.21  7.60</span>
<span class="co">#&gt;  6  55.9          0  9.78  5.36  7.56  8.07  5.90  6.39  8.87  8.22  7.35</span>
<span class="co">#&gt;  7  58.0          0  9.22  5.05  8.20  7.80  5.55  6.86  8.28  8.97  7.43</span>
<span class="co">#&gt;  8  54.0          0 10.3   5.33  7.82  7.89  6.27  6.25  8.66  9.71  7.38</span>
<span class="co">#&gt;  9   0.427        1 10.8   5.07  7.63  7.69  5.48  7.57  8.36  9.69  6.66</span>
<span class="co">#&gt; 10  41.4          0  9.52  5.50  7.48  7.53  5.71  7.33  8.54  8.14  6.88</span>
<span class="co">#&gt; # ... with 240 more rows, and 647 more variables: NRG1 &lt;dbl&gt;, PAK2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   MTOR &lt;dbl&gt;, PAK4 &lt;dbl&gt;, MAP2K4 &lt;dbl&gt;, EIF4EBP1 &lt;dbl&gt;, BAD &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PRKCG &lt;dbl&gt;, NRG3 &lt;dbl&gt;, MAPK9 &lt;dbl&gt;, ERBB4 &lt;dbl&gt;, MAPK10 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PTK2 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, ERBB3 &lt;dbl&gt;, MAP2K2 &lt;dbl&gt;, TGFA &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BRAF &lt;dbl&gt;, MAP2K1 &lt;dbl&gt;, MAP2K7 &lt;dbl&gt;, ABL1 &lt;dbl&gt;, NRG2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   AKT1 &lt;dbl&gt;, ABL2 &lt;dbl&gt;, AKT2 &lt;dbl&gt;, SHC4 &lt;dbl&gt;, RPS6KB1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   RPS6KB2 &lt;dbl&gt;, AKT3 &lt;dbl&gt;, NRAS &lt;dbl&gt;, GRB2 &lt;dbl&gt;, AREG &lt;dbl&gt;,</span>
<span class="co">#&gt; #   STAT5B &lt;dbl&gt;, MAPK3 &lt;dbl&gt;, STAT5A &lt;dbl&gt;, PAK6 &lt;dbl&gt;, SOS1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   MYC &lt;dbl&gt;, MAPK1 &lt;dbl&gt;, NCK1 &lt;dbl&gt;, PIK3R5 &lt;dbl&gt;, NRG4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   HRAS &lt;dbl&gt;, MAPK8 &lt;dbl&gt;, EGFR &lt;dbl&gt;, GSK3B &lt;dbl&gt;, CBLB &lt;dbl&gt;,</span>
<span class="co">#&gt; #   KRAS &lt;dbl&gt;, CBL &lt;dbl&gt;, SHC3 &lt;dbl&gt;, CDKN1B &lt;dbl&gt;, CDKN1A &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EGF &lt;dbl&gt;, EREG &lt;dbl&gt;, ARAF &lt;dbl&gt;, NCK2 &lt;dbl&gt;, SRC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PIK3R3 &lt;dbl&gt;, CAMK2A &lt;dbl&gt;, CAMK2B &lt;dbl&gt;, CAMK2D &lt;dbl&gt;, CAMK2G &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PAK1 &lt;dbl&gt;, CBLC &lt;dbl&gt;, CRK &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, PIK3CB &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CRKL &lt;dbl&gt;, PIK3CD &lt;dbl&gt;, GAB1 &lt;dbl&gt;, PLCG1 &lt;dbl&gt;, PLCG2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SHC2 &lt;dbl&gt;, HBEGF &lt;dbl&gt;, PIK3CG &lt;dbl&gt;, PIK3R1 &lt;dbl&gt;, PIK3R2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EPHB2 &lt;dbl&gt;, EPHB4 &lt;dbl&gt;, EFNA5 &lt;dbl&gt;, PXN &lt;dbl&gt;, CDC42 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB3 &lt;dbl&gt;, RRAS &lt;dbl&gt;, GRB7 &lt;dbl&gt;, SYNJ1 &lt;dbl&gt;, EPHB3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB1 &lt;dbl&gt;, DNM1 &lt;dbl&gt;, MAP4K4 &lt;dbl&gt;, GRIA1 &lt;dbl&gt;, EPHB1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   ROCK1 &lt;dbl&gt;, ITSN1 &lt;dbl&gt;, RAP1A &lt;dbl&gt;, RAC1 &lt;dbl&gt;, RAP1B &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB2 &lt;dbl&gt;, WASL &lt;dbl&gt;, TF &lt;dbl&gt;, …</span></code></pre></div>
<p>We also have a small list of 15 pathways which correspond to our example colon cancer assay. To create a toy example, we have curated this list to include seven significant pathways and eight non-significant pathways.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"colon_pathwaySet"</span>)
colon_pathwaySet
<span class="co">#&gt; Object with Class(es) 'pathwaySet', 'list' [package 'pathwayPCA'] with 2 elements: </span>
<span class="co">#&gt;  $ pathways:List of 15</span>
<span class="co">#&gt;  $ TERMS   : chr [1:15] "KEGG_PENTOSE_PHOSPHATE_PATHWAY" ...</span>
<span class="kw">str</span>(colon_pathwaySet<span class="op">$</span>pathways, <span class="dt">list.len =</span> <span class="dv">10</span>)
<span class="co">#&gt; List of 15</span>
<span class="co">#&gt;  $ pathway3   : chr [1:27] "RPE" "RPIA" "PGM2" "PGLS" ...</span>
<span class="co">#&gt;  $ pathway60  : chr [1:64] "RPE65" "CYP3A5" "UGT2B28" "CYP4A11" ...</span>
<span class="co">#&gt;  $ pathway87  : chr [1:87] "JUN" "SOS2" "PAK3" "RAF1" ...</span>
<span class="co">#&gt;  $ pathway120 : chr [1:89] "HLA-DOA" "HLA-DOB" "KLRC3" "KLRD1" ...</span>
<span class="co">#&gt;  $ pathway176 : chr [1:54] "CASP9" "SOS2" "E2F1" "PRKCB" ...</span>
<span class="co">#&gt;  $ pathway177 : chr [1:30] "HLA-DRB4" "HLA-DRB5" "HLA-DOA" "HLA-DOB" ...</span>
<span class="co">#&gt;  $ pathway187 : chr [1:16] "IKBKG" "CHUK" "EP300" "RELA" ...</span>
<span class="co">#&gt;  $ pathway266 : chr [1:11] "PRF1" "DFFA" "DFFB" "HMGB2" ...</span>
<span class="co">#&gt;  $ pathway390 : chr [1:29] "JUN" "BAG4" "CASP8" "MAPK8" ...</span>
<span class="co">#&gt;  $ pathway413 : chr [1:23] "PLD1" "RAF1" "EPHB2" "VAV1" ...</span>
<span class="co">#&gt;   [list output truncated]</span></code></pre></div>
<p>The pathway set list and tidy assay (with matched phenotype information) are all the information we need to create an <code>Omics</code>-class data object.</p>
<hr>
<p></p>
</div>
<div id="review" class="section level1">
<h1 class="hasAnchor">
<a href="#review" class="anchor"></a>6. Review</h1>
<p>We now summarize our steps so far. We have</p>
<ol style="list-style-type: decimal">
<li>Imported a <code>.gmt</code> file and saved the gene sets therein as a <code>pathwaySet</code> object using the <code>read_gmt</code> function.</li>
<li>Imported an assay <code>.csv</code> file with the <code>read_csv</code> function from the <code>readr</code> package, and transposed this data frame into “tidy” form with the <code>transpose_assay</code> function.</li>
<li>Imported a phenotype information <code>.csv</code> file, and joined it to the assay data frame with the <code>inner_join</code> function from the <code>dplyr</code> package.</li>
</ol>
<p>Now we are prepared to create our first <code>Omics</code>-class object for analysis with either AES-PCA or Supervised PCA. Please read the next vignette: <a href="https://gabrielodom.github.io/pathwayPCA/articles/C3-Create_Omics_Objects.html">Creating <code>Omics</code> Data Objects</a>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">1. Overview</a></li>
      <li>
<a href="#gmt-files">2. GMT Files</a><ul class="nav nav-pills nav-stacked">
<li><a href="#gmt-format-description">2.1 GMT Format Description</a></li>
      <li><a href="#import-gmt-files-with-read_gmt">2.2 Import GMT files with <code>read_gmt</code></a></li>
      <li><a href="#creating-your-own-pathwayset-list">2.3 Creating Your Own <code>pathwaySet</code> List</a></li>
      </ul>
</li>
      <li>
<a href="#import-and-tidy-an-assay-matrix">3. Import and Tidy an Assay Matrix</a><ul class="nav nav-pills nav-stacked">
<li><a href="#import-with-readr">3.1 Import with <code>readr</code></a></li>
      <li><a href="#tidy-the-assay-data-frame">3.2 Tidy the Assay Data Frame</a></li>
      <li><a href="#subsetting-a-tidy-data-frame">3.3 Subsetting a Tidy Data Frame</a></li>
      </ul>
</li>
      <li><a href="#import-and-join-response-data">4. Import and Join Response Data</a></li>
      <li><a href="#example-tidy-assay-and-gene-set">5. Example Tidy Assay and Gene Set</a></li>
      <li><a href="#review">6. Review</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d">Lily Wang</a>, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
