<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Integrative Pathway Analysis with pathwayPCA • pathwayPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Integrative Pathway Analysis with pathwayPCA">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathwayPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    User Guides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction_to_pathwayPCA.html">Integrative Pathway Analysis with pathwayPCA</a>
    </li>
    <li>
      <a href="../articles/Supplement1-Quickstart_Guide.html">Suppl. Ch. 1 - Quickstart Guide for New R Users</a>
    </li>
    <li>
      <a href="../articles/Supplement2-Importing_Data.html">Suppl. Ch. 2 - Import and Tidy Data</a>
    </li>
    <li>
      <a href="../articles/Supplement3-Create_Omics_Objects.html">Suppl. Ch. 3 - Creating Data Objects</a>
    </li>
    <li>
      <a href="../articles/Supplement4-Methods_Walkthrough.html">Suppl. Ch. 4 - Test Pathway Significance</a>
    </li>
    <li>
      <a href="../articles/Supplement5-Analyse_Results.html">Suppl. Ch.  5 - Visualizing the Results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Version Logs</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Integrative Pathway Analysis with pathwayPCA</h1>
                        <h4 class="author">Lily Wang and Gabriel Odom</h4>
            
            <h4 class="date">2018-12-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gabrielodom/pathwayPCA/blob/master/vignettes/Introduction_to_pathwayPCA.Rmd"><code>vignettes/Introduction_to_pathwayPCA.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction_to_pathwayPCA.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>1. Introduction</h1>
<p><code>pathwayPCA</code> provides implementation of the methodogly described in Chen (2011) and Chen et al. (2010). <code>pathwayPCA</code> allows users to:</p>
<ol style="list-style-type: decimal">
<li>Test pathway association with binary, continuous, or survival phenotypes.</li>
<li>Extract relevant genes in the pathways using the SPCA and AESPCA approaches.</li>
<li>Compute PCAs based on the selected genes. These estimated latent variables represent pathway activities for individual subjects, which can then be used to perform integrative pathway analysis, such as multi-omics analysis, or predicting survival time.</li>
<li>Extract relevant genes that drive pathway significance as well as data corresponding to these relevant genes for additional <em>ad hoc</em> anlaysis.</li>
<li>Perform analyses with enhanced computational efficiency with parallel computing and enhanced data safety with S4-class data objects.</li>
<li>Can be extended to studies with complex experimental designs, with multiple covariates, and with interaction effects, e.g., testing whether pathway association with clinical phenotype is different between male and female subjects.</li>
</ol>
<p>Throughout this vignette, we will make use of the <code>tidyverse</code> suite of utility packages (<a href="https://www.tidyverse.org/" class="uri">https://www.tidyverse.org/</a>). The <code>tidyverse</code> and <code>pathwayPCA</code> can be loaded into R using:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(pathwayPCA)</code></pre></div>
<p></p>
<hr>
<p></p>
</div>
<div id="case-study-identifying-significant-pathways-in-protein-expressions-associated-with-survival-outcome-in-ovarian-cancer-data" class="section level1">
<h1 class="hasAnchor">
<a href="#case-study-identifying-significant-pathways-in-protein-expressions-associated-with-survival-outcome-in-ovarian-cancer-data" class="anchor"></a>2. Case study: identifying significant pathways in protein expressions associated with survival outcome in ovarian cancer data</h1>
<div id="overian-cancer-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#overian-cancer-dataset" class="anchor"></a>2.1. Overian cancer dataset</h2>
<p>For this example, we will use the mass spectrometry based global proteomics data for ovarian cancer recently generated by the Clinical Proteomic Tumor Analysis Consortium (CPTAC). The normalized protein abundance expression dataset can be obtained from the LinkedOmics database at <a href="http://linkedomics.org/data_download/TCGA-OV/" class="uri">http://linkedomics.org/data_download/TCGA-OV/</a>. We used the dataset “Proteome (PNNL, Gene level)” which was generated by the Pacific Northwest National Laboratory (PNNL). We then removed one subject due to missingness, we retained the least-missing 80% of proteins, and we imputed the remaining 4.5% of missing values via k-nearest-neighbors (as implemented in the Bioconductor package <code>impute</code> under default settings). This yields 83 complete observations of 5162 proteins.</p>
</div>
<div id="creating-an-omics-data-object-for-pathway-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-an-omics-data-object-for-pathway-analysis" class="anchor"></a>2.2. Creating an <code>Omics</code> data object for pathway analysis</h2>
<p>First, we need to create an <code>Omics</code>-class data object that stores</p>
<ol style="list-style-type: decimal">
<li>the expression dataset</li>
<li>phenotype information for the samples</li>
<li>a collection of pathways</li>
</ol>
<div id="expression-and-phenotype-data" class="section level3">
<h3 class="hasAnchor">
<a href="#expression-and-phenotype-data" class="anchor"></a>2.2.1 Expression and Phenotype Data</h3>
<p>We can obtain (1) and (2) for the ovarian cancer dataset by loading the dataset <code>ovarian_PNNL_survival.RDS</code> included in the <code>pathwayPCA</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dataDir_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(
  <span class="st">"extdata"</span>, <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>
)
ovSurv_df &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(dataDir_path, <span class="st">"/ovarian_PNNL_survival.RDS"</span>))</code></pre></div>
<p>The <code>ovSurv_df</code> dataset is a data frame with protein expression levels and survival outcome matched by sample IDs. The variables (columns) include overall survival time and censoring status, as well as expression data for 5162 proteins for each of the 83 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovSurv_df</code></pre></div>
</div>
<div id="pathway-collections" class="section level3">
<h3 class="hasAnchor">
<a href="#pathway-collections" class="anchor"></a>2.2.2 Pathway Collections</h3>
<p>For the collection of pathways in (3), we need to specify a <code>.gmt</code> file. Gene Matrix formatted files store each pathway as a single line. Each line contains an ID (column <code>TERMS</code>), an optional description (column <code>description</code>), and the genes in the pathway (all subsequent columns). Pathway collections in <code>.gmt</code> form can be downloaded from the MSigDB database at <a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp" class="uri">http://software.broadinstitute.org/gsea/msigdb/collections.jsp</a>. Additionally, for <a href="https://www.wikipathways.org/index.php/WikiPathways">WikiPathways</a>, one can download monthly data releases in <code>.gmt</code> format using the <code>dowloadPathwayArchive()</code> function in the <code>rWikiPathways</code> package <a href="https://bioconductor.org/packages/release/bioc/html/rWikiPathways.html">from Bioconductor</a>. For example, the following commands downloads to your current directory the October 2017 release of WikiPathways:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rWikiPathways)
<span class="kw">downloadPathwayArchive</span>(
  <span class="dt">date =</span> <span class="st">"20171010"</span>, <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>, <span class="dt">format =</span> <span class="st">"gmt"</span>
) </code></pre></div>
<pre><code>trying URL 'http://data.wikipathways.org/20171010/gmt/wikipathways-20171010-gmt-Homo_sapiens.gmt'
Content type '' length 137254 bytes (134 KB)
downloaded 134 KB</code></pre>
<pre><code>#&gt; [1] "wikipathways-20171010-gmt-Homo_sapiens.gmt"</code></pre>
<p>As an example, <code>pathwayPCA</code> includes the June 2018 Wikipathways collection for <em>homo sapiens</em>, which can be loaded using the read_gmt function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wikipathways_PC &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_gmt.html">read_gmt</a></span>(
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(dataDir_path, <span class="st">"/wikipathways_human_symbol.gmt"</span>),
  <span class="dt">description =</span> <span class="ot">TRUE</span>
)</code></pre></div>
</div>
<div id="create-an-omicssurv-data-container" class="section level3">
<h3 class="hasAnchor">
<a href="#create-an-omicssurv-data-container" class="anchor"></a>2.2.3 Create an <code>OmicsSurv</code> Data Container</h3>
<p>Now that we have these three data components, we create an <code>OmicsSurv</code> data container. Note that when <code>assayData_df</code> and <code>response</code> are supplied from two different files, the user need to make sure these data sets are match by sample IDs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ov_OmicsSurv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CreateOmics.html">CreateOmics</a></span>(
  <span class="co"># protein expression data</span>
  <span class="dt">assayData_df =</span> ovSurv_df[, <span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)], 
  <span class="co"># pathway collection</span>
  <span class="dt">pathwayCollection_ls =</span> wikipathways_PC,
  <span class="co"># survival phenotypes</span>
  <span class="dt">response =</span> ovSurv_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],
  <span class="co"># phenotype is survival data</span>
  <span class="dt">respType =</span> <span class="st">"survival"</span>,
  <span class="co"># retain pathways with &gt; 5 proteins</span>
  <span class="dt">minPathSize =</span> <span class="dv">5</span>  
)

  <span class="op">==</span><span class="er">====</span><span class="st">  </span>Creating object of class OmicsSurv  <span class="op">==</span><span class="er">=====</span>
The input pathway database included <span class="dv">5831</span> unique features.
The input assay dataset included <span class="dv">5162</span> features.
Only pathways with at least <span class="dv">5</span> or more features included <span class="cf">in</span> the assay dataset are
  <span class="kw">tested</span> (specified by minPathSize parameter). There are <span class="dv">324</span> pathways which meet
  this criterion.
Because pathwayPCA is a self<span class="op">-</span>contained <span class="kw">test</span> (PMID<span class="op">:</span><span class="st"> </span><span class="dv">17303618</span>), only features <span class="cf">in</span>
  both assay data and pathway database are considered <span class="cf">for</span> analysis. There are <span class="dv">2064</span>
  such features shared by the input assay and pathway database.</code></pre></div>
<p>To see a summary of the omics data object we just created, simply type the name of the object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ov_OmicsSurv
<span class="co">#&gt; Formal class 'OmicsSurv' [package "pathwayPCA"] with 6 slots</span>
<span class="co">#&gt;   ..@ eventTime            : num [1:83] 2279 3785 2154 2553 2856 ...</span>
<span class="co">#&gt;   ..@ eventObserved        : logi [1:83] TRUE TRUE TRUE TRUE TRUE TRUE ...</span>
<span class="co">#&gt;   ..@ assayData_df         :Classes 'tbl_df', 'tbl' and 'data.frame':    83 obs. of  5162 variables:</span>
<span class="co">#&gt;   ..@ sampleIDs_char       : chr [1:83] "TCGA.09.1664" "TCGA.13.1484" "TCGA.13.1488" "TCGA.13.1489" ...</span>
<span class="co">#&gt;   ..@ pathwayCollection    :List of 4</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span>
<span class="co">#&gt;   ..@ trimPathwayCollection:List of 5</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span></code></pre></div>
</div>
</div>
<div id="testing-pathway-association-with-phenotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#testing-pathway-association-with-phenotypes" class="anchor"></a>2.3. Testing pathway association with phenotypes</h2>
<div id="method-description" class="section level3">
<h3 class="hasAnchor">
<a href="#method-description" class="anchor"></a>2.3.1 Method Description</h3>
<p>Once we have a valid <code>Omics</code> object, we can perform pathway analysis using the AES-PCA (Adaptive, Elastic-net, Sparse PCA) or Supervised PCA methodology described in Chen (2011) and Chen et al. (2010).</p>
<p>Briefly, in the AES-PCA method, we first extract <em>principal components</em> (PCs) representing activities within each pathway using a dimension reduction approach based on adaptive, elastic-net, sparse principal component analysis (<a href="https://doi.org/10.2202/1544-6115.1697" class="uri">https://doi.org/10.2202/1544-6115.1697</a>). The estimated latent variables are then tested against phenotypes using a permutation test that permutes sample labels. Note that the AESPCA approach does not use the response information to estiamte pathway PCs, so it is an unsupervised approach.</p>
<p>This is in contrast to the Supervised PCA approach, where a selected subset of genes most associated with disease outcome are used to estimate the latent variable for a pathway (<a href="https://doi.org/10.1002/gepi.20532" class="uri">https://doi.org/10.1002/gepi.20532</a>). Because of this gene selection step, the test statistics from the Supervised PCA model can no longer be approximated well using the Student’s <span class="math inline">\(t\)</span>-distribution. To account for the gene selection step, <code>pathwayPCA</code> estimates <span class="math inline">\(p\)</span>-values from a two-component mixture of Gumbel exteme value distributions instead.</p>
</div>
<div id="implementation" class="section level3">
<h3 class="hasAnchor">
<a href="#implementation" class="anchor"></a>2.3.2 Implementation</h3>
<p>Because the syntax for performing Supervised PCA is nearly identical to the AES-PCA syntax, we will illustrate only the AES-PCA workflow below. Note that when the value supplied to the <code>numReps</code> argument is greater than 0, the <code>AESPCA_pvals()</code> function employs a parametric test when estimating pathway significance via the following model: “phenotype ~ PC1”. Pathway <span class="math inline">\(p\)</span>-values are estimated based on a likelihood ratio test that compares this model to a null model (with intercept only). Because pathways should represent a single biological effect, we recommend extracting only one PC be estimated (i.e., <code>numPCs = 1</code>). Additionally, we can specify a method to control the false discovery rate (FDR) or family-wise error rate with the <code>adjustment</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovarian_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="co"># The Omics data container</span>
  <span class="dt">object =</span> ov_OmicsSurv,
  <span class="co"># One principal component per pathway</span>
  <span class="dt">numPCs =</span> <span class="dv">1</span>,
  <span class="co"># Use parallel computing with 2 cores</span>
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="co"># Estimate the p-values parametrically</span>
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="co"># Control FDR via Benjamini-Hochberg</span>
  <span class="dt">adjustment =</span> <span class="st">"BH"</span>
)
Part <span class="dv">1</span><span class="op">:</span><span class="st"> </span>Calculate Pathway AES<span class="op">-</span>PCs
Initializing Computing Cluster<span class="op">:</span><span class="st"> </span>DONE
Extracting Pathway PCs <span class="cf">in</span> Parallel<span class="op">:</span><span class="st"> </span>DONE

Part <span class="dv">2</span><span class="op">:</span><span class="st"> </span>Calculate Pathway p<span class="op">-</span>Values
Initializing Computing Cluster<span class="op">:</span><span class="st"> </span>DONE
Extracting Pathway p<span class="op">-</span>Values <span class="cf">in</span> Parallel<span class="op">:</span><span class="st"> </span>DONE

Part <span class="dv">3</span><span class="op">:</span><span class="st"> </span>Adjusting p<span class="op">-</span>Values and Sorting Pathway p<span class="op">-</span>Value Data Frame
DONE</code></pre></div>
<p>This <code>ovarian_aespcOut</code> object contains 3 components: a table of pathway <span class="math inline">\(p\)</span>-values, AESPCA-estimated PCs of each sample from each pathway, and the loadings of each protein onto the AESPCs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(ovarian_aespcOut)
<span class="co">#&gt; [1] "pVals_df"    "PCs_ls"      "loadings_ls"</span></code></pre></div>
</div>
<div id="show-the-top-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#show-the-top-pathways" class="anchor"></a>2.2.3 Show the Top Pathways</h3>
<p>For this ovarian cancer dataset, the top 10 most significant pathways identified by AES-PCA are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovarian_aespcOut<span class="op">$</span>pVals_df
<span class="co">#&gt; # A tibble: 324 x 6</span>
<span class="co">#&gt;    pathways n_tested terms  description                        rawp FDR_BH</span>
<span class="co">#&gt;  * &lt;chr&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;                             &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 path22         30 WP195  IL-1 signaling pathway          0.00247  0.279</span>
<span class="co">#&gt;  2 path253        16 WP3858 Toll-like Receptor Signaling    0.00463  0.279</span>
<span class="co">#&gt;  3 path410        18 WP363  Wnt Signaling Pathway           0.00485  0.279</span>
<span class="co">#&gt;  4 path33         24 WP2447 Amyotrophic lateral sclerosis ~ 0.00582  0.279</span>
<span class="co">#&gt;  5 path128        22 WP2036 TNF related weak inducer of ap~ 0.00641  0.279</span>
<span class="co">#&gt;  6 path95         12 WP3298 Melatonin metabolism and effec~ 0.00688  0.279</span>
<span class="co">#&gt;  7 path249        13 WP3850 Factors and pathways affecting~ 0.00700  0.279</span>
<span class="co">#&gt;  8 path208        10 WP3617 Photodynamic therapy-induced N~ 0.00714  0.279</span>
<span class="co">#&gt;  9 path133         8 WP34   Ovarian Infertility Genes       0.00801  0.279</span>
<span class="co">#&gt; 10 path242        14 WP3851 TLR4 Signaling and Tolerance    0.00870  0.279</span>
<span class="co">#&gt; # ... with 314 more rows</span></code></pre></div>
<p>In this table, <code>n_tested</code> indicates the number of genes in both the pathway and assay data, i.e. the number of genes in a pathway that we actually tested for association with phenotype in AESPCA.</p>
<p>We are interested in graphing the top dozen or so pathways by their level of significance as well. This code takes in the pathway <span class="math inline">\(p\)</span>-values data frame from the AESPCA method output and <a href="https://rpubs.com/mm-c/gather-and-spread">gathers</a> it into a “tidy” data frame (compatible with <code>ggplot</code>). We also add on a column for the negative natural logarithm of the pathway <span class="math inline">\(p\)</span>-values (called <code>score</code>) and recode the label of the <span class="math inline">\(p\)</span>-value adjustment method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovOutGather_df &lt;-
<span class="st">  </span><span class="co"># Take in the results data frame,</span>
<span class="st">  </span>ovarian_aespcOut<span class="op">$</span>pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># add the score variable,</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(rawp)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># subset the top 20 pathways, and</span>
<span class="st">  </span><span class="kw">slice</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># remove the columns we don't need</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>pathways, <span class="op">-</span>n_tested, <span class="op">-</span>rawp, <span class="op">-</span>FDR_BH)</code></pre></div>
<p>Now we plot the pathway significance level for the top 20 pathways.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(ovOutGather_df) <span class="op">+</span>
<span class="st">  </span><span class="co"># set overall appearance of the plot</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="co"># Define the dependent and independent variables</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(terms, score), <span class="dt">y =</span> score) <span class="op">+</span>
<span class="st">  </span><span class="co"># From the defined variables, create a vertical bar chart</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">position =</span> <span class="st">"dodge"</span>, <span class="dt">fill =</span> <span class="st">"#F47321"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add pathway labels</span>
<span class="st">  </span><span class="kw">geom_text</span>(
    <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(terms, score), <span class="dt">label =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(description, score), <span class="dt">y =</span> <span class="dv">2</span>),
    <span class="dt">color =</span> <span class="st">"white"</span>,
    <span class="dt">size =</span> <span class="dv">3</span>
  ) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set main and axis titles</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"AES-PCA Significant Pathways: Ovarian Cancer"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Negative Log p-Value"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add a line showing the alpha = 0.01 level</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(<span class="fl">0.01</span>), <span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">color =</span> <span class="st">"#005030"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Flip the x and y axes</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/pathway_bar_plots-1.png" width="1008"></p>
</div>
<div id="extract-relevant-genes-from-significant-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-relevant-genes-from-significant-pathways" class="anchor"></a>2.2.4 Extract relevant genes from significant pathways</h3>
<p>Because pathways are defined <em>a priori</em>—typically within each pathway—only a subset of genes are relevant to the phenotype and contribute to pathway significance. In AESPCA, these relevant genes are the genes with nonzero loadings in the first PC of AESPCs.</p>
<p>For example, for the “IL-1 signaling pathway” (Wikipathways <a href="https://www.wikipathways.org/index.php/Pathway:WP195">WP195</a>), we can extract the principal components and their protein loadings (PC &amp; L) using the <code><a href="../reference/getPathPCLs.html">getPathPCLs()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wp195PCLs_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(ovarian_aespcOut, <span class="st">"WP195"</span>)
wp195PCLs_ls
<span class="co">#&gt; $PCs</span>
<span class="co">#&gt; # A tibble: 83 x 2</span>
<span class="co">#&gt;    sampleID         V1</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA.09.1664 -2.66 </span>
<span class="co">#&gt;  2 TCGA.13.1484 -0.459</span>
<span class="co">#&gt;  3 TCGA.13.1488  0.603</span>
<span class="co">#&gt;  4 TCGA.13.1489 -1.70 </span>
<span class="co">#&gt;  5 TCGA.13.1494 -0.726</span>
<span class="co">#&gt;  6 TCGA.13.1495 -0.316</span>
<span class="co">#&gt;  7 TCGA.13.1499  0.252</span>
<span class="co">#&gt;  8 TCGA.23.1123 -0.679</span>
<span class="co">#&gt;  9 TCGA.23.1124  1.83 </span>
<span class="co">#&gt; 10 TCGA.24.1103 -1.93 </span>
<span class="co">#&gt; # ... with 73 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Loadings</span>
<span class="co">#&gt; # A tibble: 30 x 2</span>
<span class="co">#&gt;    featureID   PC1</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ATF2      0    </span>
<span class="co">#&gt;  2 MAPK14    0    </span>
<span class="co">#&gt;  3 AKT1      0    </span>
<span class="co">#&gt;  4 IKBKB     0.527</span>
<span class="co">#&gt;  5 NFKB1     0.483</span>
<span class="co">#&gt;  6 PIK3R1    0.352</span>
<span class="co">#&gt;  7 PLCG1     0    </span>
<span class="co">#&gt;  8 MAPK1     0    </span>
<span class="co">#&gt;  9 MAPK9     0    </span>
<span class="co">#&gt; 10 MAP2K1    0.289</span>
<span class="co">#&gt; # ... with 20 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $pathway</span>
<span class="co">#&gt; [1] "path22"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $term</span>
<span class="co">#&gt; [1] "WP195"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $description</span>
<span class="co">#&gt; [1] "IL-1 signaling pathway"</span></code></pre></div>
<p>From this output list, we can see the data frame of loadings. We will delete proteins with loadings that are equal to 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wp195Loadings_df &lt;-<span class="st"> </span>
<span class="st">  </span>wp195PCLs_ls<span class="op">$</span>Loadings <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(PC1 <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)
wp195Loadings_df
<span class="co">#&gt; # A tibble: 7 x 2</span>
<span class="co">#&gt;   featureID   PC1</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;dbl&gt;</span>
<span class="co">#&gt; 1 IKBKB     0.527</span>
<span class="co">#&gt; 2 NFKB1     0.483</span>
<span class="co">#&gt; 3 PIK3R1    0.352</span>
<span class="co">#&gt; 4 MAP2K1    0.289</span>
<span class="co">#&gt; 5 MAP2K2    0.150</span>
<span class="co">#&gt; 6 TAB1      0.328</span>
<span class="co">#&gt; 7 MYD88     0.389</span></code></pre></div>
<p>We can also graph these loadings in a barchart. We will first add an indicator for inhibition.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wp195Loadings_df &lt;-<span class="st"> </span>
<span class="st">  </span>wp195Loadings_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Sort Loading from Strongest to Weakest</span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(PC1))) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Order the Genes by Loading Strength</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">featureID =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(featureID, <span class="dt">levels =</span> featureID)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Add Directional Indicator (for Colour)</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">Direction =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(PC1 <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">"Up"</span>, <span class="st">"Down"</span>))) </code></pre></div>
<p>Now we will construct a <strong>col</strong>umn chart with <code>ggplot2</code>’s <code>geom_col()</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> wp195Loadings_df) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set overall appearance</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="co"># Define the dependent and independent variables</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> featureID, <span class="dt">y =</span> PC1, <span class="dt">fill =</span> Direction) <span class="op">+</span>
<span class="st">  </span><span class="co"># From the defined variables, create a vertical bar chart</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">width =</span> <span class="fl">0.5</span>, <span class="dt">fill =</span> <span class="st">"#005030"</span>, <span class="dt">color =</span> <span class="st">"#f47321"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set main and axis titles</span>
<span class="st">  </span><span class="kw">labs</span>(
    <span class="dt">title =</span> <span class="st">"Gene Loadings on IL-1 Signaling Pathway"</span>,
    <span class="dt">x =</span> <span class="st">"Protein"</span>
  ) <span class="op">+</span>
<span class="st">  </span><span class="co"># Remove the legend</span>
<span class="st">  </span><span class="kw">guides</span>(<span class="dt">fill =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/graphLoads-1.png" width="624"></p>
<p>Alternatively, we can also plot the correlation of each gene with first PC for each gene. These correlations can be computed by using the <code>TidyCorrelation()</code> function in Section 3.3. of additional vignette “Chapter 5 - Visualizing the Results”.</p>
</div>
<div id="subject-specific-pca-estimates" class="section level3">
<h3 class="hasAnchor">
<a href="#subject-specific-pca-estimates" class="anchor"></a>2.2.5 Subject-specific PCA estimates</h3>
<p>In the study of complex diseases, there is often considerable heterogeneity among different subjects with regard to underlying causes of disease and benefit of particular treatment. Therefore, in addition to identifying disease-relevant pathways for the entire patient group, successful (personalized) treatment regimes will also depend upon knowing if a particular pathway is dysregulated for an individual patient.</p>
<p>To this end, we can also assess subject-specific pathway activity. As we saw earlier, the <code><a href="../reference/getPathPCLs.html">getPathPCLs()</a></code> function also returns subject-specific estimates for the individual pathway PCs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> wp195PCLs_ls<span class="op">$</span>PCs) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set overall appearance</span>
<span class="st">  </span><span class="kw">theme_classic</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Define the independent variable</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> V1) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add the histogram layer</span>
<span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">bins =</span> <span class="dv">10</span>, <span class="dt">color =</span> <span class="st">"#005030"</span>, <span class="dt">fill =</span> <span class="st">"#f47321"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set main and axis titles</span>
<span class="st">  </span><span class="kw">labs</span>(
    <span class="dt">title =</span> <span class="st">"Distribution of Sample-specific Estimate of Pathway Activities"</span>,
    <span class="dt">subtitle =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(wp195PCLs_ls<span class="op">$</span>pathway, <span class="st">": "</span>, wp195PCLs_ls<span class="op">$</span>description),
    <span class="dt">x =</span> <span class="st">"PC1 Value for Each Sample"</span>,
    <span class="dt">y =</span> <span class="st">"Count"</span>
  )</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/wp195PCs-1.png" width="624"></p>
</div>
<div id="extract-analysis-dataset-for-significant-pathways" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-analysis-dataset-for-significant-pathways" class="anchor"></a>2.2.6 Extract analysis dataset for significant pathways</h3>
<p>Users are often also interested in examining the actual data used for analysis of the top pathways, especially for the relevant genes with the pathway. To extract this dataset, use the <code><a href="../reference/SubsetPathwayData.html">SubsetPathwayData()</a></code> function on the <code>Omics</code> data container.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">wp195Data_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SubsetPathwayData.html">SubsetPathwayData</a></span>(ov_OmicsSurv, <span class="st">"WP195"</span>)
wp195Data_df
<span class="co">#&gt; # A tibble: 83 x 33</span>
<span class="co">#&gt;    sampleID EventTime EventObs   ATF2  MAPK14    AKT1   IKBKB   NFKB1</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;lgl&gt;     &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 TCGA.09~      2279 TRUE      0.623 -0.995   0.958  -1.71   -0.730 </span>
<span class="co">#&gt;  2 TCGA.13~      3785 TRUE      1.17   1.71    0.575  -0.751   0.783 </span>
<span class="co">#&gt;  3 TCGA.13~      2154 TRUE     -0.130 -0.127  -0.542  -0.915   0.418 </span>
<span class="co">#&gt;  4 TCGA.13~      2553 TRUE      0.482 -0.0588 -0.292  -0.422  -0.488 </span>
<span class="co">#&gt;  5 TCGA.13~      2856 TRUE      0.130  0.610   1.25   -0.950   0.393 </span>
<span class="co">#&gt;  6 TCGA.13~      2749 TRUE     -0.823  0.266   1.79   -0.522  -0.0498</span>
<span class="co">#&gt;  7 TCGA.13~      3500 FALSE    -0.704  0.516  -0.519  -0.494   0.104 </span>
<span class="co">#&gt;  8 TCGA.23~      1018 TRUE      0.270 -0.419  -3.63    0.0331 -0.114 </span>
<span class="co">#&gt;  9 TCGA.23~      1768 TRUE     -2.61  -1.81   -0.0214  0.811   1.17  </span>
<span class="co">#&gt; 10 TCGA.24~      1646 TRUE      0.270 -0.622   1.23   -0.924  -0.954 </span>
<span class="co">#&gt; # ... with 73 more rows, and 25 more variables: PIK3R1 &lt;dbl&gt;, PLCG1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   MAPK1 &lt;dbl&gt;, MAPK9 &lt;dbl&gt;, MAP2K1 &lt;dbl&gt;, MAP2K2 &lt;dbl&gt;, MAP2K6 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PTPN11 &lt;dbl&gt;, REL &lt;dbl&gt;, RELA &lt;dbl&gt;, MAP3K7 &lt;dbl&gt;, IKBKG &lt;dbl&gt;,</span>
<span class="co">#&gt; #   TAB1 &lt;dbl&gt;, IRAK1 &lt;dbl&gt;, MYD88 &lt;dbl&gt;, IRAK4 &lt;dbl&gt;, ECSIT &lt;dbl&gt;,</span>
<span class="co">#&gt; #   TOLLIP &lt;dbl&gt;, MAPK3 &lt;dbl&gt;, MAP2K3 &lt;dbl&gt;, MAP2K4 &lt;dbl&gt;, TRAF6 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   UBE2N &lt;dbl&gt;, SQSTM1 &lt;dbl&gt;, MAPKAPK2 &lt;dbl&gt;</span></code></pre></div>
<p>We can use this data subset to perform pasthway-specific analysis to estimate the effect of individual genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(survival)
NFKB1_df &lt;-<span class="st"> </span>
<span class="st">  </span>wp195Data_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(EventTime, EventObs, NFKB1)

wp195_mod &lt;-<span class="st"> </span><span class="kw">coxph</span>(
  <span class="kw">Surv</span>(EventTime, EventObs) <span class="op">~</span><span class="st"> </span>NFKB1,
  <span class="dt">data =</span> NFKB1_df
)

<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(wp195_mod)
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; coxph(formula = Surv(EventTime, EventObs) ~ NFKB1, data = NFKB1_df)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   n= 83, number of events= 64 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;         coef exp(coef) se(coef)     z Pr(&gt;|z|)   </span>
<span class="co">#&gt; NFKB1 0.4606    1.5850   0.1521 3.028  0.00246 **</span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;       exp(coef) exp(-coef) lower .95 upper .95</span>
<span class="co">#&gt; NFKB1     1.585     0.6309     1.176     2.136</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Concordance= 0.65  (se = 0.045 )</span>
<span class="co">#&gt; Rsquare= 0.121   (max possible= 0.994 )</span>
<span class="co">#&gt; Likelihood ratio test= 10.7  on 1 df,   p=0.001</span>
<span class="co">#&gt; Wald test            = 9.17  on 1 df,   p=0.002</span>
<span class="co">#&gt; Score (logrank) test = 9.49  on 1 df,   p=0.002</span></code></pre></div>
<p>Additionally, we can estimate Kaplan-Meier survival curves based on up-regulation or down-regulation of NFKB1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add the direction</span>
NFKB1_df &lt;-<span class="st"> </span>
<span class="st">  </span>NFKB1_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">NFKB1_Expr =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(NFKB1 <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/median">median</a></span>(NFKB1), <span class="st">"High"</span>, <span class="st">"Low"</span>))

<span class="co"># Fit the survival model</span>
NFKB1_fit &lt;-<span class="st"> </span><span class="kw">survfit</span>(
  <span class="kw">Surv</span>(EventTime, EventObs) <span class="op">~</span><span class="st"> </span>NFKB1_Expr,
  <span class="dt">data =</span> NFKB1_df
)</code></pre></div>
<p>Finally, we can plot these K-M curves over NFKB1 expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(survminer)

<span class="kw">ggsurvplot</span>(
  NFKB1_fit, <span class="dt">conf.int =</span> <span class="ot">FALSE</span>, 
  <span class="dt">title =</span> <span class="st">"Kaplan-Meier Survival Curves for Ovarian Cancer by NFKB1 Expression"</span>,
  <span class="dt">xlab =</span> <span class="st">"Time in Days"</span>,
  <span class="dt">palette =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"#f47321"</span>, <span class="st">"#005030"</span>)
)</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/NFKB1SurvPlot-1.png" width="700"></p>
<p></p>
<hr>
<p></p>
</div>
</div>
</div>
<div id="case-study-integrative-multi-omics-pathway-analysis-of-ovarian-cancer-data" class="section level1">
<h1 class="hasAnchor">
<a href="#case-study-integrative-multi-omics-pathway-analysis-of-ovarian-cancer-data" class="anchor"></a>3. Case study: integrative multi-omics pathway analysis of ovarian cancer data</h1>
<p>While copy number alterations are common genomic aberrations in ovarian carcer, recent studies have shown these changes do not necessarily lead to concordant changes in protein expression. In section 2.2.4 above, we illustrated testing pathway activities in protein expression against survival outcome. In this section, we will additionally test pathway activities in copy number against survial outcome. Moreover, we will perform integrative analysis to identify those survial-associated protein pathways driven by copy number alterations.</p>
<div id="creating-copy-number-omics-data-object-for-pathway-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-copy-number-omics-data-object-for-pathway-analysis" class="anchor"></a>3.1 Creating copy number <code>Omics</code> data object for pathway analysis</h2>
<p>We can identify copy number (CNV) pathways significantly associated with survival in the same way as we did for protein expressions. This CNV data is from from <strong>DATA SOURCE HERE</strong>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">copyNumberClean_df &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(dataDir_path, <span class="st">"/OV_surv_x_CNV2.RDS"</span>))</code></pre></div>
<p>And now we create an <code>Omics</code> data container.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovCNV_Surv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CreateOmics.html">CreateOmics</a></span>(
  <span class="dt">assayData_df =</span> copyNumberClean_df[, <span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)],
  <span class="dt">pathwayCollection_ls =</span> wikipathways_PC,
  <span class="dt">response =</span> copyNumberClean_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],
  <span class="dt">respType =</span> <span class="st">"survival"</span>,
  <span class="dt">minPathSize =</span> <span class="dv">5</span>
)</code></pre></div>
<pre><code>  ======  Creating object of class OmicsSurv  =======
The input pathway database included 5831 unique features.
The input assay dataset included 24776 features.
Only pathways with at least 5 or more features included in the assay dataset are
  tested (specified by minPathSize parameter). There are 424 pathways which meet
  this criterion.
Because pathwayPCA is a self-contained test (PMID: 17303618), only features in
  both assay data and pathway database are considered for analysis. There are 5637
  such features shared by the input assay and pathway database.</code></pre>
<p>Finally, we can apply the AESPCA method to this copy-number data container. Due to the large sample size, this will take a few moments.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ovCNV_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> ovCNV_Surv,
  <span class="dt">numPCs =</span> <span class="dv">1</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">adjustment =</span> <span class="st">"BH"</span>
)</code></pre></div>
<pre><code>Part 1: Calculate Pathway AES-PCs
Initializing Computing Cluster: DONE
Extracting Pathway PCs in Parallel: DONE

Part 2: Calculate Pathway p-Values
Initializing Computing Cluster: DONE
Extracting Pathway p-Values in Parallel: DONE

Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame
DONE</code></pre>
<!-- The copy number data frame is 109.6 Mb; the copy number `OmicsSurv` is 112.3 Mb; and the copy number `aespcOut` is 3.8 Mb (*and* it takes 3 minutes to compute). We're saving the output. -->
</div>
<div id="identifying-significant-pathways-and-relevant-genes-in-both-cnv-and-protein-level" class="section level2">
<h2 class="hasAnchor">
<a href="#identifying-significant-pathways-and-relevant-genes-in-both-cnv-and-protein-level" class="anchor"></a>3.2 Identifying significant pathways and relevant genes in both CNV and protein level</h2>
<p>Next, we identify the intersection of significant pathways based on both CNV and protein data. First, we will create a data frame of the pathway <span class="math inline">\(p\)</span>-values from both CNV and proteomics.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Copy Number</span>
CNVpvals_df &lt;-<span class="st"> </span>
<span class="st">  </span>ovCNV_aespcOut<span class="op">$</span>pVals_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">rawp_CNV =</span> rawp) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(description, rawp_CNV) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(rawp_CNV <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)

<span class="co"># Proteomics</span>
PROTpvals_df &lt;-<span class="st"> </span>
<span class="st">  </span>ovarian_aespcOut<span class="op">$</span>pVals_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">rawp_PROT =</span> rawp) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(terms, description, rawp_PROT) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(rawp_PROT <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)

<span class="co"># Intersection</span>
SigBoth_df &lt;-<span class="st"> </span><span class="kw">inner_join</span>(PROTpvals_df, CNVpvals_df, <span class="dt">by =</span> <span class="st">"description"</span>)
<span class="co"># WnT Signaling Pathway is listed as WP363 and WP428</span></code></pre></div>
<!-- We can also create a Venn diagram of these two sets. -->
<!-- ```{r VennOverlap, message=FALSE} -->
<!-- library(VennDiagram) -->
<!-- venn.diagram( -->
<!--   list(CopyNum = CNVpvals_df$description, Protein = PROTpvals_df$description), -->
<!--   filename = "Figures/signif_Paths_Venn.png", -->
<!--   imagetype = "png", -->
<!--   col = c("#f47321", "#005030") -->
<!-- ) -->
<!-- ``` -->
<p>The results showed there are 23 pathways significantly associated with survival in both CNV and protein data. Here are the top-10 most significant pathways (sorted by protein data significance):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">SigBoth_df
<span class="co">#&gt; # A tibble: 23 x 4</span>
<span class="co">#&gt;    terms  description                                   rawp_PROT rawp_CNV</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;                                             &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1 WP195  IL-1 signaling pathway                          0.00247 0.0222  </span>
<span class="co">#&gt;  2 WP363  Wnt Signaling Pathway                           0.00485 0.0486  </span>
<span class="co">#&gt;  3 WP2036 TNF related weak inducer of apoptosis (TWEAK~   0.00641 0.0226  </span>
<span class="co">#&gt;  4 WP3850 Factors and pathways affecting insulin-like ~   0.00700 0.0366  </span>
<span class="co">#&gt;  5 WP2203 Thymic Stromal LymphoPoietin (TSLP) Signalin~   0.0156  0.0181  </span>
<span class="co">#&gt;  6 WP4136 Fibrin Complement Receptor 3 Signaling Pathw~   0.0160  0.000358</span>
<span class="co">#&gt;  7 WP384  Apoptosis Modulation by HSP70                   0.0165  0.0311  </span>
<span class="co">#&gt;  8 WP2018 RANKL/RANK (Receptor activator of NFKB (liga~   0.0170  0.000418</span>
<span class="co">#&gt;  9 WP560  TGF-beta Receptor Signaling                     0.0205  0.0355  </span>
<span class="co">#&gt; 10 WP2380 Brain-Derived Neurotrophic Factor (BDNF) sig~   0.0206  0.0113  </span>
<span class="co">#&gt; # ... with 13 more rows</span></code></pre></div>
<p>As discussed in section 2.2.3, we can also identify relevant genes with nonzero loadings that drives pathway significance in both CNV and protein pathway analysis. The “IL-1 signaling pathway” (WP195) makes our cutoff for both CNV and protein data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Copy Number Loadings</span>
CNVwp195_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(ovCNV_aespcOut, <span class="st">"WP195"</span>)
CNV195load_df &lt;-<span class="st"> </span>
<span class="st">  </span>CNVwp195_ls<span class="op">$</span>Loadings <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(PC1) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">PC1_CNV =</span> PC1)

<span class="co"># Protein Loadings</span>
PROTwp195_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(ovarian_aespcOut, <span class="st">"WP195"</span>)
PROT195load_df &lt;-<span class="st"> </span>
<span class="st">  </span>PROTwp195_ls<span class="op">$</span>Loadings <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(PC1) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">rename</span>(<span class="dt">PC1_PROT =</span> PC1)

<span class="co"># Intersection</span>
<span class="kw">inner_join</span>(CNV195load_df, PROT195load_df)
<span class="co">#&gt; Joining, by = "featureID"</span>
<span class="co">#&gt; # A tibble: 1 x 3</span>
<span class="co">#&gt;   featureID PC1_CNV PC1_PROT</span>
<span class="co">#&gt;   &lt;chr&gt;       &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt; 1 NFKB1       0.113    0.483</span></code></pre></div>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/26345753">NFKB1</a> contributes both to CNV and protein pathway significance.</p>
<p>Because the proteomics data were recorded on a subset of the subjects shown in the copy number data, we can further examine the relationship between CNV and protein expressions for this gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">NFKB1data_df &lt;-<span class="st"> </span><span class="kw">inner_join</span>(
  copyNumberClean_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">select</span>(Sample, NFKB1) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">rename</span>(<span class="dt">CNV_NFKB1 =</span> NFKB1),
  ovSurv_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">select</span>(Sample, NFKB1) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">rename</span>(<span class="dt">PROT_NFKB1 =</span> NFKB1)
)
<span class="co">#&gt; Joining, by = "Sample"</span></code></pre></div>
<p>A correlation test shows CNV is highly correlated with protein expression for this gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">NFKB1_cor &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor.test">cor.test</a></span>(NFKB1data_df<span class="op">$</span>CNV_NFKB1, NFKB1data_df<span class="op">$</span>PROT_NFKB1)
NFKB1_cor

    Pearson<span class="st">'s product-moment correlation</span>

<span class="st">data:  NFKB1data_df$CNV_NFKB1 and NFKB1data_df$PROT_NFKB1</span>
<span class="st">t = 4.8873, df = 81, p-value = 5.081e-06</span>
<span class="st">alternative hypothesis: true correlation is not equal to 0</span>
<span class="st">95 percent confidence interval:</span>
<span class="st"> 0.2915318 0.6282385</span>
<span class="st">sample estimates:</span>
<span class="st">      cor </span>
<span class="st">0.4772138 </span></code></pre></div>
<p>We can then visualize the multi-omics relationship for the NFKB1 gene using a scatterplot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(<span class="dt">data =</span> NFKB1data_df) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set overall appearance</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="co"># Define the dependent and independent variables</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> CNV_NFKB1, <span class="dt">y =</span> PROT_NFKB1) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add the scatterplot</span>
<span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add a trendline</span>
<span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> lm, <span class="dt">se =</span> <span class="ot">FALSE</span>, <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set main and axis titles</span>
<span class="st">  </span><span class="kw">labs</span>(
    <span class="dt">title =</span> <span class="st">"NFKB1 Expression by Multi-Omics Data"</span>,
    <span class="dt">x =</span> <span class="st">"Copy Number"</span>,
    <span class="dt">y =</span> <span class="st">"Proteomics"</span>
  ) <span class="op">+</span>
<span class="st">  </span><span class="co"># Include the correlation on the graph</span>
<span class="st">  </span><span class="kw">annotate</span>(
    <span class="dt">geom =</span> <span class="st">"text"</span>, <span class="dt">x =</span> <span class="fl">0.5</span>, <span class="dt">y =</span> <span class="op">-</span><span class="fl">0.6</span>,
    <span class="dt">label =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"rho = "</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(NFKB1_cor<span class="op">$</span>estimate, <span class="dv">3</span>))
  ) <span class="op">+</span>
<span class="st">  </span><span class="kw">annotate</span>(
    <span class="dt">geom =</span> <span class="st">"text"</span>, <span class="dt">x =</span> <span class="fl">0.5</span>, <span class="dt">y =</span> <span class="op">-</span><span class="fl">0.65</span>,
    <span class="dt">label =</span> <span class="st">"p-val &lt; 10^-5"</span>
  )</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/NFKB1plot-1.png" width="624"></p>
</div>
<div id="an-integrative-view-on-patient-specific-pathway-activities" class="section level2">
<h2 class="hasAnchor">
<a href="#an-integrative-view-on-patient-specific-pathway-activities" class="anchor"></a>3.3 An integrative view on patient-specific pathway activities</h2>
<p>In section 2.2.5 above, we have also seen that there can be considerable heterogeneity in pathway activities between the patients. One possible reason could be that copy number changes might not have lead to changes in protein expression for some of the patients. <code>pathwayPCA</code> can be used to estimate pathway activities for each patient, both for protein expressions and copy number expressions separately. These estimates can then be viewed jointly using a circo plot, as illustrated in section xx of the additional vignette “Chapter 5 - Visualizing the Results”.</p>
<p><strong>paste a picture here</strong></p>
<p></p>
<hr>
<p></p>
</div>
</div>
<div id="case-study-analysis-of-studies-with-complex-designs" class="section level1">
<h1 class="hasAnchor">
<a href="#case-study-analysis-of-studies-with-complex-designs" class="anchor"></a>4. Case study: analysis of studies with complex designs</h1>
<p><code>pathwayPCA</code> is capable of analyzing studies with multiple experimental factors. In this section, we illustrate using <code>pathwayPCA</code> to test differential association of pathway expression with survial outcome in male and female subjects.</p>
<div id="data-setup-and-aespca-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#data-setup-and-aespca-analysis" class="anchor"></a>4.1 Data setup and AESPCA analysis</h2>
<p>For this example, we use TCGA KIRP RNAseq dataset downloaded from Xena datahub. First, we load the KIRP data, create an <code>Omics</code> data container, and extract first AESPC from each pathway-specific data subset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load Data</span>
kidney_df &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(dataDir_path, <span class="st">"/KIRP_Surv_RNAseq_inner.RDS"</span>))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Create Omics Container</span>
kidney_Omics &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CreateOmics.html">CreateOmics</a></span>(
  <span class="dt">assayData_df =</span> kidney_df[, <span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">4</span>)],
  <span class="dt">pathwayCollection_ls =</span> wikipathways_PC,
  <span class="dt">response =</span> kidney_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],
  <span class="dt">respType =</span> <span class="st">"surv"</span>,
  <span class="dt">minPathSize =</span> <span class="dv">5</span>
)</code></pre></div>
<pre><code>  ======  Creating object of class OmicsSurv  =======
The input pathway database included 5831 unique features.
The input assay dataset included 20211 features.
Only pathways with at least 5 or more features included in the assay dataset are
  tested (specified by minPathSize parameter). There are 423 pathways which meet
  this criterion.
Because pathwayPCA is a self-contained test (PMID: 17303618), only features in
  both assay data and pathway database are considered for analysis. There are 5566
  such features shared by the input assay and pathway database.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># AESPCA</span>
kidney_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> kidney_Omics,
  <span class="dt">numPCs =</span> <span class="dv">1</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">adjustment =</span> <span class="st">"BH"</span>
)</code></pre></div>
<pre><code>Part 1: Calculate Pathway AES-PCs
Initializing Computing Cluster: DONE
Extracting Pathway PCs in Parallel: DONE

Part 2: Calculate Pathway p-Values
Initializing Computing Cluster: DONE
Extracting Pathway p-Values in Parallel: DONE

Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame
DONE</code></pre>
<!-- This takes 1 minute and 40 seconds.  -->
</div>
<div id="test-for-sex-interaction-with-first-pc" class="section level2">
<h2 class="hasAnchor">
<a href="#test-for-sex-interaction-with-first-pc" class="anchor"></a>4.2 Test for sex interaction with first PC</h2>
<p>Next we can test whether there is differential pathway association with survival outcome for males and females by the following model, <span class="math display">\[
h(t) = h_0(t) \exp\left[ \beta_0 + \beta_1\text{PC}_1 + \beta_2\text{male} + \beta_3(\text{PC}_1 \times \text{male}) \right].
\]</span></p>
<p>In order to test the sex interaction effect for all pathways, we will write a function which tests the interaction for a single pathway-subset of the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">TestIntxn &lt;-<span class="st"> </span><span class="cf">function</span>(pathway, pcaOut, resp_df){
  
  <span class="co"># For the given pathway, extract the PCs and loadings from the pcaOut list</span>
  PCL_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(pcaOut, pathway)
  <span class="co"># Select and rename the PC</span>
  PC_df &lt;-<span class="st"> </span>PCL_ls<span class="op">$</span>PCs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="dt">PC1 =</span> V1)
  <span class="co"># Bind this PC to the phenotype data</span>
  data_df &lt;-<span class="st"> </span><span class="kw">bind_cols</span>(resp_df, PC_df)
  
  <span class="co"># Construct a survival model with sex interaction</span>
  sex_mod &lt;-<span class="st"> </span><span class="kw">coxph</span>(
    <span class="kw">Surv</span>(time, status) <span class="op">~</span><span class="st"> </span>PC1 <span class="op">+</span><span class="st"> </span>male <span class="op">+</span><span class="st"> </span>PC1 <span class="op">*</span><span class="st"> </span>male, <span class="dt">data =</span> data_df
  )
  <span class="co"># Extract the model fit statistics for the interaction</span>
  modStats_mat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(
    <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(sex_mod))[<span class="st">"PC1:maleTRUE"</span>, ]
  )
 
  <span class="co"># Return a data frame of the pathway and model statistics</span>
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(
    <span class="dt">statistics =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(
      <span class="dt">terms =</span> pathway,
      <span class="dt">description =</span> PCL_ls<span class="op">$</span>description,
      modStats_mat,
      <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>
    ),
    <span class="dt">model =</span> sex_mod
  )

}</code></pre></div>
<p>To make sure this function gives us the information we need, we can test it on one pathway.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">TestIntxn</span>(<span class="st">"WP195"</span>, kidney_aespcOut, kidney_df[, <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>])<span class="op">$</span>model
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; coxph(formula = Surv(time, status) ~ PC1 + male + PC1 * male, </span>
<span class="co">#&gt;     data = data_df)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                  coef exp(coef) se(coef)      z     p</span>
<span class="co">#&gt; PC1           0.03278   1.03332  0.08206  0.399 0.690</span>
<span class="co">#&gt; maleTRUE     -0.38822   0.67826  0.32827 -1.183 0.237</span>
<span class="co">#&gt; PC1:maleTRUE  0.05649   1.05812  0.10438  0.541 0.588</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Likelihood ratio test=3.97  on 3 df, p=0.2652</span>
<span class="co">#&gt; n= 320, number of events= 51</span></code></pre></div>
<p>Now we can apply this function across the list of pathways, and wrangle the results into a sorted data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">paths_char &lt;-<span class="st"> </span>kidney_aespcOut<span class="op">$</span>pVals_df<span class="op">$</span>terms
interactions_ls &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(
    paths_char,
    <span class="dt">FUN =</span> TestIntxn,
    <span class="dt">pcaOut =</span> kidney_aespcOut,
    <span class="dt">resp_df =</span> kidney_df[, <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>]
)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(interactions_ls) &lt;-<span class="st"> </span>paths_char

interactions_df &lt;-<span class="st"> </span>
<span class="st">  </span>interactions_ls <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">bind_rows</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">as.tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="st">`</span><span class="dt">Pr...z..</span><span class="st">`</span>)

interactions_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(<span class="st">`</span><span class="dt">Pr...z..</span><span class="st">`</span> <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)
<span class="co">#&gt; # A tibble: 14 x 7</span>
<span class="co">#&gt;    terms  description               coef exp.coef. se.coef.     z Pr...z..</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;chr&gt;                    &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;</span>
<span class="co">#&gt;  1 WP1559 TFs Regulate miRNAs re~ -0.597     0.550    0.216 -2.76  0.00573</span>
<span class="co">#&gt;  2 WP453  Inflammatory Response ~ -0.272     0.762    0.109 -2.50  0.0126 </span>
<span class="co">#&gt;  3 WP3893 Development and hetero~ -0.329     0.719    0.137 -2.40  0.0166 </span>
<span class="co">#&gt;  4 WP3929 Chemokine signaling pa~ -0.252     0.777    0.109 -2.32  0.0206 </span>
<span class="co">#&gt;  5 WP2849 Hematopoietic Stem Cel~ -0.278     0.757    0.129 -2.16  0.0305 </span>
<span class="co">#&gt;  6 WP1423 Ganglio Sphingolipid M~ -0.487     0.615    0.230 -2.11  0.0346 </span>
<span class="co">#&gt;  7 WP3892 Development of pulmona~ -0.316     0.729    0.151 -2.10  0.0360 </span>
<span class="co">#&gt;  8 WP3678 Amplification and Expa~ -0.357     0.700    0.171 -2.09  0.0363 </span>
<span class="co">#&gt;  9 WP4141 PI3K/AKT/mTOR - VitD3 ~  0.314     1.37     0.153  2.05  0.0407 </span>
<span class="co">#&gt; 10 WP3941 Oxidative Damage        -0.354     0.702    0.176 -2.01  0.0442 </span>
<span class="co">#&gt; 11 WP3863 T-Cell antigen Recepto~ -0.233     0.792    0.117 -1.99  0.0462 </span>
<span class="co">#&gt; 12 WP3872 Regulation of Apoptosi~  0.267     1.31     0.135  1.98  0.0472 </span>
<span class="co">#&gt; 13 WP3672 LncRNA-mediated mechan~ -0.392     0.675    0.198 -1.98  0.0475 </span>
<span class="co">#&gt; 14 WP3967 miR-509-3p alteration ~ -0.299     0.741    0.151 -1.98  0.0481</span></code></pre></div>
<p>The results showed the most significant pathway is <a href="https://www.wikipathways.org/index.php/Pathway:WP1559">WP1559</a> (“TFs Regulate miRNAs related to cardiac hypertrophy”). We can inspect the model results for this pathway directly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(interactions_ls[[<span class="st">"WP1559"</span>]]<span class="op">$</span>model)
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; coxph(formula = Surv(time, status) ~ PC1 + male + PC1 * male, </span>
<span class="co">#&gt;     data = data_df)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   n= 320, number of events= 51 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                 coef exp(coef) se(coef)      z Pr(&gt;|z|)    </span>
<span class="co">#&gt; PC1           0.7534    2.1242   0.1782  4.228 2.36e-05 ***</span>
<span class="co">#&gt; maleTRUE      0.1825    1.2002   0.4268  0.427  0.66903    </span>
<span class="co">#&gt; PC1:maleTRUE -0.5974    0.5502   0.2162 -2.763  0.00573 ** </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;              exp(coef) exp(-coef) lower .95 upper .95</span>
<span class="co">#&gt; PC1             2.1242     0.4708    1.4980    3.0122</span>
<span class="co">#&gt; maleTRUE        1.2002     0.8332    0.5199    2.7705</span>
<span class="co">#&gt; PC1:maleTRUE    0.5502     1.8174    0.3602    0.8406</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Concordance= 0.654  (se = 0.047 )</span>
<span class="co">#&gt; Rsquare= 0.068   (max possible= 0.792 )</span>
<span class="co">#&gt; Likelihood ratio test= 22.51  on 3 df,   p=5e-05</span>
<span class="co">#&gt; Wald test            = 30.03  on 3 df,   p=1e-06</span>
<span class="co">#&gt; Score (logrank) test = 33.09  on 3 df,   p=3e-07</span></code></pre></div>
<p>The results showed that although sex is not significantly associated with survial outcome, the association of pathway gene expression (PC1) with survival is highly dependent on sex of the samples.</p>
</div>
<div id="survival-curves-by-sex-interaction" class="section level2">
<h2 class="hasAnchor">
<a href="#survival-curves-by-sex-interaction" class="anchor"></a>4.3 Survival curves by sex interaction</h2>
<p>We can also divide subjects according to sex and high or low PC-expression, and add a group indicator for the four groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Bind the Pheno Data to WP1559's First PC</span>
kidneyWP1559_df &lt;-<span class="st"> </span><span class="kw">bind_cols</span>(
  kidney_df[, <span class="dv">2</span><span class="op">:</span><span class="dv">4</span>],
  <span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(kidney_aespcOut, <span class="st">"WP1559"</span>)<span class="op">$</span>PCs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="dt">PC1 =</span> V1)
)

<span class="co"># Add Grouping Feature</span>
kidneySurvWP1559grouped_df &lt;-<span class="st"> </span>
<span class="st">  </span>kidneyWP1559_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># add strength indicator</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">direction =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(PC1 <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/median">median</a></span>(PC1), <span class="st">"High"</span>, <span class="st">"Low"</span>)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># group by interaction of sex and strength on PC</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">group =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(direction, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(male, <span class="st">"_Male"</span>, <span class="st">"_Female"</span>))) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># remove summarized columns</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>male, <span class="op">-</span>PC1, <span class="op">-</span>direction)</code></pre></div>
<p>Now we can plot survial curves for the four groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Fit the survival model</span>
fit &lt;-<span class="st"> </span><span class="kw">survfit</span>(
  <span class="kw">Surv</span>(time, status) <span class="op">~</span><span class="st"> </span>group,
  <span class="dt">data =</span> kidneySurvWP1559grouped_df
)

<span class="kw">ggsurvplot</span>(
  fit, <span class="dt">conf.int =</span> <span class="ot">FALSE</span>, 
  <span class="dt">title =</span> <span class="st">"Kaplan-Meier Curves of Kidney Cancer Survival by Sex * PC Interaction"</span>,
  <span class="dt">xlab =</span> <span class="st">"Time in Days"</span>,
  <span class="dt">ylim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.4</span>, <span class="dv">1</span>)
)</code></pre></div>
<p><img src="Introduction_to_pathwayPCA_files/figure-html/plotKidneyGroupedSurv-1.png" width="1008"></p>
<p>As shown by these Kaplan-Meier curves, for the “TFs Regulate miRNAs related to cardiac hypertrophy” pathway, first-PC expression above the median was more strongly associated with shorter survival times than those below the median. Moreover, this effect appears stronger for females than males.</p>
<p></p>
<hr>
<p></p>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>5. Conclusion</h1>
<p>Now that you have an idea of what questions can be answered with this package, please see each of our supplementary vignette chapters for detailed and thorough commentary and guiding information on each of the three topics discussed herein. The vignettes are:</p>
<ul>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html">Chapter 2: Import Data</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html">Chapter 3: Create <code>Omics</code> Data Objects</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html">Chapter 4: Test Pathway Significance</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html">Chapter 5: Visualizing the Results</a></li>
</ul>
<p>Here is the R session information for this vignette:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.5.1 (2018-07-02)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 7 x64 (build 7601) Service Pack 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_United States.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                          </span>
<span class="co">#&gt; [5] LC_TIME=English_United States.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] survminer_0.4.3       ggpubr_0.2            magrittr_1.5         </span>
<span class="co">#&gt;  [4] survival_2.43-3       bindrcpp_0.2.2        pathwayPCA_0.0.0.9000</span>
<span class="co">#&gt;  [7] forcats_0.3.0         stringr_1.3.1         dplyr_0.7.8          </span>
<span class="co">#&gt; [10] purrr_0.2.5           readr_1.3.0           tidyr_0.8.2          </span>
<span class="co">#&gt; [13] tibble_1.4.2          ggplot2_3.1.0         tidyverse_1.2.1      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.0        lubridate_1.7.4   lattice_0.20-38  </span>
<span class="co">#&gt;  [4] zoo_1.8-4         assertthat_0.2.0  rprojroot_1.3-2  </span>
<span class="co">#&gt;  [7] digest_0.6.18     utf8_1.1.4        R6_2.3.0         </span>
<span class="co">#&gt; [10] cellranger_1.1.0  plyr_1.8.4        backports_1.1.2  </span>
<span class="co">#&gt; [13] lars_1.2          evaluate_0.12     httr_1.4.0       </span>
<span class="co">#&gt; [16] pillar_1.3.0      rlang_0.3.0.1     lazyeval_0.2.1   </span>
<span class="co">#&gt; [19] readxl_1.1.0      data.table_1.11.8 rstudioapi_0.8   </span>
<span class="co">#&gt; [22] Matrix_1.2-15     rmarkdown_1.11    pkgdown_1.3.0    </span>
<span class="co">#&gt; [25] desc_1.2.0        labeling_0.3      splines_3.5.1    </span>
<span class="co">#&gt; [28] munsell_0.5.0     broom_0.5.1       compiler_3.5.1   </span>
<span class="co">#&gt; [31] modelr_0.1.2      xfun_0.4          pkgconfig_2.0.2  </span>
<span class="co">#&gt; [34] htmltools_0.3.6   tidyselect_0.2.5  km.ci_0.5-2      </span>
<span class="co">#&gt; [37] gridExtra_2.3     roxygen2_6.1.1    fansi_0.4.0      </span>
<span class="co">#&gt; [40] crayon_1.3.4      withr_2.1.2       MASS_7.3-51.1    </span>
<span class="co">#&gt; [43] commonmark_1.7    grid_3.5.1        xtable_1.8-3     </span>
<span class="co">#&gt; [46] nlme_3.1-137      jsonlite_1.6      gtable_0.2.0     </span>
<span class="co">#&gt; [49] KMsurv_0.1-5      scales_1.0.0      cli_1.0.1        </span>
<span class="co">#&gt; [52] stringi_1.2.4     fs_1.2.6          xml2_1.2.0       </span>
<span class="co">#&gt; [55] survMisc_0.5.5    generics_0.0.2    cmprsk_2.2-7     </span>
<span class="co">#&gt; [58] tools_3.5.1       glue_1.3.0        hms_0.4.2        </span>
<span class="co">#&gt; [61] parallel_3.5.1    yaml_2.2.0        colorspace_1.3-2 </span>
<span class="co">#&gt; [64] rvest_0.3.2       memoise_1.1.0     knitr_1.21       </span>
<span class="co">#&gt; [67] bindr_0.1.1       haven_2.0.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">1. Introduction</a></li>
      <li>
<a href="#case-study-identifying-significant-pathways-in-protein-expressions-associated-with-survival-outcome-in-ovarian-cancer-data">2. Case study: identifying significant pathways in protein expressions associated with survival outcome in ovarian cancer data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#overian-cancer-dataset">2.1. Overian cancer dataset</a></li>
      <li><a href="#creating-an-omics-data-object-for-pathway-analysis">2.2. Creating an <code>Omics</code> data object for pathway analysis</a></li>
      <li><a href="#testing-pathway-association-with-phenotypes">2.3. Testing pathway association with phenotypes</a></li>
      </ul>
</li>
      <li>
<a href="#case-study-integrative-multi-omics-pathway-analysis-of-ovarian-cancer-data">3. Case study: integrative multi-omics pathway analysis of ovarian cancer data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#creating-copy-number-omics-data-object-for-pathway-analysis">3.1 Creating copy number <code>Omics</code> data object for pathway analysis</a></li>
      <li><a href="#identifying-significant-pathways-and-relevant-genes-in-both-cnv-and-protein-level">3.2 Identifying significant pathways and relevant genes in both CNV and protein level</a></li>
      <li><a href="#an-integrative-view-on-patient-specific-pathway-activities">3.3 An integrative view on patient-specific pathway activities</a></li>
      </ul>
</li>
      <li>
<a href="#case-study-analysis-of-studies-with-complex-designs">4. Case study: analysis of studies with complex designs</a><ul class="nav nav-pills nav-stacked">
<li><a href="#data-setup-and-aespca-analysis">4.1 Data setup and AESPCA analysis</a></li>
      <li><a href="#test-for-sex-interaction-with-first-pc">4.2 Test for sex interaction with first PC</a></li>
      <li><a href="#survival-curves-by-sex-interaction">4.3 Survival curves by sex interaction</a></li>
      </ul>
</li>
      <li><a href="#conclusion">5. Conclusion</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d">Lily Wang</a>, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
