<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Suppl. Ch. 1 - Quickstart Guide for New R Users • pathwayPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Suppl. Ch. 1 - Quickstart Guide for New R Users">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathwayPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    User Guides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction_to_pathwayPCA.html">Integrative Pathway Analysis with pathwayPCA</a>
    </li>
    <li>
      <a href="../articles/Supplement1-Quickstart_Guide.html">Suppl. Ch. 1 - Quickstart Guide for New R Users</a>
    </li>
    <li>
      <a href="../articles/Supplement2-Importing_Data.html">Suppl. Ch. 2 - Import and Tidy Data</a>
    </li>
    <li>
      <a href="../articles/Supplement3-Create_Omics_Objects.html">Suppl. Ch. 3 - Creating Data Objects</a>
    </li>
    <li>
      <a href="../articles/Supplement4-Methods_Walkthrough.html">Suppl. Ch. 4 - Test Pathway Significance</a>
    </li>
    <li>
      <a href="../articles/Supplement5-Analyse_Results.html">Suppl. Ch.  5 - Visualizing the Results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Version Logs</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Suppl. Ch. 1 - Quickstart Guide for New R Users</h1>
                        <h4 class="author">Gabriel Odom</h4>
            
            <h4 class="date">2018-12-17</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gabrielodom/pathwayPCA/blob/master/vignettes/Supplement1-Quickstart_Guide.Rmd"><code>vignettes/Supplement1-Quickstart_Guide.Rmd</code></a></small>
      <div class="hidden name"><code>Supplement1-Quickstart_Guide.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>1. Overview</h1>
<p>This guide will serve as a brief overview to the pathway significance testing workflow with the <code>pathwayPCA</code> package. We will discuss the four basic steps of pathway significance testing with the <code>pathwayPCA</code> package. These steps are: importing data, creating an <code>Omics</code> data object, testing pathways for significance, and inspecting the results. For detailed discussions of these steps, see the following appropriate vignettes:</p>
<ol style="list-style-type: decimal">
<li>Download Packages</li>
<li>Import and Tidy Data (<a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html"><em>vignette</em></a>)</li>
<li>Create Data Objects (<a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html"><em>vignette</em></a>)</li>
<li>Test Pathway Significance (<a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html"><em>vignette</em></a>)</li>
<li>Visualize Results (<a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html"><em>vignette</em></a>)</li>
</ol>
<p>Before we get started, you need the <code>pathwayPCA</code> package to run your analysis. Because we are currently in the development phase for this package, you can install the package from GitHub. In order to install a package from GitHub, you will need the <code>devtools::</code> package (<a href="https://github.com/r-lib/devtools" class="uri">https://github.com/r-lib/devtools</a>) and either <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (for Windows) or <a href="https://developer.apple.com/xcode/">Xcode</a> (for Mac). Then you can install the development version of the <a href="https://github.com/gabrielodom/pathwayPCA"><code>pathwayPCA</code> package</a> from <a href="https://github.com/">GitHub</a>:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">devtools::install_github("gabrielodom/pathwayPCA")</a></code></pre>
<p>Also, if you want your analysis to be performed with parallel computing, you will need a package to help you. We recommend the <code>parallel</code> package (it comes with <code>R</code> automatically). We also recommend the <code>tidyverse</code> package to help you run some of the examples in these vignettes (while the <code>tidyverse</code> package suite is required for many of the examples in the vignettes, it is not required for any of the functions in this package).</p>
<pre><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("tidyverse")</a></code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(parallel)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(pathwayPCA)</code></pre></div>
<hr>
<p></p>
</div>
<div id="import-data" class="section level1">
<h1 class="hasAnchor">
<a href="#import-data" class="anchor"></a>2. Import Data</h1>
<p>This section is a quick overview of the material covered in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html">Import and Tidy Data</a> vignette. Here we show how to import pathway information, assay and phenotype data, and how to join the assay and phenotype data into one data frame.</p>
<div id="import--gmt-files" class="section level2">
<h2 class="hasAnchor">
<a href="#import--gmt-files" class="anchor"></a>2.1 Import <code>.gmt</code> Files</h2>
<p>The <code>.gmt</code> format is a commonly used file format for storing <a href="https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29">pathway information</a>. Lists of pathways in the Molecular Signatures Database (MSigDB) can be downloaded from the <a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp">MSigDB Collections page</a>, and you can use the <code>read_gmt</code> function to import such a <code>.gmt</code> file into R. All <code>.gmt</code> files have a “description” field, which contains additional information on the pathway. However, this field can be left empty. In this example, we use <code>description = FALSE</code> to skip importing the “description” field.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gmt_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"c2.cp.v6.0.symbols.gmt"</span>,
                         <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
cp_pathwayCollection &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_gmt.html">read_gmt</a></span>(gmt_path, <span class="dt">description =</span> <span class="ot">FALSE</span>)
cp_pathwayCollection
<span class="co">#&gt; Object with Class(es) 'pathwayCollection', 'list' [package 'pathwayPCA'] with 2 elements: </span>
<span class="co">#&gt;  $ pathways:List of 1329</span>
<span class="co">#&gt;  $ TERMS   : chr [1:1329] "KEGG_GLYCOLYSIS_GLUCONEOGENESIS" ...</span></code></pre></div>
<p>The imported <code>.gmt</code> file is stored as a <code>pathwayCollection</code> list object. This list contains:</p>
<ul>
<li>a list of the gene symbols or gene IDs contained in each pathway (<code>pathways</code>),</li>
<li>the names of the pathways (<code>TERMS</code>), and</li>
<li>(<em>OPTIONAL</em>) the description of each pathway (<code>description</code>). For Canonical Pathways files, if you specify <code>description = TRUE</code>, this is the hyperlink to the pathway description card on the GSEA website.</li>
</ul>
</div>
<div id="import-and-tidy-assay-data" class="section level2">
<h2 class="hasAnchor">
<a href="#import-and-tidy-assay-data" class="anchor"></a>2.2 Import and Tidy Assay Data</h2>
<p>We assume that the assay data (e.g. transcriptomic data) is either in an Excel file or flat text file. For example, your data may look like this:</p>
<div class="figure">
<img src="example_assay_data.PNG">
</div>
<p>In this example data set, the columns are individual samples. The values in each row are the gene expression measurements for that gene. Use the <code>read_csv</code> function from the <code>readr</code> package to import <code>.csv</code> files into <code>R</code> as <a href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">tibble</a> (table <em>and</em> data frame) objects. The <code>read_csv</code> function prints messages to the screen informing you of the assumptions it makes when importing your data. Specifically, this message tells us that all the imported data is numeric (<code>col_double()</code>) except for the gene name column (<code>X1 = col_character()</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assay_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_assay_subset.csv"</span>,
                          <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
assay_df &lt;-<span class="st"> </span><span class="kw">read_csv</span>(assay_path)
<span class="co">#&gt; Warning: Missing column names filled in: 'X1' [1]</span>
<span class="co">#&gt; Parsed with column specification:</span>
<span class="co">#&gt; cols(</span>
<span class="co">#&gt;   .default = col_double(),</span>
<span class="co">#&gt;   X1 = col_character()</span>
<span class="co">#&gt; )</span>
<span class="co">#&gt; See spec(...) for full column specifications.</span></code></pre></div>
<p>Incidentally, we consider gene symbols to adhere to the following conditions:</p>
<ul>
<li>gene symbols must start with an English letter (a-z or A-Z), and</li>
<li>gene symbols can only contain English letters, Arabic numerals (0-9), and possibly a dash (-).</li>
</ul>
<p>Furthermore, if your data has samples in the columns and -omic feature measurements in the rows (like the data set above), you’ll need to “tidy” the imported assay with the <code>TransposeAssay</code> function. The transposed data set will appear similar to the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assayT_df &lt;-<span class="st"> </span><span class="kw"><a href="../reference/TransposeAssay.html">TransposeAssay</a></span>(assay_df)
assayT_df
<span class="co">#&gt; # A tibble: 36 x 18</span>
<span class="co">#&gt;    Sample SOAT1   LSS  SQLE   EBP CYP51A1 DHCR7 CYP27B1 DHCR24 HSD17B7</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 T2110…  5.37  9.77  7.74  4.68    8.27  8.32    6.78   4.70    7.63</span>
<span class="co">#&gt;  2 T2110…  5.52  9.78  8.06  5.12    8.21  8.33    6.47   5.06    7.63</span>
<span class="co">#&gt;  3 T2110…  5.89  8.11  7.00  5.78    8.20  8.39    6.57   4.89    8.15</span>
<span class="co">#&gt;  4 T2110…  5.62  8.67  8.59  5.64    8.07  8.64    6.47   4.98    8.13</span>
<span class="co">#&gt;  5 T2110…  5.49  9.83  8.13  5.73    9.38  8.15    6.43   5.03    7.70</span>
<span class="co">#&gt;  6 T2110…  5.58  9.85  8.55  5.13    9.40  8.71    6.56   4.87    7.75</span>
<span class="co">#&gt;  7 T2110…  5.32 10.0   6.99  5.86    8.08  9.25    6.86   4.72    7.88</span>
<span class="co">#&gt;  8 T2110…  5.49  9.72  7.47  5.16    6.67  7.37    6.70   4.92    7.50</span>
<span class="co">#&gt;  9 T2110…  5.57  9.88  7.97  5.40    7.91  8.06    6.58   5.06    8.16</span>
<span class="co">#&gt; 10 T2110…  5.16  9.87  7.42  5.50    7.43  8.68    6.55   4.85    8.20</span>
<span class="co">#&gt; # ... with 26 more rows, and 8 more variables: MSMO1 &lt;dbl&gt;, FDFT1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SC5DL &lt;dbl&gt;, LIPA &lt;dbl&gt;, CEL &lt;dbl&gt;, TM7SF2 &lt;dbl&gt;, NSDHL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SOAT2 &lt;dbl&gt;</span></code></pre></div>
</div>
<div id="import-phenotype-info" class="section level2">
<h2 class="hasAnchor">
<a href="#import-phenotype-info" class="anchor"></a>2.3 Import Phenotype Info</h2>
<p>Use the <code>read_csv</code> function to import the phenotype data. Once again, the <code>read_csv</code> function displays a message informing us of the types of data in each column. The following phenotype dataset for subject survival information contains the subject ID (<code>Sample</code>), survival time after disease onset in months (<code>eventTime</code>), and a logical (or binary) variable indicating if the subject died (<code>TRUE</code> or 1) or was lost to follow up (<code>eventObserved</code>; 0 or <code>FALSE</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pInfo_path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"ex_pInfo_subset.csv"</span>,
                          <span class="dt">package =</span> <span class="st">"pathwayPCA"</span>, <span class="dt">mustWork =</span> <span class="ot">TRUE</span>)
pInfo_df &lt;-<span class="st"> </span><span class="kw">read_csv</span>(pInfo_path)
<span class="co">#&gt; Parsed with column specification:</span>
<span class="co">#&gt; cols(</span>
<span class="co">#&gt;   Sample = col_character(),</span>
<span class="co">#&gt;   eventTime = col_double(),</span>
<span class="co">#&gt;   eventObserved = col_logical()</span>
<span class="co">#&gt; )</span>
pInfo_df
<span class="co">#&gt; # A tibble: 36 x 3</span>
<span class="co">#&gt;    Sample    eventTime eventObserved</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;dbl&gt; &lt;lgl&gt;        </span>
<span class="co">#&gt;  1 T21101311     14.2  TRUE         </span>
<span class="co">#&gt;  2 T21101312      1    TRUE         </span>
<span class="co">#&gt;  3 T21101313      6.75 FALSE        </span>
<span class="co">#&gt;  4 T21101314      8.5  TRUE         </span>
<span class="co">#&gt;  5 T21101315      7.25 FALSE        </span>
<span class="co">#&gt;  6 T21101316      5    TRUE         </span>
<span class="co">#&gt;  7 T21101317     20    TRUE         </span>
<span class="co">#&gt;  8 T21101318     13.2  FALSE        </span>
<span class="co">#&gt;  9 T21101319      7.75 FALSE        </span>
<span class="co">#&gt; 10 T21101320      9    FALSE        </span>
<span class="co">#&gt; # ... with 26 more rows</span></code></pre></div>
</div>
<div id="match-the-phenotype-and-assay-data" class="section level2">
<h2 class="hasAnchor">
<a href="#match-the-phenotype-and-assay-data" class="anchor"></a>2.4 Match the Phenotype and Assay Data</h2>
<p>Now that you have the assay data in tidy form (<code>assayT_df</code>) and the phenotype data (<code>pInfo_df</code>), you can use the <code>inner_join</code> function from the <code>dplyr</code> package to match the assay measurements to phenotype information by subject identifier.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exSurv_df &lt;-<span class="st"> </span><span class="kw">inner_join</span>(pInfo_df, assayT_df, <span class="dt">by =</span> <span class="st">"Sample"</span>)
exSurv_df
<span class="co">#&gt; # A tibble: 36 x 20</span>
<span class="co">#&gt;    Sample eventTime eventObserved SOAT1   LSS  SQLE   EBP CYP51A1 DHCR7</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;lgl&gt;         &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 T2110…     14.2  TRUE           5.37  9.77  7.74  4.68    8.27  8.32</span>
<span class="co">#&gt;  2 T2110…      1    TRUE           5.52  9.78  8.06  5.12    8.21  8.33</span>
<span class="co">#&gt;  3 T2110…      6.75 FALSE          5.89  8.11  7.00  5.78    8.20  8.39</span>
<span class="co">#&gt;  4 T2110…      8.5  TRUE           5.62  8.67  8.59  5.64    8.07  8.64</span>
<span class="co">#&gt;  5 T2110…      7.25 FALSE          5.49  9.83  8.13  5.73    9.38  8.15</span>
<span class="co">#&gt;  6 T2110…      5    TRUE           5.58  9.85  8.55  5.13    9.40  8.71</span>
<span class="co">#&gt;  7 T2110…     20    TRUE           5.32 10.0   6.99  5.86    8.08  9.25</span>
<span class="co">#&gt;  8 T2110…     13.2  FALSE          5.49  9.72  7.47  5.16    6.67  7.37</span>
<span class="co">#&gt;  9 T2110…      7.75 FALSE          5.57  9.88  7.97  5.40    7.91  8.06</span>
<span class="co">#&gt; 10 T2110…      9    FALSE          5.16  9.87  7.42  5.50    7.43  8.68</span>
<span class="co">#&gt; # ... with 26 more rows, and 11 more variables: CYP27B1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   DHCR24 &lt;dbl&gt;, HSD17B7 &lt;dbl&gt;, MSMO1 &lt;dbl&gt;, FDFT1 &lt;dbl&gt;, SC5DL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   LIPA &lt;dbl&gt;, CEL &lt;dbl&gt;, TM7SF2 &lt;dbl&gt;, NSDHL &lt;dbl&gt;, SOAT2 &lt;dbl&gt;</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="create-an-omics-data-object" class="section level1">
<h1 class="hasAnchor">
<a href="#create-an-omics-data-object" class="anchor"></a>3. Create an <code>Omics</code> Data Object</h1>
<p>This section is a quick overview of the material covered in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html">Creating Data Objects</a> vignette.</p>
<div id="create-an-object" class="section level2">
<h2 class="hasAnchor">
<a href="#create-an-object" class="anchor"></a>3.1 Create an Object</h2>
<p>Using the data you just imported, create a data object specific to survival, regression, or categorical responses. For our example dataset, we will create a survival <code>Omics</code> object to hold our assay, pathways, and survival responses. If your indicator is a binary variable, the <code>CreateOmics</code> function will attempt to coerce it to a logical variable. Therefore, death indicators should be coded as 0-1, not 1-2.</p>
<p>This package includes a subject-matched colon cancer survival assay subset (<code>colonSurv_df</code>) and a toy pathway collection with 15 pathways (<code>colon_pathwayCollection</code>). When we create this <code>OmicsSurv</code> object, the <code>pathwayPCA</code> package checks the overlap between the -omes recorded in the assay data and the gene symbols in the supplied pathway collection. The <code><a href="../reference/CreateOmics.html">CreateOmics()</a></code> function also prints some diagnostic messages to inform you of how well your pathway collection overlaps your data.</p>
<!-- There are three classes of `Omics` objects to match the three types of response, but only one function to learn: the `CreateOmics` function. Each class contains assay information (in tidy format) and `pathwayCollection` (in list format), but the classes differ in the type of response information they can hold. The classes, and their responses, are -->
<!-- 1. `OmicsSurv`---a data object for survival information, which includes event time (the time of last follow-up with a subject) and event indicator (did the subject die, or was the observation right-censored). -->
<!-- 2. `OmicsReg`---a data object for continuous responses (usually a vector of linear regression responses). -->
<!-- 3. `OmicsCateg`---a data object for categorical responses, the dependent variable of a generalized linear model. Currently, we only support binary classification (through logistic regression). -->
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"colonSurv_df"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"colon_pathwayCollection"</span>)

colon_OmicsSurv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CreateOmics.html">CreateOmics</a></span>(
  <span class="dt">assayData_df =</span> colonSurv_df[, <span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)],
  <span class="dt">pathwayCollection_ls =</span> colon_pathwayCollection,
  <span class="dt">response =</span> colonSurv_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],
  <span class="dt">respType =</span> <span class="st">"survival"</span>
)
<span class="co">#&gt; </span>
<span class="co">#&gt;   ======  Creating object of class OmicsSurv  =======</span>
<span class="co">#&gt; The input pathway database included 676 unique features.</span>
<span class="co">#&gt; The input assay dataset included 656 features.</span>
<span class="co">#&gt; Only pathways with at least 3 or more features included in the assay dataset are</span>
<span class="co">#&gt;   tested (specified by minPathSize parameter). There are 15 pathways which meet</span>
<span class="co">#&gt;   this criterion.</span>
<span class="co">#&gt; Because pathwayPCA is a self-contained test (PMID: 17303618), only features in</span>
<span class="co">#&gt;   both assay data and pathway database are considered for analysis. There are 615</span>
<span class="co">#&gt;   such features shared by the input assay and pathway database.</span></code></pre></div>
</div>
<div id="inspect-the-object" class="section level2">
<h2 class="hasAnchor">
<a href="#inspect-the-object" class="anchor"></a>3.2 Inspect the Object</h2>
<p>After you create an <code>Omics</code> object, print the object to the screen to view a summary of the data contained therein.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_OmicsSurv
<span class="co">#&gt; Formal class 'OmicsSurv' [package "pathwayPCA"] with 6 slots</span>
<span class="co">#&gt;   ..@ eventTime            : num [1:250] 64.9 59.8 62.4 54.5 46.3 ...</span>
<span class="co">#&gt;   ..@ eventObserved        : logi [1:250] FALSE FALSE FALSE FALSE TRUE FALSE ...</span>
<span class="co">#&gt;   ..@ assayData_df         :Classes 'tbl_df', 'tbl' and 'data.frame':    250 obs. of  656 variables:</span>
<span class="co">#&gt;   ..@ sampleIDs_char       : chr [1:250] "subj1" "subj2" "subj3" "subj4" ...</span>
<span class="co">#&gt;   ..@ pathwayCollection    :List of 3</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span>
<span class="co">#&gt;   ..@ trimPathwayCollection:List of 4</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span></code></pre></div>
</div>
<div id="detailed-object-views" class="section level2">
<h2 class="hasAnchor">
<a href="#detailed-object-views" class="anchor"></a>3.3 Detailed Object Views</h2>
<p>Because the printing procedure for <code>Omics</code> objects is to show a summary of the contents, you need to use the <code>get*()</code> functions to view the individual components of the <code>colon_OmicsSurv</code> object we just created. Overall, you can use accessor functions to extract, edit, or replace data contained in the object. The accessor functions are listed in more detail in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html#table-of-accessors">Table of Accessors</a> subsection of Chapter 3. Use these functions to confirm that the data object you created accurately reflects the data you intend to analyze.</p>
<div id="view-the-assay" class="section level3">
<h3 class="hasAnchor">
<a href="#view-the-assay" class="anchor"></a>3.3.1 View the Assay</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsPathway.html">getAssay</a></span>(colon_OmicsSurv)
<span class="co">#&gt; # A tibble: 250 x 656</span>
<span class="co">#&gt;       JUN    SOS2   PAK3    RAF1  PRKCB    BTC     SHC1   PRKCA   ELK1</span>
<span class="co">#&gt;     &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 -0.880  0.367   1.30   0.480  -0.867 -0.790 -2.84e-1  0.0104  1.04 </span>
<span class="co">#&gt;  2 -1.11   2.15    1.62   0.165   0.547 -0.261 -4.91e-4  1.18    1.49 </span>
<span class="co">#&gt;  3 -0.752  0.764   0.330 -0.541   0.164  0.452  6.91e-1 -0.950   0.667</span>
<span class="co">#&gt;  4  1.03  -0.751   2.75  -0.878  -1.09  -0.481 -1.34e+0  1.52    2.23 </span>
<span class="co">#&gt;  5 -1.73   0.613   2.66   0.550   0.197 -1.02   1.56e-1 -1.29    1.68 </span>
<span class="co">#&gt;  6 -0.162  0.126  -0.324  0.627  -0.106 -1.17   1.11e+0 -1.27    0.964</span>
<span class="co">#&gt;  7 -0.980 -0.520   1.28  -0.343  -0.762 -0.489 -2.57e-1  0.0518  1.21 </span>
<span class="co">#&gt;  8  0.616  0.0633  0.343 -0.0195  0.578 -1.37   6.38e-1  1.37    1.05 </span>
<span class="co">#&gt;  9  1.35  -0.467  -0.149 -0.718  -0.898  0.527 -6.04e-2  1.32   -1.000</span>
<span class="co">#&gt; 10 -0.538  0.413  -0.503 -1.27   -0.466  0.178  3.45e-1 -1.40   -0.378</span>
<span class="co">#&gt; # ... with 240 more rows, and 647 more variables: NRG1 &lt;dbl&gt;, PAK2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   MTOR &lt;dbl&gt;, PAK4 &lt;dbl&gt;, MAP2K4 &lt;dbl&gt;, EIF4EBP1 &lt;dbl&gt;, BAD &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PRKCG &lt;dbl&gt;, NRG3 &lt;dbl&gt;, MAPK9 &lt;dbl&gt;, ERBB4 &lt;dbl&gt;, MAPK10 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PTK2 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, ERBB3 &lt;dbl&gt;, MAP2K2 &lt;dbl&gt;, TGFA &lt;dbl&gt;,</span>
<span class="co">#&gt; #   BRAF &lt;dbl&gt;, MAP2K1 &lt;dbl&gt;, MAP2K7 &lt;dbl&gt;, ABL1 &lt;dbl&gt;, NRG2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   AKT1 &lt;dbl&gt;, ABL2 &lt;dbl&gt;, AKT2 &lt;dbl&gt;, SHC4 &lt;dbl&gt;, RPS6KB1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   RPS6KB2 &lt;dbl&gt;, AKT3 &lt;dbl&gt;, NRAS &lt;dbl&gt;, GRB2 &lt;dbl&gt;, AREG &lt;dbl&gt;,</span>
<span class="co">#&gt; #   STAT5B &lt;dbl&gt;, MAPK3 &lt;dbl&gt;, STAT5A &lt;dbl&gt;, PAK6 &lt;dbl&gt;, SOS1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   MYC &lt;dbl&gt;, MAPK1 &lt;dbl&gt;, NCK1 &lt;dbl&gt;, PIK3R5 &lt;dbl&gt;, NRG4 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   HRAS &lt;dbl&gt;, MAPK8 &lt;dbl&gt;, EGFR &lt;dbl&gt;, GSK3B &lt;dbl&gt;, CBLB &lt;dbl&gt;,</span>
<span class="co">#&gt; #   KRAS &lt;dbl&gt;, CBL &lt;dbl&gt;, SHC3 &lt;dbl&gt;, CDKN1B &lt;dbl&gt;, CDKN1A &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EGF &lt;dbl&gt;, EREG &lt;dbl&gt;, ARAF &lt;dbl&gt;, NCK2 &lt;dbl&gt;, SRC &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PIK3R3 &lt;dbl&gt;, CAMK2A &lt;dbl&gt;, CAMK2B &lt;dbl&gt;, CAMK2D &lt;dbl&gt;, CAMK2G &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PAK1 &lt;dbl&gt;, CBLC &lt;dbl&gt;, CRK &lt;dbl&gt;, PIK3CA &lt;dbl&gt;, PIK3CB &lt;dbl&gt;,</span>
<span class="co">#&gt; #   CRKL &lt;dbl&gt;, PIK3CD &lt;dbl&gt;, GAB1 &lt;dbl&gt;, PLCG1 &lt;dbl&gt;, PLCG2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   SHC2 &lt;dbl&gt;, HBEGF &lt;dbl&gt;, PIK3CG &lt;dbl&gt;, PIK3R1 &lt;dbl&gt;, PIK3R2 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EPHB2 &lt;dbl&gt;, EPHB4 &lt;dbl&gt;, EFNA5 &lt;dbl&gt;, PXN &lt;dbl&gt;, CDC42 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB3 &lt;dbl&gt;, RRAS &lt;dbl&gt;, GRB7 &lt;dbl&gt;, SYNJ1 &lt;dbl&gt;, EPHB3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB1 &lt;dbl&gt;, DNM1 &lt;dbl&gt;, MAP4K4 &lt;dbl&gt;, GRIA1 &lt;dbl&gt;, EPHB1 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   ROCK1 &lt;dbl&gt;, ITSN1 &lt;dbl&gt;, RAP1A &lt;dbl&gt;, RAC1 &lt;dbl&gt;, RAP1B &lt;dbl&gt;,</span>
<span class="co">#&gt; #   EFNB2 &lt;dbl&gt;, WASL &lt;dbl&gt;, TF &lt;dbl&gt;, …</span></code></pre></div>
</div>
<div id="view-the-pathwaycollection-list" class="section level3">
<h3 class="hasAnchor">
<a href="#view-the-pathwaycollection-list" class="anchor"></a>3.3.2 View the <code>pathwayCollection</code> List</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsPathway.html">getPathwayCollection</a></span>(colon_OmicsSurv)
<span class="co">#&gt; Object with Class(es) 'pathwayCollection', 'list' [package 'pathwayPCA'] with 3 elements: </span>
<span class="co">#&gt;  $ pathways:List of 15</span>
<span class="co">#&gt;  $ TERMS   : Named chr [1:15] "KEGG_PENTOSE_PHOSPHATE_PATHWAY" ...</span>
<span class="co">#&gt;  $ setsize : Named int [1:15] 27 64 ...</span></code></pre></div>
</div>
<div id="view-the-event-time" class="section level3">
<h3 class="hasAnchor">
<a href="#view-the-event-time" class="anchor"></a>3.3.3 View the Event Time</h3>
<p>We can use the vector subsetting mechanic in R (<code>vector[]</code>) to view only the first ten event times.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsSurv.html">getEventTime</a></span>(colon_OmicsSurv)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]
<span class="co">#&gt;  [1] 64.8657534 59.7698630 62.4000000 54.5095890 46.2904110 55.8575343</span>
<span class="co">#&gt;  [7] 57.9616438 54.0493151  0.4273973 41.4246575</span></code></pre></div>
</div>
<div id="view-the-event-indicator" class="section level3">
<h3 class="hasAnchor">
<a href="#view-the-event-indicator" class="anchor"></a>3.3.4 View the Event Indicator</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsSurv.html">getEvent</a></span>(colon_OmicsSurv)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]
<span class="co">#&gt;  [1] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE FALSE</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
</div>
<div id="test-pathways-for-significance" class="section level1">
<h1 class="hasAnchor">
<a href="#test-pathways-for-significance" class="anchor"></a>4. Test Pathways for Significance</h1>
<p>After you have confirmed that the <code>CreateOmics</code> function created the <code>Omics</code> object you wanted, you can analyze the object with adaptive, elastic-net, sparse (AES) PCA or supervised PCA. This section is a quick overview of the material covered in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html">Test Pathway Significance</a> vignette. For details of these methods functions, please see their respective sections in Chapter 4.</p>
<p>The function arguments are as follows. Both the <code>AESPCA_pVals</code> and <code>SuperPCA_pVals</code> functions take in an <code>Omics</code> object as the value to the <code>object</code> argument. AES-PCA can use permutation-based <span class="math inline">\(p\)</span>-values, so the <code>numReps</code> argument controls how many permutations to take. If we set the number of permutations to 0, then the <span class="math inline">\(p\)</span>-values will be calculated parametrically. The <code>numPCs</code> argument specifies how many principal components will be extracted from each pathway. The <code>parallel</code> and <code>numCores</code> arguments are used to control if and how the functions make use of parallel computing. Finally, the <code>adjustment</code> argument allows you to specify a family-wise error rate (FWER) or false discovery rate (FDR) adjustment for the pathway <span class="math inline">\(p\)</span>-values. These options are documented in the <code>adjustRaw_pVals</code> function (see the <a href="https://gabrielodom.github.io/pathwayPCA/reference/adjustRaw_pVals.html">help documentation</a> for details).</p>
<div id="aes-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#aes-pca" class="anchor"></a>4.1 AES-PCA</h2>
<p>Perform AES-PCA pathway significance testing on the <code>Omics</code> object with the <code>AESPCA_pVals</code> function. For more details on this function, see the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html#aes-pca">AES-PCA</a> section of Chapter 4. We will adjust the pathway <span class="math inline">\(p\)</span>-values by the estimated FDR calculated with the <code>"BH"</code> procedure (Benjamini and Hochberg, 1995).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsSurv,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="st">"BH"</span>
)
<span class="co">#&gt; Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 2: Calculate Pathway p-Values</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span></code></pre></div>
</div>
<div id="supervised-pca" class="section level2">
<h2 class="hasAnchor">
<a href="#supervised-pca" class="anchor"></a>4.2 Supervised PCA</h2>
<p>Perform Supervised PCA pathway significance testing on the <code>Omics</code> object with the <code>SuperPCA_pVals</code> function. For more details on this function, see the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html#supervised-pca">Supervised PCA</a> section of Chapter 4.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_superpcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SuperPCA_pVals.html">SuperPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsSurv,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="st">"BH"</span>
)
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Calculating Pathway Test Statistics in Parallel: DONE</span>
<span class="co">#&gt; Calculating Pathway Critical Values in Parallel: DONE</span>
<span class="co">#&gt; Calculating Pathway p-Values: DONE</span>
<span class="co">#&gt; Adjusting p-Values and Sorting Pathway p-Value Data Frame: DONE</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="inspect-results" class="section level1">
<h1 class="hasAnchor">
<a href="#inspect-results" class="anchor"></a>5. Inspect Results</h1>
<p>This section is a quick overview of the material covered in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html">Visualizing the Results</a> vignette. The output of <code><a href="../reference/AESPCA_pVals.html">AESPCA_pVals()</a></code> is list object with class <code>aespcOut</code>. The output of <code><a href="../reference/SuperPCA_pVals.html">SuperPCA_pVals()</a></code> is a list object with class <code>superpcOut</code>. Both of these objects have the following three elements:</p>
<ul>
<li>
<code>pVals_df</code>: a data frame with <span class="math inline">\(p\)</span>-values and their details each pathway</li>
<li>
<code>PCs_ls</code>: a list of the data frames of the first selected principal component(s) extracted from the assay data subset corresponding to each pathway.</li>
<li>
<code>loadings_ls</code>: a list of the matrices of the loading vectors that correspond to the principal components in <code>PCs_ls</code>.</li>
</ul>
<div id="analysis-output-table" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-output-table" class="anchor"></a>5.1 Analysis Output Table</h2>
<p>For a quick and easy view of the pathway significance testing results, we can simply print the output data frame. If you are not using the <code>tidyverse</code> package suite, your results will print differently (use the <code>head</code> function to print the top pathways instead).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_superpcOut<span class="op">$</span>pVals_df
<span class="co">#&gt; # A tibble: 15 x 5</span>
<span class="co">#&gt;    pathways    n_tested terms                                 rawp  FDR_BH</span>
<span class="co">#&gt;  * &lt;chr&gt;          &lt;int&gt; &lt;chr&gt;                                &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt;  1 pathway87         86 KEGG_ERBB_SIGNALING_PATHWAY        4.34e-5 6.51e-4</span>
<span class="co">#&gt;  2 pathway491        40 PID_EPHB_FWD_PATHWAY               1.61e-4 9.04e-4</span>
<span class="co">#&gt;  3 pathway757        83 REACTOME_INSULIN_RECEPTOR_SIGNAL…  1.81e-4 9.04e-4</span>
<span class="co">#&gt;  4 pathway176        54 KEGG_NON_SMALL_CELL_LUNG_CANCER    5.60e-4 2.10e-3</span>
<span class="co">#&gt;  5 pathway390        29 BIOCARTA_TNFR1_PATHWAY             4.84e-3 1.45e-2</span>
<span class="co">#&gt;  6 pathway177        26 KEGG_ASTHMA                        8.64e-3 2.16e-2</span>
<span class="co">#&gt;  7 pathway413        23 ST_GA12_PATHWAY                    2.09e-2 4.48e-2</span>
<span class="co">#&gt;  8 pathway3          26 KEGG_PENTOSE_PHOSPHATE_PATHWAY     3.56e-2 5.99e-2</span>
<span class="co">#&gt;  9 pathway60         45 KEGG_RETINOL_METABOLISM            3.60e-2 5.99e-2</span>
<span class="co">#&gt; 10 pathway1211      104 REACTOME_SIGNALING_BY_INSULIN_RE…  4.25e-2 6.37e-2</span>
<span class="co">#&gt; 11 pathway536        44 PID_TNF_PATHWAY                    4.74e-2 6.46e-2</span>
<span class="co">#&gt; 12 pathway781       180 REACTOME_PHOSPHOLIPID_METABOLISM   6.01e-2 7.52e-2</span>
<span class="co">#&gt; 13 pathway187        16 BIOCARTA_RELA_PATHWAY              1.75e-1 2.02e-1</span>
<span class="co">#&gt; 14 pathway120        73 KEGG_ANTIGEN_PROCESSING_AND_PRES…  2.64e-1 2.83e-1</span>
<span class="co">#&gt; 15 pathway266        11 BIOCARTA_SET_PATHWAY               3.86e-1 3.86e-1</span></code></pre></div>
</div>
<div id="graph-of-top-pathways" class="section level2">
<h2 class="hasAnchor">
<a href="#graph-of-top-pathways" class="anchor"></a>5.2 Graph of Top Pathways</h2>
<p>To visualize the significance of the pathways based on FDR or uncorrected <span class="math inline">\(p\)</span>-values, we can use the <a href="http://ggplot2.org/"><code>ggplot2</code> package</a> to create summary graphics of the analysis results.</p>
<div id="tidy-up-the-data" class="section level3">
<h3 class="hasAnchor">
<a href="#tidy-up-the-data" class="anchor"></a>5.2.1 Tidy Up the Data</h3>
<p>In order to take advantage of the publication-quality graphics created with the <code>ggplot2</code> package, we first need to tidy the data frames returned by the <code>AESPCA_pVals</code> and <code>SuperPCA_pVals</code> functions. The following code takes in the <span class="math inline">\(p\)</span>-values data frame (<code>colon_superpcOut$pVals_df</code>) from the Supervised PCA method, modifies it to be compatible with the <code>ggplot</code> function, and saves the new data frame (<code>colonOutGather_df</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonOutGather_df &lt;-<span class="st"> </span>colon_superpcOut<span class="op">$</span>pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>pathways, <span class="op">-</span>n_tested) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">gather</span>(variable, value, <span class="op">-</span>terms) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">variable =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(variable)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">variable =</span> <span class="kw">recode_factor</span>(variable,
                                  <span class="dt">rawp =</span> <span class="st">"None"</span>,
                                  <span class="dt">FDR_BH =</span> <span class="st">"FDR"</span>))

graphMax &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">ceiling</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Extremes">max</a></span>(colonOutGather_df<span class="op">$</span>score))

colonOutGather_df
<span class="co">#&gt; # A tibble: 30 x 4</span>
<span class="co">#&gt;    terms                                        variable     value score</span>
<span class="co">#&gt;    &lt;chr&gt;                                        &lt;fct&gt;        &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 KEGG_ERBB_SIGNALING_PATHWAY                  None     0.0000434 10.0 </span>
<span class="co">#&gt;  2 PID_EPHB_FWD_PATHWAY                         None     0.000161   8.74</span>
<span class="co">#&gt;  3 REACTOME_INSULIN_RECEPTOR_SIGNALLING_CASCADE None     0.000181   8.62</span>
<span class="co">#&gt;  4 KEGG_NON_SMALL_CELL_LUNG_CANCER              None     0.000560   7.49</span>
<span class="co">#&gt;  5 BIOCARTA_TNFR1_PATHWAY                       None     0.00484    5.33</span>
<span class="co">#&gt;  6 KEGG_ASTHMA                                  None     0.00864    4.75</span>
<span class="co">#&gt;  7 ST_GA12_PATHWAY                              None     0.0209     3.87</span>
<span class="co">#&gt;  8 KEGG_PENTOSE_PHOSPHATE_PATHWAY               None     0.0356     3.34</span>
<span class="co">#&gt;  9 KEGG_RETINOL_METABOLISM                      None     0.0360     3.33</span>
<span class="co">#&gt; 10 REACTOME_SIGNALING_BY_INSULIN_RECEPTOR       None     0.0425     3.16</span>
<span class="co">#&gt; # ... with 20 more rows</span></code></pre></div>
</div>
<div id="graph-pathway-ranks" class="section level3">
<h3 class="hasAnchor">
<a href="#graph-pathway-ranks" class="anchor"></a>5.2.2 Graph Pathway Ranks</h3>
<p>Now that our output is tidy, we can make a bar chart of the pathway significance based on the unadjusted <span class="math inline">\(p\)</span>-values.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">raw_df &lt;-<span class="st"> </span>colonOutGather_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(variable <span class="op">==</span><span class="st"> "None"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>variable, <span class="op">-</span>value)

<span class="kw">ggplot</span>(raw_df) <span class="op">+</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(terms, score), <span class="dt">y =</span> score) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">"identity"</span>, <span class="dt">position =</span> <span class="st">"dodge"</span>, <span class="dt">fill =</span> <span class="st">"#005030"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">guide =</span> <span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"Supervised PCA Significant Colon Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="st">"Negative Log p-Value"</span>, <span class="dt">limits =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, graphMax)) <span class="op">+</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(<span class="fl">0.01</span>), <span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="Supplement1-Quickstart_Guide_files/figure-html/surv_spr_pval_plot-1.png" width="100%" height="60%"></p>
<hr>
<p></p>
</div>
</div>
</div>
<div id="links-to-detailed-information" class="section level1">
<h1 class="hasAnchor">
<a href="#links-to-detailed-information" class="anchor"></a>6. Links to Detailed Information</h1>
<p>Now that you have an idea of how to use this package, please see each of our vignettes for detailed and thorough commentary and guiding information on each of the three topics discussed herein. The vignettes are:</p>
<ul>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html">Chapter 2: Import Data</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html">Chapter 3: Create <code>Omics</code> Data Objects</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html">Chapter 4: Test Pathway Significance</a></li>
<li><a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html">Chapter 5: Visualizing the Results</a></li>
</ul>
<p>Here is the R session information for this vignette:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.4.4 (2018-03-15)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.1 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] bindrcpp_0.2.2     pathwayPCA_1.0.0.0 forcats_0.3.0     </span>
<span class="co">#&gt;  [4] stringr_1.3.1      dplyr_0.7.8        purrr_0.2.5       </span>
<span class="co">#&gt;  [7] readr_1.1.1        tidyr_0.8.2        tibble_1.4.2      </span>
<span class="co">#&gt; [10] ggplot2_3.1.0.9000 tidyverse_1.2.1   </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] tidyselect_0.2.5 splines_3.4.4    haven_1.1.2      lattice_0.20-38 </span>
<span class="co">#&gt;  [5] colorspace_1.3-2 htmltools_0.3.6  yaml_2.2.0       utf8_1.1.4      </span>
<span class="co">#&gt;  [9] survival_2.43-1  rlang_0.3.0.1    pkgdown_1.2.0    pillar_1.3.0    </span>
<span class="co">#&gt; [13] withr_2.1.2      glue_1.3.0       modelr_0.1.2     readxl_1.1.0    </span>
<span class="co">#&gt; [17] bindr_0.1.1      plyr_1.8.4       munsell_0.5.0    commonmark_1.6  </span>
<span class="co">#&gt; [21] gtable_0.2.0     cellranger_1.1.0 rvest_0.3.2      memoise_1.1.0   </span>
<span class="co">#&gt; [25] evaluate_0.12    labeling_0.3     knitr_1.20       fansi_0.4.0     </span>
<span class="co">#&gt; [29] lars_1.2         broom_0.5.0      Rcpp_1.0.0       backports_1.1.2 </span>
<span class="co">#&gt; [33] scales_1.0.0     desc_1.2.0       jsonlite_1.5     fs_1.2.6        </span>
<span class="co">#&gt; [37] hms_0.4.2        digest_0.6.18    stringi_1.2.4    grid_3.4.4      </span>
<span class="co">#&gt; [41] rprojroot_1.3-2  cli_1.0.1        tools_3.4.4      magrittr_1.5    </span>
<span class="co">#&gt; [45] lazyeval_0.2.1   crayon_1.3.4     pkgconfig_2.0.2  Matrix_1.2-14   </span>
<span class="co">#&gt; [49] MASS_7.3-51.1    xml2_1.2.0       lubridate_1.7.4  assertthat_0.2.0</span>
<span class="co">#&gt; [53] rmarkdown_1.10   roxygen2_6.1.1   httr_1.3.1       rstudioapi_0.8  </span>
<span class="co">#&gt; [57] R6_2.3.0         nlme_3.1-137     compiler_3.4.4</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">1. Overview</a></li>
      <li>
<a href="#import-data">2. Import Data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#import--gmt-files">2.1 Import <code>.gmt</code> Files</a></li>
      <li><a href="#import-and-tidy-assay-data">2.2 Import and Tidy Assay Data</a></li>
      <li><a href="#import-phenotype-info">2.3 Import Phenotype Info</a></li>
      <li><a href="#match-the-phenotype-and-assay-data">2.4 Match the Phenotype and Assay Data</a></li>
      </ul>
</li>
      <li>
<a href="#create-an-omics-data-object">3. Create an <code>Omics</code> Data Object</a><ul class="nav nav-pills nav-stacked">
<li><a href="#create-an-object">3.1 Create an Object</a></li>
      <li><a href="#inspect-the-object">3.2 Inspect the Object</a></li>
      <li><a href="#detailed-object-views">3.3 Detailed Object Views</a></li>
      </ul>
</li>
      <li>
<a href="#test-pathways-for-significance">4. Test Pathways for Significance</a><ul class="nav nav-pills nav-stacked">
<li><a href="#aes-pca">4.1 AES-PCA</a></li>
      <li><a href="#supervised-pca">4.2 Supervised PCA</a></li>
      </ul>
</li>
      <li>
<a href="#inspect-results">5. Inspect Results</a><ul class="nav nav-pills nav-stacked">
<li><a href="#analysis-output-table">5.1 Analysis Output Table</a></li>
      <li><a href="#graph-of-top-pathways">5.2 Graph of Top Pathways</a></li>
      </ul>
</li>
      <li><a href="#links-to-detailed-information">6. Links to Detailed Information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, Lizhong Liu, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d">Lily Wang</a>, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
