<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Suppl. Ch. 5 - Visualizing the Results • pathwayPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Suppl. Ch. 5 - Visualizing the Results">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathwayPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    User Guides
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction_to_pathwayPCA.html">Integrative Pathway Analysis with pathwayPCA</a>
    </li>
    <li>
      <a href="../articles/Supplement1-Quickstart_Guide.html">Suppl. Ch. 1 - Quickstart Guide for New R Users</a>
    </li>
    <li>
      <a href="../articles/Supplement2-Importing_Data.html">Suppl. Ch. 2 - Import and Tidy Data</a>
    </li>
    <li>
      <a href="../articles/Supplement3-Create_Omics_Objects.html">Suppl. Ch. 3 - Creating Data Objects</a>
    </li>
    <li>
      <a href="../articles/Supplement4-Methods_Walkthrough.html">Suppl. Ch. 4 - Test Pathway Significance</a>
    </li>
    <li>
      <a href="../articles/Supplement5-Analyse_Results.html">Suppl. Ch.  5 - Visualizing the Results</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Version Logs</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Suppl. Ch. 5 - Visualizing the Results</h1>
                        <h4 class="author">Gabriel Odom</h4>
            
            <h4 class="date">2019-04-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gabrielodom/pathwayPCA/blob/master/vignettes/Supplement5-Analyse_Results.Rmd"><code>vignettes/Supplement5-Analyse_Results.Rmd</code></a></small>
      <div class="hidden name"><code>Supplement5-Analyse_Results.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>1. Overview</h1>
<p>This vignette is the fifth chapter in the “Pathway Significance Testing with <code>pathwayPCA</code>” workflow, providing a detailed perspective to the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement1-Quickstart_Guide.html#inspect-results">Inspect Results</a> section of the Quickstart Guide. This vignette will discuss graphing the pathway-specific <span class="math inline">\(p\)</span>-values, FDR values, and associated log-scores. Also, we will discuss how to plot the gene- or protein-specific loadings within a certain pathway and further how to plot the correlation between a gene or protein within a pathway and the principal component representing that pathway.</p>
<p>Before we move on, we will outline our steps. After reading this vignette, you should be able to</p>
<ol style="list-style-type: decimal">
<li>Plot the log-score rank of the top pathways.</li>
<li>Plot the feature loadings for a single pathway.</li>
<li>Plot the correlations between a pathway principal component and the features within that pathway.</li>
</ol>
<div id="packages" class="section level2">
<h2 class="hasAnchor">
<a href="#packages" class="anchor"></a>1.1 Packages</h2>
<p>First, load the <code>pathwayPCA</code> package and the <a href="https://www.tidyverse.org/"><code>tidyverse</code> package suite</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(pathwayPCA)</code></pre></div>
</div>
<div id="example-results" class="section level2">
<h2 class="hasAnchor">
<a href="#example-results" class="anchor"></a>1.2 Example Results</h2>
<p>First, we will replicate the data setup from the previous vignette chapters on importing data and creating data objects ( <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html">Chapter 2: Import and Tidy Data</a> and <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html">Chapter 3: Creating -Omics Data Objects</a>, respectively). See these chapters for detailed explination of the code below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load Data: see Chapter 2</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"colonSurv_df"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"colon_pathwayCollection"</span>)

<span class="co"># Create -Omics Container: see Chapter 3</span>
colon_OmicsSurv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/CreateOmics.html">CreateOmics</a></span>(
  <span class="dt">assayData_df =</span> colonSurv_df[, <span class="op">-</span>(<span class="dv">2</span><span class="op">:</span><span class="dv">3</span>)],
  <span class="dt">pathwayCollection_ls =</span> colon_pathwayCollection,
  <span class="dt">response =</span> colonSurv_df[, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>],
  <span class="dt">respType =</span> <span class="st">"surv"</span>
)
<span class="co">#&gt; </span>
<span class="co">#&gt;   ======  Creating object of class OmicsSurv  =======</span>
<span class="co">#&gt; The input pathway database included 676 unique features.</span>
<span class="co">#&gt; The input assay dataset included 656 features.</span>
<span class="co">#&gt; Only pathways with at least 3 or more features included in the assay dataset are</span>
<span class="co">#&gt;   tested (specified by minPathSize parameter). There are 15 pathways which meet</span>
<span class="co">#&gt;   this criterion.</span>
<span class="co">#&gt; Because pathwayPCA is a self-contained test (PMID: 17303618), only features in</span>
<span class="co">#&gt;   both assay data and pathway database are considered for analysis. There are 615</span>
<span class="co">#&gt;   such features shared by the input assay and pathway database.</span></code></pre></div>
<p>We will resume inspection of the analysis results from the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement4-Methods_Walkthrough.html">chapter 4 vignette</a> for the AESPCA method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># AESPCA Analysis: see Chapter 4</span>
colonSurv_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsSurv,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"BH"</span>, <span class="st">"Bonf"</span>)
)
<span class="co">#&gt; Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 2: Calculate Pathway p-Values</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span></code></pre></div>
<p></p>
<hr>
<p></p>
</div>
</div>
<div id="plot-pathway-significance-levels" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-pathway-significance-levels" class="anchor"></a>2. Plot Pathway Significance Levels</h1>
<p>The first element of this results data object is a data frame of pathways and their significance levels (<code>pVals_df</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonSurv_aespcOut<span class="op">$</span>pVals_df
<span class="co">#&gt; # A tibble: 15 x 6</span>
<span class="co">#&gt;    pathways   n_tested terms                   rawp  FDR_BH FWER_Bonferroni</span>
<span class="co">#&gt;  * &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;                  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt;  1 pathway491       40 PID_EPHB_FWD_PATHW~  6.53e-6 9.80e-5       0.0000980</span>
<span class="co">#&gt;  2 pathway781      180 REACTOME_PHOSPHOLI~  1.96e-4 1.47e-3       0.00295  </span>
<span class="co">#&gt;  3 pathway757       83 REACTOME_INSULIN_R~  4.90e-4 2.45e-3       0.00735  </span>
<span class="co">#&gt;  4 pathway177       26 KEGG_ASTHMA          8.21e-4 3.08e-3       0.0123   </span>
<span class="co">#&gt;  5 pathway87        86 KEGG_ERBB_SIGNALIN~  1.47e-3 4.42e-3       0.0221   </span>
<span class="co">#&gt;  6 pathway536       44 PID_TNF_PATHWAY      2.60e-3 6.51e-3       0.0390   </span>
<span class="co">#&gt;  7 pathway12~      104 REACTOME_SIGNALING~  4.42e-3 9.08e-3       0.0662   </span>
<span class="co">#&gt;  8 pathway3         26 KEGG_PENTOSE_PHOSP~  4.84e-3 9.08e-3       0.0726   </span>
<span class="co">#&gt;  9 pathway413       23 ST_GA12_PATHWAY      1.48e-2 2.46e-2       0.222    </span>
<span class="co">#&gt; 10 pathway60        45 KEGG_RETINOL_METAB~  2.55e-2 3.82e-2       0.382    </span>
<span class="co">#&gt; 11 pathway176       54 KEGG_NON_SMALL_CEL~  4.69e-2 6.39e-2       0.703    </span>
<span class="co">#&gt; 12 pathway120       73 KEGG_ANTIGEN_PROCE~  7.37e-2 9.21e-2       1        </span>
<span class="co">#&gt; 13 pathway187       16 BIOCARTA_RELA_PATH~  7.09e-1 8.01e-1       1        </span>
<span class="co">#&gt; 14 pathway390       29 BIOCARTA_TNFR1_PAT~  7.76e-1 8.01e-1       1        </span>
<span class="co">#&gt; 15 pathway266       11 BIOCARTA_SET_PATHW~  8.01e-1 8.01e-1       1</span></code></pre></div>
<p>Given the <span class="math inline">\(p\)</span>-values from these pathways, we will now graph their score (negative log of the <span class="math inline">\(p\)</span>-value) and description for the top 15 most-significant pathways.</p>
<div id="trim-pathway-names" class="section level2">
<h2 class="hasAnchor">
<a href="#trim-pathway-names" class="anchor"></a>2.1 Trim Pathway Names</h2>
<p>Because pathway names are quite long, we truncate names longer than 35 characters for better display in the graphs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pVals_df &lt;-<span class="st"> </span>
<span class="st">  </span>colonSurv_aespcOut<span class="op">$</span>pVals_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(
    <span class="dt">terms =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(
      <span class="kw">str_length</span>(terms) <span class="op">&gt;</span><span class="st"> </span><span class="dv">35</span>,
      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="kw">str_sub</span>(terms, <span class="dv">1</span>, <span class="dv">33</span>), <span class="st">"..."</span>),
      terms
    )
  )

pVals_df
<span class="co">#&gt; # A tibble: 15 x 6</span>
<span class="co">#&gt;    pathways   n_tested terms                   rawp  FDR_BH FWER_Bonferroni</span>
<span class="co">#&gt;    &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;                  &lt;dbl&gt;   &lt;dbl&gt;           &lt;dbl&gt;</span>
<span class="co">#&gt;  1 pathway491       40 PID_EPHB_FWD_PATHW~  6.53e-6 9.80e-5       0.0000980</span>
<span class="co">#&gt;  2 pathway781      180 REACTOME_PHOSPHOLI~  1.96e-4 1.47e-3       0.00295  </span>
<span class="co">#&gt;  3 pathway757       83 REACTOME_INSULIN_R~  4.90e-4 2.45e-3       0.00735  </span>
<span class="co">#&gt;  4 pathway177       26 KEGG_ASTHMA          8.21e-4 3.08e-3       0.0123   </span>
<span class="co">#&gt;  5 pathway87        86 KEGG_ERBB_SIGNALIN~  1.47e-3 4.42e-3       0.0221   </span>
<span class="co">#&gt;  6 pathway536       44 PID_TNF_PATHWAY      2.60e-3 6.51e-3       0.0390   </span>
<span class="co">#&gt;  7 pathway12~      104 REACTOME_SIGNALING~  4.42e-3 9.08e-3       0.0662   </span>
<span class="co">#&gt;  8 pathway3         26 KEGG_PENTOSE_PHOSP~  4.84e-3 9.08e-3       0.0726   </span>
<span class="co">#&gt;  9 pathway413       23 ST_GA12_PATHWAY      1.48e-2 2.46e-2       0.222    </span>
<span class="co">#&gt; 10 pathway60        45 KEGG_RETINOL_METAB~  2.55e-2 3.82e-2       0.382    </span>
<span class="co">#&gt; 11 pathway176       54 KEGG_NON_SMALL_CEL~  4.69e-2 6.39e-2       0.703    </span>
<span class="co">#&gt; 12 pathway120       73 KEGG_ANTIGEN_PROCE~  7.37e-2 9.21e-2       1        </span>
<span class="co">#&gt; 13 pathway187       16 BIOCARTA_RELA_PATH~  7.09e-1 8.01e-1       1        </span>
<span class="co">#&gt; 14 pathway390       29 BIOCARTA_TNFR1_PAT~  7.76e-1 8.01e-1       1        </span>
<span class="co">#&gt; 15 pathway266       11 BIOCARTA_SET_PATHW~  8.01e-1 8.01e-1       1</span></code></pre></div>
</div>
<div id="tidy-the-pathway-results" class="section level2">
<h2 class="hasAnchor">
<a href="#tidy-the-pathway-results" class="anchor"></a>2.2 Tidy the Pathway Results</h2>
<p>For the data frame containing the survival data pathway <span class="math inline">\(p\)</span>-values, we will transform the data for better graphics. This code takes in the pathway <span class="math inline">\(p\)</span>-values data frame from the AESPCA method output and <a href="https://rpubs.com/mm-c/gather-and-spread">gathers</a> it into a “tidy” data frame (compatible with <code>ggplot</code>). We also add on a column for the negative natural logarithm of the pathway <span class="math inline">\(p\)</span>-values (called <code>score</code>) and recode the label of the <span class="math inline">\(p\)</span>-value adjustment method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonOutGather_df &lt;-
<span class="st">  </span><span class="co"># Take in the results data frame,</span>
<span class="st">  </span>pVals_df <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># remove the columns we don't need,</span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>pathways, <span class="op">-</span>n_tested) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># "tidy" the data,</span>
<span class="st">  </span><span class="kw">gather</span>(variable, value, <span class="op">-</span>terms) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># add the score variable, and</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">score =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(value)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># store the adjustment methods as a factor</span>
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">variable =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(variable)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">mutate</span>(
    <span class="dt">variable =</span> <span class="kw">recode_factor</span>(
      variable,
      <span class="dt">rawp =</span> <span class="st">"None"</span>,
      <span class="dt">FDR_BH =</span> <span class="st">"Hochberg"</span>,
      <span class="dt">FWER_Bonferroni =</span> <span class="st">"Bonferroni"</span>
    )
  )</code></pre></div>
</div>
<div id="plot-significant-survival-pathways-for-one-adjustment" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-significant-survival-pathways-for-one-adjustment" class="anchor"></a>2.3 Plot Significant Survival Pathways for One Adjustment</h2>
<p>Now we will plot the pathway <span class="math inline">\(p\)</span>-values for the most significant pathways as a horizontal bar chart. For more information on how to modify <code>ggplot</code> graphs, or to learn how to create your own, please see Chang’s <a href="https://www.amazon.com/dp/1449316956/ref=cm_sw_su_dp?tag=ggplot2-20">R Graphics Cookbook</a> or Wickham’s <a href="http://ggplot2.org/book/"><code>ggplot2</code>: Elegant Graphics for Data Analysis</a>.</p>
<p>First, we select the rows of the <span class="math inline">\(p\)</span>-values data frame which correspond to the adjustment method we are interested in. We will select the Benjamini and Hochberg FDR-adjustment method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BHpVals_df &lt;-<span class="st"> </span>
<span class="st">  </span>colonOutGather_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(variable <span class="op">==</span><span class="st"> "Hochberg"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>variable)</code></pre></div>
<p>Now we plot the pathway significance level for the pathways based on this FDR-adjustment method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(BHpVals_df) <span class="op">+</span>
<span class="st">  </span><span class="co"># set overall appearance of the plot</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="co"># Define the dependent and independent variables</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(terms, score), <span class="dt">y =</span> score) <span class="op">+</span>
<span class="st">  </span><span class="co"># From the defined variables, create a vertical bar chart</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">position =</span> <span class="st">"dodge"</span>, <span class="dt">fill =</span> <span class="st">"#F47321"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set main and axis titles</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"AES-PCA Significant Pathways: Colon Cancer"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Negative Log BH-FDR"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add a line showing the alpha = 0.01 level</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(<span class="fl">0.01</span>), <span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">color =</span> <span class="st">"#005030"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Flip the x and y axes</span>
<span class="st">  </span><span class="kw">coord_flip</span>()</code></pre></div>
<p><img src="Supplement5-Analyse_Results_files/figure-html/bar_plots-1.png" width="100%" height="60%"></p>
</div>
<div id="plot-significant-survival-pathways-for-all-adjustments" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-significant-survival-pathways-for-all-adjustments" class="anchor"></a>2.4 Plot Significant Survival Pathways for All Adjustments</h2>
<p>If we were interested in comparing adjustment methods, we can. This figure shows that a few of the simulated pathways are significant at the <span class="math inline">\(\alpha = 0.01\)</span> for either the Benjamini and Hochberg or Bonferroni FWER approaches. The vertical black line is at <span class="math inline">\(-\log(p = 0.01)\)</span>. This figure is slightly different from the figures shown in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement1-Quickstart_Guide.html#graph-of-top-pathways">Graph Top Pathways</a> subsection of the Quickstart Guide.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(colonOutGather_df) <span class="op">+</span>
<span class="st">  </span><span class="co"># set overall appearance of the plot</span>
<span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span>
<span class="st">  </span><span class="co"># Define the dependent and independent variables</span>
<span class="st">  </span><span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/reorder.factor">reorder</a></span>(terms, score), <span class="dt">y =</span> score) <span class="op">+</span>
<span class="st">  </span><span class="co"># From the defined variables, create a vertical bar chart</span>
<span class="st">  </span><span class="kw">geom_col</span>(<span class="dt">position =</span> <span class="st">"dodge"</span>, <span class="kw">aes</span>(<span class="dt">fill =</span> variable)) <span class="op">+</span>
<span class="st">  </span><span class="co"># Set the legend, main titles, and axis titles</span>
<span class="st">  </span><span class="kw">scale_fill_discrete</span>(<span class="dt">guide =</span> <span class="ot">FALSE</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="st">"AES-PCA Significant Colon Pathways by FWER Adjustment"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Pathways"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">ylab</span>(<span class="st">"Negative Log p-Value"</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Add a line showing the alpha = 0.01 level</span>
<span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(<span class="fl">0.01</span>), <span class="dt">size =</span> <span class="dv">1</span>) <span class="op">+</span>
<span class="st">  </span><span class="co"># Flip the x and y axes</span>
<span class="st">  </span><span class="kw">coord_flip</span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="co"># Create a subplot for each p-value adjustment method</span>
<span class="st">  </span><span class="kw">facet_grid</span>(. <span class="op">~</span><span class="st"> </span>variable)</code></pre></div>
<p><img src="Supplement5-Analyse_Results_files/figure-html/surv_aes_pval_plot-1.png" width="100%" height="60%"></p>
<p></p>
<hr>
<p></p>
</div>
</div>
<div id="inspecting-the-driving-genes" class="section level1">
<h1 class="hasAnchor">
<a href="#inspecting-the-driving-genes" class="anchor"></a>3. Inspecting the Driving Genes</h1>
<p>Now that we have a few significant pathways, we can look at the loadings of each gene onto the first AES-PC from these pathways.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pVals_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(FDR_BH <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(pathways)
<span class="co">#&gt; # A tibble: 8 x 1</span>
<span class="co">#&gt;   pathways   </span>
<span class="co">#&gt;   &lt;chr&gt;      </span>
<span class="co">#&gt; 1 pathway491 </span>
<span class="co">#&gt; 2 pathway781 </span>
<span class="co">#&gt; 3 pathway757 </span>
<span class="co">#&gt; 4 pathway177 </span>
<span class="co">#&gt; 5 pathway87  </span>
<span class="co">#&gt; 6 pathway536 </span>
<span class="co">#&gt; 7 pathway1211</span>
<span class="co">#&gt; 8 pathway3</span></code></pre></div>
<div id="extract-pathway-decomposition" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-pathway-decomposition" class="anchor"></a>3.1 Extract Pathway Decomposition</h2>
<p>We will chose the top two significant pathways for closer inspection, and we want to ascertain which genes load onto the pathway PCs. Notice that the pathway loadings are named by the internal pathway key, so we need to use the <code>getPathPCL()</code> to match this key to its pathway and extract the pathway PCs and loadings (PC &amp; L) list from the <code>colonSurv_aespcOut</code> object. Here are the loading vectors and pathway details from <code>pathway491</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pathway491_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(colonSurv_aespcOut, <span class="st">"pathway491"</span>)
pathway491_ls
<span class="co">#&gt; $PCs</span>
<span class="co">#&gt; # A tibble: 250 x 3</span>
<span class="co">#&gt;    sampleID      V1     V2</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 subj1    -0.712   0.525</span>
<span class="co">#&gt;  2 subj2    -0.0580  0.293</span>
<span class="co">#&gt;  3 subj3    -1.65    0.367</span>
<span class="co">#&gt;  4 subj4     3.63    1.78 </span>
<span class="co">#&gt;  5 subj5    -1.60   -0.279</span>
<span class="co">#&gt;  6 subj6     1.95    3.12 </span>
<span class="co">#&gt;  7 subj7     2.38    1.48 </span>
<span class="co">#&gt;  8 subj8    -1.48   -0.126</span>
<span class="co">#&gt;  9 subj9    -0.102   0.364</span>
<span class="co">#&gt; 10 subj10    1.89    2.05 </span>
<span class="co">#&gt; # ... with 240 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Loadings</span>
<span class="co">#&gt; # A tibble: 40 x 3</span>
<span class="co">#&gt;    featureID    PC1   PC2</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 EPHB2      0.242 0.123</span>
<span class="co">#&gt;  2 EPHB4      0.168 0    </span>
<span class="co">#&gt;  3 EFNA5      0     0    </span>
<span class="co">#&gt;  4 MAPK1     -0.265 0.251</span>
<span class="co">#&gt;  5 SRC        0     0.416</span>
<span class="co">#&gt;  6 GRB2       0     0    </span>
<span class="co">#&gt;  7 MAP2K1     0     0    </span>
<span class="co">#&gt;  8 PXN        0     0.471</span>
<span class="co">#&gt;  9 CRK        0     0    </span>
<span class="co">#&gt; 10 CDC42      0     0    </span>
<span class="co">#&gt; # ... with 30 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $pathway</span>
<span class="co">#&gt; [1] "pathway491"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $term</span>
<span class="co">#&gt; [1] "PID_EPHB_FWD_PATHWAY"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $description</span>
<span class="co">#&gt; [1] NA</span></code></pre></div>
<p>This tells us that <code>pathway491</code> represents the <a href="http://software.broadinstitute.org/gsea/msigdb/geneset_page.jsp?geneSetName=PID_EPHB_FWD_PATHWAY">EPHB forward signaling</a> pathway.</p>
<p>We can also extract the PCs and loadings vectors for <code>pathway177</code> (the <a href="http://software.broadinstitute.org/gsea/msigdb/geneset_page.jsp?geneSetName=KEGG_ASTHMA">Asthma</a> pathway), which we will compare later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pathway177_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(colonSurv_aespcOut, <span class="st">"pathway177"</span>)</code></pre></div>
</div>
<div id="gene-loadings" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-loadings" class="anchor"></a>3.2 Gene Loadings</h2>
<div id="wrangle-pathway-loadings" class="section level3">
<h3 class="hasAnchor">
<a href="#wrangle-pathway-loadings" class="anchor"></a>3.2.1 Wrangle Pathway Loadings</h3>
<p>Given the loading vectors from the EPHB forward signaling pathway, we need to wrangle them into a data frame that the <code>ggplot()</code> function can use. For this, we’ve written a function that you can modify as you need. This function takes in the pathway PC &amp; L list returned by the <code><a href="../reference/getPathPCLs.html">getPathPCLs()</a></code> function and returns a data frame with all the components necessary for a plot of the gene loadings. For more aesthetically-pleasing plots, we set the default for one PC or loading vector per figure.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">TidyLoading &lt;-<span class="st"> </span><span class="cf">function</span>(pathwayPCL, <span class="dt">numPCs =</span> <span class="dv">1</span>){
  
  <span class="co"># Remove rows with 0 loading from the first numPCs columns</span>
  loadings_df &lt;-<span class="st"> </span>pathwayPCL<span class="op">$</span>Loadings
  totLoad_num  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(
    loadings_df[, <span class="dv">2</span><span class="op">:</span>(numPCs <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">drop =</span> <span class="ot">FALSE</span>]
  ))
  keepRows_idx &lt;-<span class="st"> </span>totLoad_num <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>
  loadings_df &lt;-<span class="st"> </span>loadings_df[keepRows_idx, , drop =<span class="st"> </span><span class="ot">FALSE</span>]
  
  <span class="co"># Sort the value on first feature</span>
  load_df &lt;-<span class="st"> </span>loadings_df <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="kw">arrange</span>(<span class="kw">desc</span>(PC1))
  
  <span class="co"># Wrangle the sorted loadings:</span>
  gg_df &lt;-<span class="st"> </span>load_df <span class="op">%&gt;%</span>
<span class="st">    </span><span class="co"># make the featureID column a factor,</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">featureID =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(featureID, <span class="dt">levels =</span> featureID)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="co"># rename the ID column,</span>
<span class="st">    </span><span class="kw">rename</span>(<span class="dt">ID =</span> featureID) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="co"># "tidy" the data, and</span>
<span class="st">    </span><span class="kw">gather</span>(Feature, Value, <span class="op">-</span>ID) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">    </span><span class="co"># add the up / down indicator.</span>
<span class="st">    </span><span class="kw">mutate</span>(<span class="dt">Direction =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(Value <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>, <span class="st">"Up"</span>, <span class="st">"Down"</span>))
  
  <span class="co"># Add the pathway name and description</span>
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(gg_df, <span class="st">"PathName"</span>) &lt;-<span class="st"> </span>pathwayPCL<span class="op">$</span>term
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(gg_df, <span class="st">"Description"</span>) &lt;-<span class="st"> </span>pathwayPCL<span class="op">$</span>description
  
  gg_df
  
}</code></pre></div>
<p>We can use this function to “tidy up” the loadings from the EPHB forward signaling pathway:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tidyLoad491_df &lt;-<span class="st"> </span><span class="kw">TidyLoading</span>(<span class="dt">pathwayPCL =</span> pathway491_ls)
tidyLoad491_df</code></pre></div>
</div>
<div id="plot-the-gene-loadings" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-the-gene-loadings" class="anchor"></a>3.2.2 Plot the Gene Loadings</h3>
<p>We also wrote a function to create a <code>ggplot</code> graph object which uses this tidy data frame.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PathwayBarChart &lt;-<span class="st"> </span><span class="cf">function</span>(ggData_df, y_lab,
                            <span class="dt">colours =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dt">Up =</span> <span class="st">"#009292"</span>, <span class="dt">Down =</span> <span class="st">"#920000"</span>),
                            <span class="dt">label_size =</span> <span class="dv">5</span>,
                            <span class="dt">sigFigs =</span> <span class="dv">0</span>,
                            ...){
  
  <span class="co"># Base plot layer</span>
  p1 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(ggData_df, <span class="kw">aes</span>(<span class="dt">x =</span> ID, <span class="dt">y =</span> Value, <span class="dt">color =</span> Direction))
  
  <span class="co"># Add geometric layer and split by Feature</span>
  p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_col</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> Direction), <span class="dt">width =</span> <span class="fl">0.5</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Feature, <span class="dt">nrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(ggData_df<span class="op">$</span>Feature)))
  
  <span class="co"># Add Gene Labels</span>
  p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_text</span>(
      <span class="kw">aes</span>(<span class="dt">y =</span> <span class="dv">0</span>, <span class="dt">label =</span> ID),
      <span class="dt">vjust =</span> <span class="st">"middle"</span>,
      <span class="dt">hjust =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(ggData_df<span class="op">$</span>Direction <span class="op">==</span><span class="st"> "Up"</span>, <span class="st">"top"</span>, <span class="st">"bottom"</span>),
      <span class="dt">size =</span> label_size, <span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">color =</span> <span class="st">"black"</span>
    )
  
  <span class="co"># Add Value Labels</span>
  p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span>
<span class="st">    </span><span class="kw">geom_text</span>(
      <span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(ggData_df<span class="op">$</span>Value <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="st">""</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>(Value, sigFigs))),
      <span class="dt">vjust =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ifelse">ifelse</a></span>(ggData_df<span class="op">$</span>Direction <span class="op">==</span><span class="st"> "Up"</span>, <span class="st">"top"</span>, <span class="st">"bottom"</span>),
      <span class="dt">hjust =</span> <span class="st">"middle"</span>,
      <span class="dt">size =</span> label_size, <span class="dt">color =</span> <span class="st">"black"</span>
    )
  
  <span class="co"># Set the Theme</span>
  p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span>
<span class="st">    </span><span class="kw">scale_y_continuous</span>(<span class="dt">expand =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.15</span>, <span class="fl">0.15</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">theme_classic</span>() <span class="op">+</span>
<span class="st">    </span><span class="kw">theme</span>(
      <span class="dt">axis.title.x =</span> <span class="kw">element_blank</span>(),
      <span class="dt">axis.text.x  =</span> <span class="kw">element_blank</span>(),
      <span class="dt">axis.ticks.x =</span> <span class="kw">element_blank</span>(),
      <span class="dt">strip.text   =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">14</span>),
      <span class="dt">legend.position =</span> <span class="st">"none"</span>
    )
  
  <span class="co"># Titles, Subtitles, and Labels</span>
  pathName &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(ggData_df, <span class="st">"PathName"</span>)
  pathDesc &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/attr">attr</a></span>(ggData_df, <span class="st">"Description"</span>)
  <span class="cf">if</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(pathDesc)){
    p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(pathName) <span class="op">+</span><span class="st"> </span><span class="kw">ylab</span>(y_lab)
  } <span class="cf">else</span> {
    p1 &lt;-<span class="st"> </span>p1 <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> pathName, <span class="dt">subtitle =</span> pathDesc, <span class="dt">y =</span> y_lab)
  }
  
  <span class="co"># Return</span>
  p1
  
}</code></pre></div>
<p>This is the resulting loading bar plot for the EPHB forward signaling pathway (because all of the loadings are between -1 and 1, we set the significant figures to 2):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#  pathway491_loadings_bar_chart, fig.height = 6, fig.width = 10.7, out.width = "100%", out.height = "60%"</span>
<span class="kw">PathwayBarChart</span>(
  <span class="dt">ggData_df =</span> tidyLoad491_df,
  <span class="dt">y_lab =</span> <span class="st">"PC Loadings"</span>,
  <span class="dt">sigFigs =</span> <span class="dv">2</span>
)</code></pre></div>
<p>For comparison, this is the bar plot for the asthma pathway (<code>pathway177</code> in our -Omics data container):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pathway177_loadings_bar_chart, fig.height = 6, fig.width = 10.7, out.width = "100%", out.height = "60%"</span>
<span class="kw">PathwayBarChart</span>(
  <span class="dt">ggData_df =</span> tidyLoad177_df,
  <span class="dt">y_lab =</span> <span class="st">"PC Loadings"</span>,
  <span class="dt">sigFigs =</span> <span class="dv">2</span>
)</code></pre></div>
</div>
</div>
<div id="gene-correlations-with-pcs" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-correlations-with-pcs" class="anchor"></a>3.3 Gene Correlations with PCs</h2>
<div id="calculate-pathway-correlations" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-pathway-correlations" class="anchor"></a>3.3.1 Calculate Pathway Correlations</h3>
<p>Additionally, we can calculate the correlation between gene expression in the assay corresponding to the EPHB forward signaling pathway and the PC vectors extracted from this pathway. Thus, we need to extract the recorded gene levels corresponding to the non-zero loadings for the PCs, calculate the Spearman correlations between these values and their extracted PCs, then wrangle these correlations into a data frame that the <code>ggplot()</code> function can use. We have written an additional function that you can modify as you need. This function takes in the <code>Omics</code> data container (that you created in Chapter 3) and the PC &amp; L list returned by <code><a href="../reference/getPathPCLs.html">getPathPCLs()</a></code>. This function returns a data frame with all the components necessary for a plot of the assay-PC correlations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">TidyCorrelation &lt;-<span class="st"> </span><span class="cf">function</span>(Omics, pathwayPCL, <span class="dt">numPCs =</span> <span class="dv">1</span>){
  
  loadings_df &lt;-<span class="st"> </span>pathwayPCL<span class="op">$</span>Loadings
  totLoad_num  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(
    loadings_df[, <span class="dv">2</span><span class="op">:</span>(numPCs <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">drop =</span> <span class="ot">FALSE</span>]
  ))
  keepRows_idx &lt;-<span class="st"> </span>totLoad_num <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>
  loadings_df &lt;-<span class="st"> </span>loadings_df[keepRows_idx, , drop =<span class="st"> </span><span class="ot">FALSE</span>]
  
  geneNames &lt;-<span class="st"> </span>loadings_df<span class="op">$</span>featureID
  geneCorr_mat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>(
    <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(
      pathwayPCL<span class="op">$</span>PCs[, <span class="op">-</span><span class="dv">1</span>],
      <span class="kw"><a href="../reference/get_set_OmicsPathway.html">getAssay</a></span>(Omics)[geneNames],
      <span class="dt">method =</span> <span class="st">"spearman"</span>
    )
  )
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(geneCorr_mat) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"PC"</span>, <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(geneCorr_mat))

  pathwayPCL<span class="op">$</span>Loadings &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(
    <span class="dt">featureID =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(geneCorr_mat),
    geneCorr_mat,
    <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>,
    <span class="dt">row.names =</span> <span class="ot">NULL</span>
  )
  
  <span class="kw">TidyLoading</span>(pathwayPCL, <span class="dt">numPCs =</span> numPCs)
  
}</code></pre></div>
<p>We can use this function to “tidy up” the loadings from the EPHB forward signaling pathway:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tidyCorr491_df &lt;-<span class="st"> </span><span class="kw">TidyCorrelation</span>(
  <span class="dt">Omics =</span> colon_OmicsSurv,
  <span class="dt">pathwayPCL =</span> pathway491_ls
)
tidyCorr491_df</code></pre></div>
</div>
<div id="plot-the-assay-pc-correlation" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-the-assay-pc-correlation" class="anchor"></a>3.3.2 Plot the Assay-PC Correlation</h3>
<p>This is the resulting correlation bar plot for the EPHB forward signaling pathway:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#  pathway491_correlations_bar_chart, fig.height = 6, fig.width = 10.7, out.width = "100%", out.height = "60%"</span>
<span class="kw">PathwayBarChart</span>(
  <span class="dt">ggData_df =</span> tidyCorr491_df,
  <span class="dt">y_lab =</span> <span class="st">"Assay-PC Correlations"</span>,
  <span class="dt">sigFigs =</span> <span class="dv">1</span>
)</code></pre></div>
<p>And this is the same figure for the asthma pathway.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pathway177_correlations_bar_chart, fig.height = 6, fig.width = 10.7, out.width = "100%", out.height = "60%"</span>
<span class="kw">PathwayBarChart</span>(
  <span class="dt">ggData_df =</span> tidyCorr177_df,
  <span class="dt">y_lab =</span> <span class="st">"Assay-PC Correlations"</span>,
  <span class="dt">sigFigs =</span> <span class="dv">1</span>
)</code></pre></div>
</div>
</div>
</div>
<div id="plot-patient-specific-pathway-activities" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-patient-specific-pathway-activities" class="anchor"></a>4. Plot patient-specific pathway activities</h1>
<p>estimate pathway activities for each patient, both for protein expressions and copy number expressions separately</p>
<p>This is the circle plot for the IL-1 signaling pathway:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Load data (already ordered by copyNumber PC1 score)</span>
circlePlotData &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/readRDS">readRDS</a></span>(<span class="st">"circlePlotData.RDS"</span>)

<span class="co"># Set color scales</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(grDevices)
bluescale &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/colorRamp">colorRampPalette</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"blue"</span>, <span class="st">"white"</span>))(<span class="dv">45</span>)
redscale  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/colorRamp">colorRampPalette</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"white"</span>, <span class="st">"red"</span>))(<span class="dv">38</span>)
Totalpalette &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(bluescale,redscale)

<span class="co"># Apply colours</span>
copyCol &lt;-<span class="st">  </span>Totalpalette[circlePlotData<span class="op">$</span>copyNumOrder]
proteoCol &lt;-<span class="st"> </span>Totalpalette[circlePlotData<span class="op">$</span>proteinOrder]</code></pre></div>
<p>Blue stands for negative PC1 score, Red stands for positive PC1 score. The inner loop stands for copy numbers, outer loop stands for protein expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(circlize)
<span class="co">#&gt; ========================================</span>
<span class="co">#&gt; circlize version 0.4.6</span>
<span class="co">#&gt; CRAN page: https://cran.r-project.org/package=circlize</span>
<span class="co">#&gt; Github page: https://github.com/jokergoo/circlize</span>
<span class="co">#&gt; Documentation: http://jokergoo.github.io/circlize_book/book/</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; If you use it in published research, please cite:</span>
<span class="co">#&gt; Gu, Z. circlize implements and enhances circular visualization </span>
<span class="co">#&gt;   in R. Bioinformatics 2014.</span>
<span class="co">#&gt; ========================================</span>

<span class="co"># Set parameters</span>
nPathway &lt;-<span class="st"> </span><span class="dv">83</span>
factors  &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>nPathway

<span class="co"># Plot Setup</span>
<span class="kw">circos.clear</span>()
<span class="kw">circos.par</span>(
  <span class="st">"gap.degree"</span> =<span class="st"> </span><span class="dv">0</span>,
  <span class="st">"cell.padding"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>),
  <span class="dt">start.degree =</span> <span class="dv">360</span><span class="op">/</span><span class="dv">40</span>,
  <span class="dt">track.margin =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">0</span>),
  <span class="st">"clock.wise"</span> =<span class="st"> </span><span class="ot">FALSE</span>
)
<span class="kw">circos.initialize</span>(<span class="dt">factors =</span> factors, <span class="dt">xlim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">3</span>))

<span class="co"># ProteoOmics</span>
<span class="kw">circos.trackPlotRegion</span>(
  <span class="dt">ylim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">3</span>),
  <span class="dt">factors =</span> factors,
  <span class="dt">bg.col =</span> proteoCol,
  <span class="dt">track.height =</span> <span class="fl">0.3</span>
)

<span class="co"># Copy Number</span>
<span class="kw">circos.trackPlotRegion</span>(
  <span class="dt">ylim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">0</span>, <span class="dv">3</span>),
  <span class="dt">factors =</span> factors,
  <span class="dt">bg.col =</span> copyCol,
  <span class="dt">track.height =</span> <span class="fl">0.3</span>
)

<span class="co"># Add labels</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressMessages</a></span>(
  <span class="kw">circos.trackText</span>(
    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="op">-</span><span class="dv">3</span>, nPathway),
    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="op">-</span><span class="fl">3.8</span>,nPathway),
    <span class="dt">labels =</span> <span class="st">"PID_IL1"</span>,
    <span class="dt">factors =</span> factors,
    <span class="dt">col =</span> <span class="st">"#2d2d2d"</span>,
    <span class="dt">font =</span> <span class="dv">2</span>,
    <span class="dt">adj =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="st">"adj"</span>),
    <span class="dt">cex =</span> <span class="fl">1.5</span>,
    <span class="dt">facing =</span> <span class="st">"downward"</span>,
    <span class="dt">niceFacing =</span> <span class="ot">TRUE</span>
  )
)</code></pre></div>
<p><img src="Supplement5-Analyse_Results_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<hr>
<p></p>
</div>
<div id="review" class="section level1">
<h1 class="hasAnchor">
<a href="#review" class="anchor"></a>5. Review</h1>
<p>We have has covered in this vignette:</p>
<ol style="list-style-type: decimal">
<li>Plotting the log-score rank of the top pathways.</li>
<li>Plotting the feature loadings for a single pathway.</li>
<li>Plotting the correlations between a pathway principal component and the features within that pathway.</li>
</ol>
<p>We are exploring options to connect this package to other pathway-testing software. We are further considering other summary functions and/or graphics to help display the output of our <code>*_pVals()</code> functions. If you have any questions, comments, complaints, or suggestions, please <a href="https://github.com/gabrielodom/pathwayPCA/issues/new">submit an issue ticket</a> and give us some feedback. Also you can visit our <a href="https://github.com/gabrielodom/pathwayPCA">development page</a>.</p>
<p>Here is the R session information for this vignette:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()
<span class="co">#&gt; R Under development (unstable) (2019-02-12 r76095)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 7 x64 (build 7601) Service Pack 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_United States.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                          </span>
<span class="co">#&gt; [5] LC_TIME=English_United States.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] circlize_0.4.6    pathwayPCA_0.99.5 forcats_0.4.0    </span>
<span class="co">#&gt;  [4] stringr_1.4.0     dplyr_0.8.0.1     purrr_0.3.2      </span>
<span class="co">#&gt;  [7] readr_1.3.1       tidyr_0.8.3       tibble_2.1.1     </span>
<span class="co">#&gt; [10] ggplot2_3.1.0     tidyverse_1.2.1  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.1          lubridate_1.7.4     lattice_0.20-38    </span>
<span class="co">#&gt;  [4] assertthat_0.2.1    rprojroot_1.3-2     digest_0.6.18      </span>
<span class="co">#&gt;  [7] utf8_1.1.4          R6_2.4.0            cellranger_1.1.0   </span>
<span class="co">#&gt; [10] plyr_1.8.4          backports_1.1.3     lars_1.2           </span>
<span class="co">#&gt; [13] evaluate_0.13       httr_1.4.0          pillar_1.3.1       </span>
<span class="co">#&gt; [16] GlobalOptions_0.1.0 rlang_0.3.3         lazyeval_0.2.2     </span>
<span class="co">#&gt; [19] readxl_1.3.1        rstudioapi_0.10     Matrix_1.2-17      </span>
<span class="co">#&gt; [22] rmarkdown_1.12      pkgdown_1.3.0       desc_1.2.0         </span>
<span class="co">#&gt; [25] labeling_0.3        splines_3.6.0       munsell_0.5.0      </span>
<span class="co">#&gt; [28] broom_0.5.1         compiler_3.6.0      modelr_0.1.4       </span>
<span class="co">#&gt; [31] xfun_0.6            pkgconfig_2.0.2     shape_1.4.4        </span>
<span class="co">#&gt; [34] htmltools_0.3.6     tidyselect_0.2.5    roxygen2_6.1.1     </span>
<span class="co">#&gt; [37] fansi_0.4.0         crayon_1.3.4        withr_2.1.2        </span>
<span class="co">#&gt; [40] MASS_7.3-51.4       commonmark_1.7      grid_3.6.0         </span>
<span class="co">#&gt; [43] nlme_3.1-137        jsonlite_1.6        gtable_0.3.0       </span>
<span class="co">#&gt; [46] magrittr_1.5        scales_1.0.0        cli_1.1.0          </span>
<span class="co">#&gt; [49] stringi_1.4.3       reshape2_1.4.3      fs_1.2.7           </span>
<span class="co">#&gt; [52] xml2_1.2.0          generics_0.0.2      tools_3.6.0        </span>
<span class="co">#&gt; [55] glue_1.3.1          hms_0.4.2           parallel_3.6.0     </span>
<span class="co">#&gt; [58] survival_2.44-1.1   yaml_2.2.0          colorspace_1.4-1   </span>
<span class="co">#&gt; [61] rvest_0.3.2         memoise_1.1.0       knitr_1.22         </span>
<span class="co">#&gt; [64] haven_2.1.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview">1. Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#packages">1.1 Packages</a></li>
      <li><a href="#example-results">1.2 Example Results</a></li>
      </ul>
</li>
      <li>
<a href="#plot-pathway-significance-levels">2. Plot Pathway Significance Levels</a><ul class="nav nav-pills nav-stacked">
<li><a href="#trim-pathway-names">2.1 Trim Pathway Names</a></li>
      <li><a href="#tidy-the-pathway-results">2.2 Tidy the Pathway Results</a></li>
      <li><a href="#plot-significant-survival-pathways-for-one-adjustment">2.3 Plot Significant Survival Pathways for One Adjustment</a></li>
      <li><a href="#plot-significant-survival-pathways-for-all-adjustments">2.4 Plot Significant Survival Pathways for All Adjustments</a></li>
      </ul>
</li>
      <li>
<a href="#inspecting-the-driving-genes">3. Inspecting the Driving Genes</a><ul class="nav nav-pills nav-stacked">
<li><a href="#extract-pathway-decomposition">3.1 Extract Pathway Decomposition</a></li>
      <li><a href="#gene-loadings">3.2 Gene Loadings</a></li>
      <li><a href="#gene-correlations-with-pcs">3.3 Gene Correlations with PCs</a></li>
      </ul>
</li>
      <li><a href="#plot-patient-specific-pathway-activities">4. Plot patient-specific pathway activities</a></li>
      <li><a href="#review">5. Review</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, Lizhong Liu, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d">Lily Wang</a>, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
