<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Suppl. Ch. 4 - Test Pathway Significance • pathwayPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Suppl. Ch. 4 - Test Pathway Significance">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pathwayPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown-header">User Guides</li>
<li class="dropdown-header">Version Logs</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/gabrielodom/pathwayPCA">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Suppl. Ch. 4 - Test Pathway Significance</h1>
                        <h4 class="author">Gabriel Odom</h4>
            
            <h4 class="date">2019-02-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/gabrielodom/pathwayPCA/blob/master/vignettes/Supplement4-Methods_Walkthrough.Rmd"><code>vignettes/Supplement4-Methods_Walkthrough.Rmd</code></a></small>
      <div class="hidden name"><code>Supplement4-Methods_Walkthrough.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>1. Overview</h1>
<p>This vignette is the fourth chapter in the “Pathway Significance Testing with <code>pathwayPCA</code>” workflow, providing a detailed perspective to the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement1-Quickstart_Guide.html#test-pathways-for-significance">Pathway Significance Testing</a> section of the Quickstart Guide. This vignette builds on the material covered in the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement2-Importing_Data.html">“Import and Tidy Data”</a> and <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement3-Create_Omics_Objects.html">“Creating -Omics Data Objects”</a> vignettes. This guide will outline the major steps needed analyze <code>Omics</code>-class objects with pathway-level adaptive, elastic-net, sparse or supervised modifications to principal components analysis (PCA), abbreviated AES-PCA and Supervised PCA, respectively. We will consider examples for three types of response information: survival, regression, and binary responses. The predictor information is subsets of assay data which correspond to individual pathways, where a pathway is a bundle of genes with shared biological function. The main goal of pathway significance testing is to discover potential relationships between a given collection of pathways and the response.</p>
<div id="outline" class="section level2">
<h2 class="hasAnchor">
<a href="#outline" class="anchor"></a>1.1 Outline</h2>
<p>Before we move on, we will outline our steps. After reading this vignette, you should</p>
<ol style="list-style-type: decimal">
<li>Understand the basics of the AES-PCA pathway-significance testing approach.</li>
<li>Understand the basics of the Supervised PCA pathway-significance testing approach.</li>
<li>Be able to apply AES-PCA or Supervised PCA to analyze <code>Omics</code> data objects with survival, regression, or classification response.</li>
</ol>
</div>
<div id="load-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages" class="anchor"></a>1.2 Load Packages</h2>
<p>Before we begin, make sure you have the latest version of the package. In order to install a package from GitHub, you will need the <code>devtools::</code> package (<a href="https://github.com/r-lib/devtools" class="uri">https://github.com/r-lib/devtools</a>) and either <a href="https://cran.r-project.org/bin/windows/Rtools/">Rtools</a> (for Windows) or <a href="https://developer.apple.com/xcode/">Xcode</a> (for Mac). Then you can install the development version of the <a href="https://github.com/gabrielodom/pathwayPCA/tree/devel"><code>pathwayPCA</code> package</a> from <a href="https://github.com/">GitHub</a>:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">devtools::install_github("gabrielodom/pathwayPCA", ref = "devel")</a></code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(parallel)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyverse)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(pathwayPCA)</code></pre></div>
</div>
<div id="load-omics-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-omics-data" class="anchor"></a>1.3 Load <code>Omics</code> Data</h2>
<p>Because you have already read through the <a href="https://gabrielodom.github.io/pathwayPCA/articles/Importing_Data.html">Import and Tidy Data</a> and <a href="https://gabrielodom.github.io/pathwayPCA/articles/Create_Omics_Objects.html">Creating -Omics Data Objects</a> vignettes, we will pick up with the <code>colon_OmicsSurv</code> object we created in the last vignette. For our pathway analysis to be meaningful, we need gene expression data (from a microarray or something similar), corresponding phenotype information (such as weight, type of cancer, or survival time and censoring indicator), and a pathways list. The <code>colon_OmicsSurv</code> data object we constructed in <a href="https://gabrielodom.github.io/pathwayPCA/articles/Create_Omics_Objects.html">Chapter 3</a> has all of this.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colon_OmicsSurv
<span class="co">#&gt; Formal class 'OmicsSurv' [package "pathwayPCA"] with 6 slots</span>
<span class="co">#&gt;   ..@ eventTime            : num [1:250] 64.9 59.8 62.4 54.5 46.3 ...</span>
<span class="co">#&gt;   ..@ eventObserved        : logi [1:250] FALSE FALSE FALSE FALSE TRUE FALSE ...</span>
<span class="co">#&gt;   ..@ assayData_df         :Classes 'tbl_df', 'tbl' and 'data.frame':    250 obs. of  656 variables:</span>
<span class="co">#&gt;   ..@ sampleIDs_char       : chr [1:250] "subj1" "subj2" "subj3" "subj4" ...</span>
<span class="co">#&gt;   ..@ pathwayCollection    :List of 3</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span>
<span class="co">#&gt;   ..@ trimPathwayCollection:List of 4</span>
<span class="co">#&gt;   .. ..- attr(*, "class")= chr [1:2] "pathwayCollection" "list"</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="pathway-testing-setup" class="section level1">
<h1 class="hasAnchor">
<a href="#pathway-testing-setup" class="anchor"></a>2. Pathway Testing Setup</h1>
<p>In this section, we will describe the workflow of the Supervised PCA (<code>SuperPCA_pVals</code>) and AES-PCA (<code>AESPCA_pVals</code>) pathway significance-testing methods. <strong>The implementation of Supervised PCA in this package does not currently support analysis of responses with missingness.</strong> If you plan to test your pathways using the Supervised PCA method, please remove observations with missing entries before analysis. Unlike the current implementation of Supervised PCA, our current implementation of AES-PCA can handle some missingness in the response.</p>
<p>Also, when we compare computing times in this vignette, we use a Dell Precision Tower 5810 with 64-bit Windows 7 Enterprise OS. This machine has 64 GB of RAM and an Intel Xeon E5-2640 v4 2.40 GHz processor with 20 threads. We use two threads for parallel computing. Please adjust your expectations of computing time accordingly.</p>
<div id="pathway-significance-testing-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#pathway-significance-testing-overview" class="anchor"></a>2.1 Pathway Significance Testing Overview</h2>
<p>Now that we have our data stored in an <code>Omics</code>-class object, we can test the significance of each pathway with AES- or Supervised PCA. These functions both</p>
<ol style="list-style-type: decimal">
<li>Extract the first principal components (PCs) from each pathway-subset of the assay design matrix.</li>
<li>Test the association between the extracted PCs and the response matrix (survival) or vector (all others).</li>
<li>Adjust the pathway <span class="math inline">\(p\)</span>-values for False Discovery Rate (FDR) or Family-wise Error Rate (FWER).</li>
<li>Return a sorted data frame of the adjusted <span class="math inline">\(p\)</span>-values, a list of the first PCs, and a list of the first loading vectors, all for each pathway.</li>
</ol>
<p>The major differences between the AES-PCA and Supervised PCA methods involve the execution of (1) and (2), which we will describe in their respective methods sections.</p>
</div>
<div id="extract-pathway-pcs" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-pathway-pcs" class="anchor"></a>2.2 Extract Pathway PCs</h2>
<p>The details of this step will depend on the method, but the overall idea remains the same. For each pathway in the trimmed pathway collection, select the columns of the assay data frame that correspond to each genes contained within that pathway. Then, given the pathway-specific assay data subset, use the chosen PCA method to extract the first PCs from that subset of the assay data. The end result of this step is a list of the first PCs and a list of the loading vectors which correspond to these PCs.</p>
</div>
<div id="test-pathway-association" class="section level2">
<h2 class="hasAnchor">
<a href="#test-pathway-association" class="anchor"></a>2.3 Test Pathway Association</h2>
<p>The details of this step will also depend on the method. At this point in the method execution, we will have a list of PCs representing the data corresponding to each pathway. We then apply simple models to test if the PCs associated with that pathway are significantly related to the output. For survival output, we use Cox Proportional Hazards (Cox PH) regression. For categorical output, (because we only support binary responses in this version) we use logistic regression to test for a relationship between pathway PCs and the response. For continuous output, we use a simple multiple regression model. The AES- and Supervised PCA methods differ on <em>how</em> the <span class="math inline">\(p\)</span>-values from these models are calculated, but the end result of this step is a <span class="math inline">\(p\)</span>-value for each of the trimmed pathways.</p>
</div>
<div id="adjust-the-pathway-p-values-for-fdr" class="section level2">
<h2 class="hasAnchor">
<a href="#adjust-the-pathway-p-values-for-fdr" class="anchor"></a>2.4 Adjust the Pathway <span class="math inline">\(p\)</span>-Values for FDR</h2>
<p>At this step, we have a vector of <span class="math inline">\(p\)</span>-values corresponding to the list of trimmed pathways. We know that repeated comparisons inflate the Type-I error rate, so we adjust these <span class="math inline">\(p\)</span>-values to control the Type-I error. We use the FDR adjustments executed in the <code>mt.rawp2adjp</code> function from the <code>multtest</code> <a href="https://www.bioconductor.org/packages/3.7/bioc/manuals/multtest/man/multtest.pdf">Bioconductor package</a>. We modified this function’s code to better fit into our package workflow. While we do not depend on this package directly, we acknowledge their work in this area and express our gratitude. Common adjustment methods to control the FWER or FDR are the Bonferroni, Sidak, Holm, or Benjamini and Hochberg techniques.</p>
</div>
<div id="output-a-sorted-data-frame-tibble" class="section level2">
<h2 class="hasAnchor">
<a href="#output-a-sorted-data-frame-tibble" class="anchor"></a>2.5 Output a Sorted Data Frame / Tibble</h2>
<p>The end result of either PCA variant is a data frame (<code>pVals_df</code>), list of PCs (<code>PCs_ls</code>), and list of loadings to match the PCs (<code>loadings_ls</code>). The <span class="math inline">\(p\)</span>-values data frame has the following columns:</p>
<ul>
<li>
<code>pathways</code>: The names of the pathways in the <code>Omics</code> object. The names will match those given in</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(<span class="kw"><a href="../reference/get_set_OmicsPathway.html">getPathwayCollection</a></span>(colon_OmicsSurv)<span class="op">$</span>pathways)
<span class="co">#&gt;  [1] "pathway3"    "pathway60"   "pathway87"   "pathway120"  "pathway176" </span>
<span class="co">#&gt;  [6] "pathway177"  "pathway187"  "pathway266"  "pathway390"  "pathway413" </span>
<span class="co">#&gt; [11] "pathway491"  "pathway536"  "pathway757"  "pathway781"  "pathway1211"</span></code></pre></div>
<ul>
<li>
<code>n_tested</code>: The number of genes in each of the pathways after trimming to match the given data assay. The number of genes per pathway given in</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsPathway.html">getTrimPathwayCollection</a></span>(colon_OmicsSurv)<span class="op">$</span>n_tested
<span class="co">#&gt;    pathway3   pathway60   pathway87  pathway120  pathway176  pathway177 </span>
<span class="co">#&gt;          26          45          86          73          54          26 </span>
<span class="co">#&gt;  pathway187  pathway266  pathway390  pathway413  pathway491  pathway536 </span>
<span class="co">#&gt;          16          11          29          23          40          44 </span>
<span class="co">#&gt;  pathway757  pathway781 pathway1211 </span>
<span class="co">#&gt;          83         180         104</span></code></pre></div>
<ul>
<li>
<code>terms</code>: The pathway description, as given in</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/get_set_OmicsPathway.html">getPathwayCollection</a></span>(colon_OmicsSurv)<span class="op">$</span>TERMS
<span class="co">#&gt;                                       pathway3 </span>
<span class="co">#&gt;               "KEGG_PENTOSE_PHOSPHATE_PATHWAY" </span>
<span class="co">#&gt;                                      pathway60 </span>
<span class="co">#&gt;                      "KEGG_RETINOL_METABOLISM" </span>
<span class="co">#&gt;                                      pathway87 </span>
<span class="co">#&gt;                  "KEGG_ERBB_SIGNALING_PATHWAY" </span>
<span class="co">#&gt;                                     pathway120 </span>
<span class="co">#&gt;     "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION" </span>
<span class="co">#&gt;                                     pathway176 </span>
<span class="co">#&gt;              "KEGG_NON_SMALL_CELL_LUNG_CANCER" </span>
<span class="co">#&gt;                                     pathway177 </span>
<span class="co">#&gt;                                  "KEGG_ASTHMA" </span>
<span class="co">#&gt;                                     pathway187 </span>
<span class="co">#&gt;                        "BIOCARTA_RELA_PATHWAY" </span>
<span class="co">#&gt;                                     pathway266 </span>
<span class="co">#&gt;                         "BIOCARTA_SET_PATHWAY" </span>
<span class="co">#&gt;                                     pathway390 </span>
<span class="co">#&gt;                       "BIOCARTA_TNFR1_PATHWAY" </span>
<span class="co">#&gt;                                     pathway413 </span>
<span class="co">#&gt;                              "ST_GA12_PATHWAY" </span>
<span class="co">#&gt;                                     pathway491 </span>
<span class="co">#&gt;                         "PID_EPHB_FWD_PATHWAY" </span>
<span class="co">#&gt;                                     pathway536 </span>
<span class="co">#&gt;                              "PID_TNF_PATHWAY" </span>
<span class="co">#&gt;                                     pathway757 </span>
<span class="co">#&gt; "REACTOME_INSULIN_RECEPTOR_SIGNALLING_CASCADE" </span>
<span class="co">#&gt;                                     pathway781 </span>
<span class="co">#&gt;             "REACTOME_PHOSPHOLIPID_METABOLISM" </span>
<span class="co">#&gt;                                    pathway1211 </span>
<span class="co">#&gt;       "REACTOME_SIGNALING_BY_INSULIN_RECEPTOR"</span></code></pre></div>
<ul>
<li>
<code>rawp</code>: The unadjusted <span class="math inline">\(p\)</span>-values of each pathway.</li>
<li>
<code>...</code>: Additional columns for each requested FDR/FWER adjustment.</li>
</ul>
<p>The data frame will have its rows sorted in increasing order by the adjusted <span class="math inline">\(p\)</span>-value corresponding to the first adjustment method requested. Ties are broken by the raw <span class="math inline">\(p\)</span>-values. Additionally, if you use the <a href="https://www.tidyverse.org/"><code>tidyverse</code></a> package suite (and have these packages loaded), then the output will be a tibble object, rather than a data frame object. This object class comes with enhanced printing methods and some other benefits.</p>
<hr>
<p></p>
</div>
</div>
<div id="aes-pca" class="section level1">
<h1 class="hasAnchor">
<a href="#aes-pca" class="anchor"></a>3. AES-PCA</h1>
<p>Now that we have described the overview of the pathway analysis methods, we can discuss and give examples in more detail.</p>
<div id="method-details" class="section level2">
<h2 class="hasAnchor">
<a href="#method-details" class="anchor"></a>3.1 Method Details</h2>
<div id="aes-pca-method-sources" class="section level3">
<h3 class="hasAnchor">
<a href="#aes-pca-method-sources" class="anchor"></a>3.1.1 AES-PCA Method Sources</h3>
<p>Adaptive, elastic-net, sparse PCA is a combination of the <a href="https://doi.org/10.1214/08-AOS625">Adaptive Elastic-Net</a> of Zou and Zhang (2009) and <a href="https://doi.org/10.1198/106186006X113430">Sparse PCA</a> of Zou et al. (2006). This method was applied to pathways association testing by <a href="https://doi.org/10.2202/1544-6115.1697">Chen (2011)</a>. Accoding to Chen (2011), the “AES-PCA method removes noisy expression signals and also account[s] for correlation structure between the genes. It is computationally efficient, and the estimation of the PCs does not depend on clinical outcomes.” This package uses a legacy version of the <a href="https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf">LARS algorithm</a> of Efron et al. (2003) to calculate the PCs.</p>
</div>
<div id="calculate-pathway-specific-model-p-values" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-pathway-specific-model-p-values" class="anchor"></a>3.1.2 Calculate Pathway-Specific Model <span class="math inline">\(p\)</span>-Values</h3>
<p>For the AES-PCA method, pathway <span class="math inline">\(p\)</span>-values can be calculated with a permutation test. Therefore, when testing the relationship between the response and the PCs extracted by AES-PCA, the accuracy of the permuted <span class="math inline">\(p\)</span>-values will depend on how many permutations you call for. We recommend 1000. Be warned, however, that this may be too few permutations to create accurate seperation in pathway significance <span class="math inline">\(p\)</span>-values. You could increase the permutations to a larger value, should your computing resources allow for that. For even moderately-sized data sets (~2000 features) and 1000 pathways, this could take half an hour or more. If you choose to calculate the pathway <span class="math inline">\(p\)</span>-values non-parametrically, about 20-30% of the computing costs will be extracting the AES-PCs from each pathway (though this proportion will increase if the LARS algorithm has convergence issues with the given pathway). The remaining 70-80% of the cost will be the permutation test (for 1000 permutations).</p>
</div>
<div id="aes-pca-pros-and-cons" class="section level3">
<h3 class="hasAnchor">
<a href="#aes-pca-pros-and-cons" class="anchor"></a>3.1.3 AES-PCA Pros and Cons</h3>
<p>Pros:</p>
<ul>
<li>The AES-PCA method can handle some missingness in the response.</li>
<li>The <span class="math inline">\(p\)</span>-values can be calculated non-parametrically.</li>
</ul>
<p>Cons:</p>
<ul>
<li>The AES-PCA algorithm requires optimization over two tuning parameters and can therefore be considerably slower than using the singular value decomposition or eigendecomposition to extract PCs.</li>
<li>The <span class="math inline">\(p\)</span>-values calculated may be too discrete for fewer than 10,000 permutations, which can affect the behavior of the adjustment procedures.</li>
</ul>
</div>
</div>
<div id="aes-pca-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#aes-pca-examples" class="anchor"></a>3.2 AES-PCA Examples</h2>
<p>Now that we have discussed both the overview of the AES-PCA method and some of its specific details, we can run some examples. We have included in this package a toy data collection: a small tidy assay and corresponding pathway collection. This assay has 656 gene expression measurements on 250 colon cancer patients. Survival responses pertaining to these patients are also included. Further, the subset of the pathways collection containts 15 pathways which match most of the genes measured in our example colon cancer assay.</p>
<div id="survival-response" class="section level3">
<h3 class="hasAnchor">
<a href="#survival-response" class="anchor"></a>3.2.1 Survival Response</h3>
<p>We will use two of our available cores with the parallel computing approach. We will adjust the <span class="math inline">\(p\)</span>-values with the Hochberg (1988) and Sidak Step-Down FWER-adjustment procedures. We will now describe the computational cost for the non-parametric approach.</p>
<p>For the tiny <span class="math inline">\(250 \times 656\)</span> assay with 15 associated pathways, calculating pathway <span class="math inline">\(p\)</span>-values with 1000 replicates completes in 28 seconds. If we increase the number of permutations from 1000 to 10,000, this calculation takes 222 seconds (<span class="math inline">\(7.9\times\)</span> longer). Even though we increased the permutations tenfold, the function completed execution less than 10 times longer (as we mentioned above, roughly a quarter of the computing time is extracting the PCs from each pathway, which does not depend on the number of permutations).</p>
<p>In the example that we show, we will calculate the pathway <span class="math inline">\(p\)</span>-values parametrically, by specifying <code>numReps = 0</code>. Furthermore, the AES-PCA and Supervised PCA functions give some messages concerning the setup and progress of the computation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonSurv_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsSurv,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Hoch"</span>, <span class="st">"SidakSD"</span>)
)
<span class="co">#&gt; Part 1: Calculate Pathway AES-PCs</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway PCs in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 2: Calculate Pathway p-Values</span>
<span class="co">#&gt; Initializing Computing Cluster: DONE</span>
<span class="co">#&gt; Extracting Pathway p-Values in Parallel: DONE</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Part 3: Adjusting p-Values and Sorting Pathway p-Value Data Frame</span>
<span class="co">#&gt; DONE</span></code></pre></div>
</div>
<div id="regression-response" class="section level3">
<h3 class="hasAnchor">
<a href="#regression-response" class="anchor"></a>3.2.2 Regression Response</h3>
<p>We can also make a mock regression data set by treating the event time as the necessary continuous response. For this example, we will adjust the <span class="math inline">\(p\)</span>-values with the Holm (1979) FWER- and Benjamini and Hochberg (1995) FDR-adjustment procedures (as an aside, note that this type of multiple testing violates the independence assumption of the <a href="https://doi.org/10.1214/193940307000000167">Simes inequality</a>). For 1000 permutations, this calculation takes 17 seconds. For 10,000 permutations, this calculation takes 102 seconds (<span class="math inline">\(6.1\times\)</span> longer).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonReg_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsReg,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Holm"</span>, <span class="st">"BH"</span>)
)</code></pre></div>
</div>
<div id="binary-classification-response" class="section level3">
<h3 class="hasAnchor">
<a href="#binary-classification-response" class="anchor"></a>3.2.3 Binary Classification Response</h3>
<p>Finally, we can simulate a mock classification data set by treating the event indicator as the necessary binary response. For this example, we will adjust the <span class="math inline">\(p\)</span>-values with the Sidak Single-Step FWER- and Benjamini and Yekutieli (2001) FDR-adjustment procedures. For 1000 permutations, this calculation takes 30 seconds. For 10,000 permutations, this calculation takes 226 seconds (<span class="math inline">\(7.6\times\)</span> longer).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonCateg_aespcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/AESPCA_pVals.html">AESPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsCateg,
  <span class="dt">numReps =</span> <span class="dv">0</span>,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"SidakSS"</span>, <span class="st">"BY"</span>)
)</code></pre></div>
<hr>
<p></p>
</div>
</div>
</div>
<div id="supervised-pca" class="section level1">
<h1 class="hasAnchor">
<a href="#supervised-pca" class="anchor"></a>4. Supervised PCA</h1>
<p>We now discuss and give examples of the Supervised PCA method.</p>
<div id="method-details-1" class="section level2">
<h2 class="hasAnchor">
<a href="#method-details-1" class="anchor"></a>4.1 Method Details</h2>
<div id="supervised-pca-method-sources" class="section level3">
<h3 class="hasAnchor">
<a href="#supervised-pca-method-sources" class="anchor"></a>4.1.1 Supervised PCA Method Sources</h3>
<p>While PCA is a commonly-applied <em>unsupervised</em> learning technique (i.e., response information is unnecessary), one limitation of this method is that ignoring response information may yield a first PC completely unrelated to outcome. In an effort to bolster this weakness, <a href="https://doi.org/10.1198/016214505000000628">Bair et al. (2006)</a> employed response information to rank predictors by the strength of their association. Then, they extracted PCs from feature design matrix subsets constructed from the predictors most strongly associated with the response. <a href="https://doi.org/10.1093/bioinformatics/btn458">Chen et al. (2008)</a> extend this technique to subsets of biological features within pre-defined biological pathways; they applied the Supervised PCA routine independently to each pathway in a pathway collection. <a href="https://doi.org/10.1002/gepi.20532">Chen et al. (2010)</a> built on this work, testing if pathways were significantly associated with a given biological or clinical response.</p>
</div>
<div id="calculate-pathway-specific-model-p-values-1" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-pathway-specific-model-p-values-1" class="anchor"></a>4.1.2 Calculate Pathway-Specific Model <span class="math inline">\(p\)</span>-Values</h3>
<p>As thoroughly discussed in Chen et al. (2008), the model fit and regression coefficient test statistics no longer come from their expected distributions. Necessarily, this is due to Supervised PCA’s strength in finding features already associated with outcome. Therefore, for the Supervised PCA method, pathway <span class="math inline">\(p\)</span>-values are calculated from a mixture of extreme value distributions. We use a constrained numerical optimization routine to calculate the maximum likelihood estimates of the mean, precision, and mixing proportion components of a mixture of two Gumbel extreme value distributions (for minima and maxima of a random normal sample). The <span class="math inline">\(p\)</span>-values from the pathways after permuting the response is used to estimate this null distribution, so result accuracy may be degraded for a very small set of pathways.</p>
</div>
<div id="supervised-pca-pros-and-cons" class="section level3">
<h3 class="hasAnchor">
<a href="#supervised-pca-pros-and-cons" class="anchor"></a>4.1.3 Supervised-PCA Pros and Cons</h3>
<p>Pros:</p>
<ul>
<li>The Supervised PCs are extracted without numerical optimization, so calculating the PCs for each pathway is considerably faster than calculating AES-PCs.</li>
<li>The <span class="math inline">\(p\)</span>-values are calculated parametrically, so calculating the <span class="math inline">\(p\)</span>-values is considerably faster than the non-parametric AES-PCA option, while holding better distributional properties than the parametric AES-PCA option.</li>
</ul>
<p>Cons:</p>
<ul>
<li>In rare cases, numerical routines used to find the maximum likelihood estimates for the mixture distribution needed to calculate the <span class="math inline">\(p\)</span>-values in Supervised PCA can fail to converge.</li>
<li>The Supervised PCA method cannot have missing values in the response.</li>
</ul>
</div>
</div>
<div id="supervised-pca-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#supervised-pca-examples" class="anchor"></a>4.2 Supervised PCA Examples</h2>
<div id="survival-response-1" class="section level3">
<h3 class="hasAnchor">
<a href="#survival-response-1" class="anchor"></a>4.2.1 Survival Response</h3>
<p>We will use two of our available cores with the parallel computing approach. We will adjust the <span class="math inline">\(p\)</span>-values with the Hochberg (1988) and Sidak Step-Down FWER-adjustment procedures. For the tiny <span class="math inline">\(250 \times 656\)</span> assay with 15 associated pathways, this calculation is completed in 6 seconds. If we compare this to AES-PCA at 1000 permutations, Supervised PCA is <span class="math inline">\(4.6\times\)</span> faster; for 10,000 permutations, it’s <span class="math inline">\(36.1\times\)</span> faster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonSurv_superpcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SuperPCA_pVals.html">SuperPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsSurv,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Hoch"</span>, <span class="st">"SidakSD"</span>)
)</code></pre></div>
<pre><code>#&gt; Initializing Computing Cluster: DONE
#&gt; Calculating Pathway Test Statistics in Parallel: DONE
#&gt; Calculating Pathway Critical Values in Parallel: DONE
#&gt; Calculating Pathway p-Values: DONE
#&gt; Adjusting p-Values and Sorting Pathway p-Value Data Frame: DONE</code></pre>
</div>
<div id="regression-response-1" class="section level3">
<h3 class="hasAnchor">
<a href="#regression-response-1" class="anchor"></a>4.2.2 Regression Response</h3>
<p>We can also make a mock regression data set by treating the event time as the necessary continuous response. For this example, we will adjust the <span class="math inline">\(p\)</span>-values with the Holm (1979) FWER- and Benjamini and Hochberg (1995) FDR-adjustment procedures. This calculation took 5 seconds. If we compare this to AES-PCA at 1000 permutations, Supervised PCA is <span class="math inline">\(3.4\times\)</span> faster; for 10,000 permutations, it’s <span class="math inline">\(20.7\times\)</span> faster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonReg_superpcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SuperPCA_pVals.html">SuperPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsReg,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Holm"</span>, <span class="st">"BH"</span>)
)</code></pre></div>
</div>
<div id="binary-classification-response-1" class="section level3">
<h3 class="hasAnchor">
<a href="#binary-classification-response-1" class="anchor"></a>4.2.3 Binary Classification Response</h3>
<p>Finally, we can simulate a mock classification data set by treating the event indicator as the necessary binary response. For this example, we will adjust the <span class="math inline">\(p\)</span>-values with the Sidak Single-Step FWER- and Benjamini and Yekutieli (2001) FDR-adjustment procedures. This calculation took 8 seconds. If we compare this to AES-PCA at 1000 permutations, Supervised PCA is <span class="math inline">\(3.7\times\)</span> faster; for 10,000 permutations, it’s <span class="math inline">\(27.6\times\)</span> faster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonCateg_superpcOut &lt;-<span class="st"> </span><span class="kw"><a href="../reference/SuperPCA_pVals.html">SuperPCA_pVals</a></span>(
  <span class="dt">object =</span> colon_OmicsCateg,
  <span class="dt">numPCs =</span> <span class="dv">2</span>,
  <span class="dt">parallel =</span> <span class="ot">TRUE</span>,
  <span class="dt">numCores =</span> <span class="dv">2</span>,
  <span class="dt">adjustpValues =</span> <span class="ot">TRUE</span>,
  <span class="dt">adjustment =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"SidakSS"</span>, <span class="st">"BY"</span>)
)</code></pre></div>
<hr>
<p></p>
</div>
</div>
</div>
<div id="inspect-the-results" class="section level1">
<h1 class="hasAnchor">
<a href="#inspect-the-results" class="anchor"></a>5. Inspect the Results</h1>
<p>Now that we have the pathway-specific <span class="math inline">\(p\)</span>-values, we can inspect the top pathways ordered by significance. Further, we can assess the loadings of each gene, or the first principal component, corresponding to each pathway.</p>
<div id="table-of-p-values" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-p-values" class="anchor"></a>5.1 Table of <span class="math inline">\(p\)</span>-Values</h2>
<p>For a quick and easy view of the pathway significance testing results, we can simply print the <code>pVals_df</code> data frame in the output object. (Note: if you are not using the <code>tidyverse</code> package suite, your results will print differently. Use the <code>head</code> function to print the top pathways instead.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">colonSurv_aespcOut<span class="op">$</span>pVals_df
<span class="co">#&gt; # A tibble: 15 x 6</span>
<span class="co">#&gt;    pathways  n_tested terms                 rawp FWER_Hochberg FWER_SidakSD</span>
<span class="co">#&gt;  * &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;                &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt;  1 pathway4~       40 PID_EPHB_FWD_PATH~ 6.53e-6     0.0000980    0.0000980</span>
<span class="co">#&gt;  2 pathway7~      180 REACTOME_PHOSPHOL~ 1.96e-4     0.00275      0.00275  </span>
<span class="co">#&gt;  3 pathway7~       83 REACTOME_INSULIN_~ 4.90e-4     0.00637      0.00635  </span>
<span class="co">#&gt;  4 pathway1~       26 KEGG_ASTHMA        8.21e-4     0.00985      0.00981  </span>
<span class="co">#&gt;  5 pathway87       86 KEGG_ERBB_SIGNALI~ 1.47e-3     0.0162       0.0161   </span>
<span class="co">#&gt;  6 pathway5~       44 PID_TNF_PATHWAY    2.60e-3     0.0260       0.0257   </span>
<span class="co">#&gt;  7 pathway1~      104 REACTOME_SIGNALIN~ 4.42e-3     0.0387       0.0390   </span>
<span class="co">#&gt;  8 pathway3        26 KEGG_PENTOSE_PHOS~ 4.84e-3     0.0387       0.0390   </span>
<span class="co">#&gt;  9 pathway4~       23 ST_GA12_PATHWAY    1.48e-2     0.103        0.0990   </span>
<span class="co">#&gt; 10 pathway60       45 KEGG_RETINOL_META~ 2.55e-2     0.153        0.143    </span>
<span class="co">#&gt; 11 pathway1~       54 KEGG_NON_SMALL_CE~ 4.69e-2     0.234        0.213    </span>
<span class="co">#&gt; 12 pathway1~       73 KEGG_ANTIGEN_PROC~ 7.37e-2     0.295        0.264    </span>
<span class="co">#&gt; 13 pathway1~       16 BIOCARTA_RELA_PAT~ 7.09e-1     0.801        0.975    </span>
<span class="co">#&gt; 14 pathway3~       29 BIOCARTA_TNFR1_PA~ 7.76e-1     0.801        0.975    </span>
<span class="co">#&gt; 15 pathway2~       11 BIOCARTA_SET_PATH~ 8.01e-1     0.801        0.975</span></code></pre></div>
</div>
<div id="pathway-pc-and-loading-vectors" class="section level2">
<h2 class="hasAnchor">
<a href="#pathway-pc-and-loading-vectors" class="anchor"></a>5.2 Pathway PC and Loading Vectors</h2>
<p>We also may be interested in which genes or proteins “drive” a specific pathway. We can extract the pathway-specific PCs and loadings (PC &amp; L) from either the AESPCA or Supervised PCA output with the <code><a href="../reference/getPathPCLs.html">getPathPCLs()</a></code> function. This function will take in either the proper name of a pathway (as given in the <code>terms</code> column) or the unique pathway identifier (as shown in the <code>pathways</code> column). Note that the PCs and Loadings are stored in tidy data frames, so they will have enhanced printing properties if you have the <code>tidyverse</code> package suite loaded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PCLs_ls &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getPathPCLs.html">getPathPCLs</a></span>(colonSurv_aespcOut, <span class="st">"KEGG_ASTHMA"</span>)
PCLs_ls
<span class="co">#&gt; $PCs</span>
<span class="co">#&gt; # A tibble: 250 x 3</span>
<span class="co">#&gt;    sampleID     V1     V2</span>
<span class="co">#&gt;    &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1 subj1    -1.36   1.98 </span>
<span class="co">#&gt;  2 subj2    -0.649  3.35 </span>
<span class="co">#&gt;  3 subj3     1.73  -0.822</span>
<span class="co">#&gt;  4 subj4    -0.156  7.11 </span>
<span class="co">#&gt;  5 subj5     2.11   3.70 </span>
<span class="co">#&gt;  6 subj6     0.284  1.78 </span>
<span class="co">#&gt;  7 subj7    -1.93   3.21 </span>
<span class="co">#&gt;  8 subj8     3.28   0.551</span>
<span class="co">#&gt;  9 subj9    -0.118 -0.458</span>
<span class="co">#&gt; 10 subj10   -1.56   0.266</span>
<span class="co">#&gt; # ... with 240 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Loadings</span>
<span class="co">#&gt; # A tibble: 26 x 3</span>
<span class="co">#&gt;    featureID    PC1   PC2</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt;  1 HLA-DRB4  0      0    </span>
<span class="co">#&gt;  2 HLA-DOA   0.228  0    </span>
<span class="co">#&gt;  3 HLA-DOB   0.0851 0.287</span>
<span class="co">#&gt;  4 IL3       0      0.318</span>
<span class="co">#&gt;  5 TNF       0      0.153</span>
<span class="co">#&gt;  6 CCL11     0      0    </span>
<span class="co">#&gt;  7 EPX       0      0.494</span>
<span class="co">#&gt;  8 FCER1G    0.331  0    </span>
<span class="co">#&gt;  9 MS4A2     0      0    </span>
<span class="co">#&gt; 10 HLA-DMB   0.347  0    </span>
<span class="co">#&gt; # ... with 16 more rows</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $pathway</span>
<span class="co">#&gt; [1] "pathway177"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $term</span>
<span class="co">#&gt; [1] "KEGG_ASTHMA"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $description</span>
<span class="co">#&gt; [1] NA</span></code></pre></div>
<p>As an example, we see that the HLA-DRA gene positively loads onto this pathway, and <a href="https://www.ncbi.nlm.nih.gov/pubmed/16367922">has been shown to be related to colorectal cancer</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">PCLs_ls<span class="op">$</span>Loadings <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/filter">filter</a></span>(PC1 <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">select</span>(<span class="op">-</span>PC2) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(PC1))
<span class="co">#&gt; # A tibble: 11 x 2</span>
<span class="co">#&gt;    featureID    PC1</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 HLA-DPB1  0.472 </span>
<span class="co">#&gt;  2 HLA-DRA   0.421 </span>
<span class="co">#&gt;  3 HLA-DPA1  0.372 </span>
<span class="co">#&gt;  4 HLA-DMB   0.347 </span>
<span class="co">#&gt;  5 FCER1G    0.331 </span>
<span class="co">#&gt;  6 HLA-DMA   0.256 </span>
<span class="co">#&gt;  7 IL10      0.254 </span>
<span class="co">#&gt;  8 HLA-DOA   0.228 </span>
<span class="co">#&gt;  9 CD40      0.163 </span>
<span class="co">#&gt; 10 HLA-DQA1  0.126 </span>
<span class="co">#&gt; 11 HLA-DOB   0.0851</span></code></pre></div>
<hr>
<p></p>
</div>
</div>
<div id="review" class="section level1">
<h1 class="hasAnchor">
<a href="#review" class="anchor"></a>6. Review</h1>
<p>We have has covered in this vignette:</p>
<ol style="list-style-type: decimal">
<li>The basics of the AES-PCA pathway-significance testing approach.</li>
<li>The basics of the Supervised PCA pathway-significance testing approach.</li>
<li>Applying AES-PCA or Supervised PCA to analyze survival, regression, or classification <code>Omics</code> data objects.</li>
</ol>
<p>Please read vignette chapter 5 next: <a href="https://gabrielodom.github.io/pathwayPCA/articles/Supplement5-Analyse_Results.html">Visualizing the Results</a>.</p>
<p>Here is the R session information for this vignette:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.5.2 (2018-12-20)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 7 x64 (build 7601) Service Pack 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_United States.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                          </span>
<span class="co">#&gt; [5] LC_TIME=English_United States.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] bindrcpp_0.2.2    pathwayPCA_0.99.1 forcats_0.3.0    </span>
<span class="co">#&gt;  [4] stringr_1.3.1     dplyr_0.7.8       purrr_0.3.0      </span>
<span class="co">#&gt;  [7] readr_1.3.1       tidyr_0.8.2       tibble_2.0.1     </span>
<span class="co">#&gt; [10] ggplot2_3.1.0     tidyverse_1.2.1  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] tidyselect_0.2.5   xfun_0.4           splines_3.5.2     </span>
<span class="co">#&gt;  [4] haven_2.0.0        lattice_0.20-38    colorspace_1.4-0  </span>
<span class="co">#&gt;  [7] generics_0.0.2     htmltools_0.3.6    yaml_2.2.0        </span>
<span class="co">#&gt; [10] utf8_1.1.4         survival_2.43-3    rlang_0.3.1       </span>
<span class="co">#&gt; [13] pkgdown_1.3.0.9000 pillar_1.3.1       withr_2.1.2       </span>
<span class="co">#&gt; [16] glue_1.3.0         modelr_0.1.2       readxl_1.2.0      </span>
<span class="co">#&gt; [19] bindr_0.1.1        plyr_1.8.4         cellranger_1.1.0  </span>
<span class="co">#&gt; [22] munsell_0.5.0      commonmark_1.7     gtable_0.2.0      </span>
<span class="co">#&gt; [25] rvest_0.3.2        memoise_1.1.0      evaluate_0.12     </span>
<span class="co">#&gt; [28] knitr_1.21         fansi_0.4.0        lars_1.2          </span>
<span class="co">#&gt; [31] broom_0.5.1        Rcpp_1.0.0         backports_1.1.3   </span>
<span class="co">#&gt; [34] scales_1.0.0       desc_1.2.0         jsonlite_1.6      </span>
<span class="co">#&gt; [37] fs_1.2.6           hms_0.4.2          digest_0.6.18     </span>
<span class="co">#&gt; [40] stringi_1.2.4      grid_3.5.2         rprojroot_1.3-2   </span>
<span class="co">#&gt; [43] cli_1.0.1          tools_3.5.2        magrittr_1.5      </span>
<span class="co">#&gt; [46] lazyeval_0.2.1     crayon_1.3.4       pkgconfig_2.0.2   </span>
<span class="co">#&gt; [49] Matrix_1.2-15      MASS_7.3-51.1      xml2_1.2.0        </span>
<span class="co">#&gt; [52] lubridate_1.7.4    assertthat_0.2.0   rmarkdown_1.11    </span>
<span class="co">#&gt; [55] roxygen2_6.1.1     httr_1.4.0         rstudioapi_0.9.0  </span>
<span class="co">#&gt; [58] R6_2.3.0           nlme_3.1-137       compiler_3.5.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#overview">1. Overview</a><ul class="nav nav-pills nav-stacked">
<li><a href="#outline">1.1 Outline</a></li>
      <li><a href="#load-packages">1.2 Load Packages</a></li>
      <li><a href="#load-omics-data">1.3 Load <code>Omics</code> Data</a></li>
      </ul>
</li>
      <li>
<a href="#pathway-testing-setup">2. Pathway Testing Setup</a><ul class="nav nav-pills nav-stacked">
<li><a href="#pathway-significance-testing-overview">2.1 Pathway Significance Testing Overview</a></li>
      <li><a href="#extract-pathway-pcs">2.2 Extract Pathway PCs</a></li>
      <li><a href="#test-pathway-association">2.3 Test Pathway Association</a></li>
      <li><a href="#adjust-the-pathway-p-values-for-fdr">2.4 Adjust the Pathway <span class="math inline">\(p\)</span>-Values for FDR</a></li>
      <li><a href="#output-a-sorted-data-frame-tibble">2.5 Output a Sorted Data Frame / Tibble</a></li>
      </ul>
</li>
      <li>
<a href="#aes-pca">3. AES-PCA</a><ul class="nav nav-pills nav-stacked">
<li><a href="#method-details">3.1 Method Details</a></li>
      <li><a href="#aes-pca-examples">3.2 AES-PCA Examples</a></li>
      </ul>
</li>
      <li>
<a href="#supervised-pca">4. Supervised PCA</a><ul class="nav nav-pills nav-stacked">
<li><a href="#method-details-1">4.1 Method Details</a></li>
      <li><a href="#supervised-pca-examples">4.2 Supervised PCA Examples</a></li>
      </ul>
</li>
      <li>
<a href="#inspect-the-results">5. Inspect the Results</a><ul class="nav nav-pills nav-stacked">
<li><a href="#table-of-p-values">5.1 Table of <span class="math inline">\(p\)</span>-Values</a></li>
      <li><a href="#pathway-pc-and-loading-vectors">5.2 Pathway PC and Loading Vectors</a></li>
      </ul>
</li>
      <li><a href="#review">6. Review</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://github.com/gabrielodom">Gabriel Odom</a>, James Ban, Lizhong Liu, <a href="http://biostat.med.miami.edu/people/primary-faculty/lily-wang-ph.d">Lily Wang</a>, <a href="http://biostat.med.miami.edu/people/primary-faculty/xi-steven-chen-ph.d">Steven Chen</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9000.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
